
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>6.3.3.7. gvf2bed &#8212; BEDOPS v2.4.36</title>
    <link rel="stylesheet" href="../../../../_static/better.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/custom.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../../../" src="../../../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../../../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../../../../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="next" title="6.3.3.8. psl2bed" href="psl2bed.html" />
    <link rel="prev" title="6.3.3.6. gtf2bed" href="gtf2bed.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  </head><body>
    <header id="pageheader"><h1><a href="../../../../index.html ">
        BEDOPS v2.4.36
    </a></h1></header>
  <div class="related top">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="gtf2bed.html" title="Previous document">6.3.3.6. <cite>gtf2bed</cite></a>
        </li>
        <li>
          <a href="psl2bed.html" title="Next document">6.3.3.8. <cite>psl2bed</cite></a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../../../../index.html">Home</a></li>
      <li>
        <a href="../../../reference.html">6. Reference</a>
      </li>
      <li>
        <a href="../../file-management.html">6.3. File management</a>
      </li>
      <li>
        <a href="../conversion.html">6.3.3. Data conversion</a>
      </li> 
    </ul>
  </nav>
  </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="gvf2bed">
<span id="id1"></span><h1>6.3.3.7. <cite>gvf2bed</cite><a class="headerlink" href="#gvf2bed" title="Permalink to this headline">¶</a></h1>
<p>The <code class="docutils literal notranslate"><span class="pre">gvf2bed</span></code> script converts 1-based, closed <code class="docutils literal notranslate"><span class="pre">[start,</span> <span class="pre">end]</span></code> <a class="reference external" href="http://www.sequenceontology.org/resources/gvf.html#summary">Genome Variation Format</a> (GVF, a type of <a class="reference external" href="http://www.sequenceontology.org/gff3.shtml">General Feature Format v3</a> or GFF3) to sorted, 0-based, half-open <code class="docutils literal notranslate"><span class="pre">[start-1,</span> <span class="pre">end)</span></code> extended BED-formatted data.</p>
<p>For convenience, we also offer <code class="docutils literal notranslate"><span class="pre">gvf2starch</span></code>, which performs the extra step of creating a <a class="reference internal" href="../compression/starch-specification.html#starch-specification"><span class="std std-ref">Starch-formatted</span></a> archive.</p>
<div class="section" id="dependencies">
<h2>6.3.3.7.1. Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">gvf2bed</span></code> script requires <a class="reference internal" href="convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a>. The <code class="docutils literal notranslate"><span class="pre">gvf2starch</span></code> script requires <a class="reference internal" href="../compression/starch.html#starch"><span class="std std-ref">starch</span></a>. Both dependencies are part of a typical BEDOPS installation.</p>
<p>This script is also dependent on input that follows the GFF3 specification. A GFF3-format validator is available <a class="reference external" href="http://modencode.oicr.on.ca/cgi-bin/validate_gff3_online">here</a> to ensure your input follows specification.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">Conversion of data which are GFF-like, but which do not follow the specification can cause parsing issues. If you run into problems, please check that your input follows the GFF3 specification. Tools such as the <a class="reference external" href="http://genometools.org/cgi-bin/gff3validator.cgi">GFF3 Online Validator</a> are useful for this task.</p>
</div>
</div>
<div class="section" id="source">
<h2>6.3.3.7.2. Source<a class="headerlink" href="#source" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">gvf2bed</span></code> and <code class="docutils literal notranslate"><span class="pre">gvf2starch</span></code> conversion scripts are part of the binary and source downloads of BEDOPS. See the <a class="reference internal" href="../../../installation.html#installation"><span class="std std-ref">Installation</span></a> documentation for more details.</p>
</div>
<div class="section" id="usage">
<h2>6.3.3.7.3. Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">gvf2bed</span></code> script parses GVF from standard input and prints sorted BED to standard output. The <code class="docutils literal notranslate"><span class="pre">gvf2starch</span></code> script uses an extra step to parse GVF to a compressed BEDOPS <a class="reference internal" href="../compression/starch-specification.html#starch-specification"><span class="std std-ref">Starch-formatted</span></a> archive, which is also directed to standard output.</p>
<p>The header data of a GVF file is usually discarded, unless you add the <code class="docutils literal notranslate"><span class="pre">--keep-header</span></code> option. In this case, BED elements are created from these data, using the chromosome name <code class="docutils literal notranslate"><span class="pre">_header</span></code> to denote content. Line numbers are specified in the start and stop coordinates, and unmodified header data are placed in the fourth column (ID field).</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">By default, all conversion scripts now output sorted BED data ready for use with BEDOPS utilities. If you do not want to sort converted output, use the <code class="docutils literal notranslate"><span class="pre">--do-not-sort</span></code> option. Run the script with the <code class="docutils literal notranslate"><span class="pre">--help</span></code> option for more details.</p>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">If sorting converted data larger than system memory, use the <code class="docutils literal notranslate"><span class="pre">--max-mem</span></code> option to limit sort memory usage to a reasonable fraction of available memory, <em>e.g.</em>, <code class="docutils literal notranslate"><span class="pre">--max-mem</span> <span class="pre">2G</span></code> or similar. See <code class="docutils literal notranslate"><span class="pre">--help</span></code> for more details.</p>
</div>
</div>
<div class="section" id="example">
<h2>6.3.3.7.4. Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>To demonstrate these scripts, we use a sample GVF input called <code class="docutils literal notranslate"><span class="pre">foo.gvf</span></code> (see the <a class="reference internal" href="#gvf2bed-downloads"><span class="std std-ref">Downloads</span></a> section to grab this file).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">##gvf-version 1.07</span>
<span class="c1">##feature-ontology http://www.sequenceontology.org/resources/obo_files/current_release.obo</span>
<span class="c1">##multi-individual NA19240,NA18507,NA12878,NA19238</span>
<span class="c1">##genome-build NCBI B36.3</span>
<span class="c1">##sequence-region chr16 1 88827254</span>

<span class="n">chr16</span> <span class="n">dbSNP</span>   <span class="n">SNV</span>     <span class="mi">49291360</span>        <span class="mi">49291360</span>        <span class="o">.</span>       <span class="o">+</span>       <span class="o">.</span>       <span class="n">ID</span><span class="o">=</span><span class="n">ID_2</span><span class="p">;</span><span class="n">Variant_seq</span><span class="o">=</span><span class="n">C</span><span class="p">,</span><span class="n">G</span><span class="p">;</span><span class="n">Individual</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">;</span><span class="n">Genotype</span><span class="o">=</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">;</span>
<span class="n">chr16</span> <span class="n">dbSNP</span>   <span class="n">SNV</span>     <span class="mi">49302125</span>        <span class="mi">49302125</span>        <span class="o">.</span>       <span class="o">+</span>       <span class="o">.</span>       <span class="n">ID</span><span class="o">=</span><span class="n">ID_3</span><span class="p">;</span><span class="n">Variant_seq</span><span class="o">=</span><span class="n">C</span><span class="p">,</span><span class="n">T</span><span class="p">;</span><span class="n">Individual</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">;</span><span class="n">Genotype</span><span class="o">=</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">;</span>
<span class="n">chr16</span> <span class="n">dbSNP</span>   <span class="n">SNV</span>     <span class="mi">49302365</span>        <span class="mi">49302365</span>        <span class="o">.</span>       <span class="o">+</span>       <span class="o">.</span>       <span class="n">ID</span><span class="o">=</span><span class="n">ID_4</span><span class="p">;</span><span class="n">Variant_seq</span><span class="o">=</span><span class="n">G</span><span class="p">;</span><span class="n">Individual</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">;</span><span class="n">Genotype</span><span class="o">=</span><span class="mi">0</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">:</span><span class="mi">0</span><span class="p">;</span>
<span class="n">chr16</span> <span class="n">dbSNP</span>   <span class="n">SNV</span>     <span class="mi">49302700</span>        <span class="mi">49302700</span>        <span class="o">.</span>       <span class="o">+</span>       <span class="o">.</span>       <span class="n">ID</span><span class="o">=</span><span class="n">ID_5</span><span class="p">;</span><span class="n">Variant_seq</span><span class="o">=</span><span class="n">C</span><span class="p">,</span><span class="n">T</span><span class="p">;</span><span class="n">Individual</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">;</span><span class="n">Genotype</span><span class="o">=</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">:</span><span class="mi">0</span><span class="p">;</span>
<span class="n">chr16</span> <span class="n">dbSNP</span>   <span class="n">SNV</span>     <span class="mi">49303084</span>        <span class="mi">49303084</span>        <span class="o">.</span>       <span class="o">+</span>       <span class="o">.</span>       <span class="n">ID</span><span class="o">=</span><span class="n">ID_6</span><span class="p">;</span><span class="n">Variant_seq</span><span class="o">=</span><span class="n">T</span><span class="p">,</span><span class="n">G</span><span class="p">,</span><span class="n">A</span><span class="p">;</span><span class="n">Individual</span><span class="o">=</span><span class="mi">3</span><span class="p">;</span><span class="n">Genotype</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">:;</span>
<span class="n">chr16</span> <span class="n">dbSNP</span>   <span class="n">SNV</span>     <span class="mi">49303427</span>        <span class="mi">49303427</span>        <span class="o">.</span>       <span class="o">+</span>       <span class="o">.</span>       <span class="n">ID</span><span class="o">=</span><span class="n">ID_8</span><span class="p">;</span><span class="n">Variant_seq</span><span class="o">=</span><span class="n">T</span><span class="p">;</span><span class="n">Individual</span><span class="o">=</span><span class="mi">0</span><span class="p">;</span><span class="n">Genotype</span><span class="o">=</span><span class="mi">0</span><span class="p">:</span><span class="mi">0</span><span class="p">;</span>
<span class="n">chr16</span> <span class="n">dbSNP</span>   <span class="n">SNV</span>     <span class="mi">49303596</span>        <span class="mi">49303596</span>        <span class="o">.</span>       <span class="o">+</span>       <span class="o">.</span>       <span class="n">ID</span><span class="o">=</span><span class="n">ID_9</span><span class="p">;</span><span class="n">Variant_seq</span><span class="o">=</span><span class="n">A</span><span class="p">,</span><span class="n">G</span><span class="p">,</span><span class="n">T</span><span class="p">;</span><span class="n">Individual</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">;</span><span class="n">Genotype</span><span class="o">=</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">;</span>
</pre></div>
</div>
<p>We can convert it to sorted BED data in the following manner:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ gvf2bed &lt; foo.gvf
chr16 49291359        49291360        ID_2    .       +       dbSNP   SNV     .       ID=ID_2;Variant_seq=C,G;Individual=0,1,2,3;Genotype=0:1,0:0,1:1,0:1;zero_length_insertion=True
chr16 49302124        49302125        ID_3    .       +       dbSNP   SNV     .       ID=ID_3;Variant_seq=C,T;Individual=0,1,3;Genotype=0:1,2:2,0:2;zero_length_insertion=True
chr16 49302364        49302365        ID_4    .       +       dbSNP   SNV     .       ID=ID_4;Variant_seq=G;Individual=0,1;Genotype=0:0,0:0;zero_length_insertion=True
chr16 49302699        49302700        ID_5    .       +       dbSNP   SNV     .       ID=ID_5;Variant_seq=C,T;Individual=2,3;Genotype=0:1,0:0;zero_length_insertion=True
chr16 49303083        49303084        ID_6    .       +       dbSNP   SNV     .       ID=ID_6;Variant_seq=T,G,A;Individual=3;Genotype=1,2:;zero_length_insertion=True
chr16 49303426        49303427        ID_8    .       +       dbSNP   SNV     .       ID=ID_8;Variant_seq=T;Individual=0;Genotype=0:0;zero_length_insertion=True
chr16 49303595        49303596        ID_9    .       +       dbSNP   SNV     .       ID=ID_9;Variant_seq=A,G,T;Individual=0,1,3;Genotype=1:2,3:3,1:3;zero_length_insertion=True
</pre></div>
</div>
<p>As shown, the default usage strips the leading pragmas (<code class="docutils literal notranslate"><span class="pre">##gvf-version</span> <span class="pre">1.07</span></code>, <em>etc.</em>), but adding the <code class="docutils literal notranslate"><span class="pre">--keep-header</span></code> option will preserve pragmas as BED elements that use <code class="docutils literal notranslate"><span class="pre">_header</span></code> as a chromosome name:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ gvf2bed --keep-header &lt; foo.gvf
_header       0       1       ##gvf-version 1.07
_header       1       2       ##feature-ontology http://www.sequenceontology.org/resources/obo_files/current_release.obo
_header       2       3       ##multi-individual NA19240,NA18507,NA12878,NA19238
_header       3       4       ##genome-build NCBI B36.3
_header       4       5       ##sequence-region chr16 1 88827254
chr16 49291359        49291360        ID_2    .       +       dbSNP   SNV     .       ID=ID_2;Variant_seq=C,G;Individual=0,1,2,3;Genotype=0:1,0:0,1:1,0:1;zero_length_insertion=True
chr16 49302124        49302125        ID_3    .       +       dbSNP   SNV     .       ID=ID_3;Variant_seq=C,T;Individual=0,1,3;Genotype=0:1,2:2,0:2;zero_length_insertion=True
chr16 49302364        49302365        ID_4    .       +       dbSNP   SNV     .       ID=ID_4;Variant_seq=G;Individual=0,1;Genotype=0:0,0:0;zero_length_insertion=True
chr16 49302699        49302700        ID_5    .       +       dbSNP   SNV     .       ID=ID_5;Variant_seq=C,T;Individual=2,3;Genotype=0:1,0:0;zero_length_insertion=True
chr16 49303083        49303084        ID_6    .       +       dbSNP   SNV     .       ID=ID_6;Variant_seq=T,G,A;Individual=3;Genotype=1,2:;zero_length_insertion=True
chr16 49303426        49303427        ID_8    .       +       dbSNP   SNV     .       ID=ID_8;Variant_seq=T;Individual=0;Genotype=0:0;zero_length_insertion=True
chr16 49303595        49303596        ID_9    .       +       dbSNP   SNV     .       ID=ID_9;Variant_seq=A,G,T;Individual=0,1,3;Genotype=1:2,3:3,1:3;zero_length_insertion=True
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Zero-length insertion elements are given an extra attribute called <code class="docutils literal notranslate"><span class="pre">zero_length_insertion</span></code> which lets a BED-to-GVF or other parser know that the element will require conversion back to a right-closed element <code class="docutils literal notranslate"><span class="pre">[a,</span> <span class="pre">b]</span></code>, where <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">b</span></code> are equal.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Note the conversion from 1- to 0-based coordinate indexing, in the transition from GVF to BED. <em>BEDOPS supports operations on input with any coordinate indexing</em>, but the coordinate change made here is believed to be convenient for most end users.</p>
</div>
</div>
<div class="section" id="column-mapping">
<span id="gvf2bed-column-mapping"></span><h2>6.3.3.7.5. Column mapping<a class="headerlink" href="#column-mapping" title="Permalink to this headline">¶</a></h2>
<p>In this section, we describe how GVF columns are mapped to BED columns. We start with the first six UCSC BED columns as follows:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="33%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">GVF field</th>
<th class="head">BED column index</th>
<th class="head">BED field</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>seqid</td>
<td>1</td>
<td>chromosome</td>
</tr>
<tr class="row-odd"><td>start</td>
<td>2</td>
<td>start</td>
</tr>
<tr class="row-even"><td>end</td>
<td>3</td>
<td>stop</td>
</tr>
<tr class="row-odd"><td>ID (via attributes)</td>
<td>4</td>
<td>id</td>
</tr>
<tr class="row-even"><td>score</td>
<td>5</td>
<td>score</td>
</tr>
<tr class="row-odd"><td>strand</td>
<td>6</td>
<td>strand</td>
</tr>
</tbody>
</table>
<p>The remaining columns are mapped as follows:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="33%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">GVF field</th>
<th class="head">BED column index</th>
<th class="head">BED field</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>source</td>
<td>7</td>
<td>&#160;</td>
</tr>
<tr class="row-odd"><td>type</td>
<td>8</td>
<td>&#160;</td>
</tr>
<tr class="row-even"><td>phase</td>
<td>9</td>
<td>&#160;</td>
</tr>
<tr class="row-odd"><td>attributes</td>
<td>10</td>
<td>&#160;</td>
</tr>
</tbody>
</table>
<p>When we encounter zero-length insertion elements (which are defined where the <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">stop</span></code> GVF field values are equivalent), the <code class="docutils literal notranslate"><span class="pre">start</span></code> coordinate is decremented to convert to 0-based, half-open indexing, and a <code class="docutils literal notranslate"><span class="pre">zero_length_insertion</span></code> attribute is added to the <code class="docutils literal notranslate"><span class="pre">attributes</span></code> field value.</p>
</div>
<div class="section" id="downloads">
<span id="gvf2bed-downloads"></span><h2>6.3.3.7.6. Downloads<a class="headerlink" href="#downloads" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Sample GVF dataset: <a class="reference download internal" download="" href="../../../../_downloads/678cfd12e0270cf861a63125ad17c9cb/reference_gvf2bed_foo.gvf"><code class="xref download docutils literal notranslate"><span class="pre">foo.gvf</span></code></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../../../index.html">
              <img class="logo" src="../../../../_static/logo_with_label_v3.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../../../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">6.3.3.7. <cite>gvf2bed</cite></a><ul>
<li><a class="reference internal" href="#dependencies">6.3.3.7.1. Dependencies</a></li>
<li><a class="reference internal" href="#source">6.3.3.7.2. Source</a></li>
<li><a class="reference internal" href="#usage">6.3.3.7.3. Usage</a></li>
<li><a class="reference internal" href="#example">6.3.3.7.4. Example</a></li>
<li><a class="reference internal" href="#column-mapping">6.3.3.7.5. Column mapping</a></li>
<li><a class="reference internal" href="#downloads">6.3.3.7.6. Downloads</a></li>
</ul>
</li>
</ul>

<form class="search" action="../../../../search.html" method="get">
  <input type="text" name="q"
   placeholder="type to search" />
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
  <div class="related bottom">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="gtf2bed.html" title="Previous document">6.3.3.6. <cite>gtf2bed</cite></a>
        </li>
        <li>
          <a href="psl2bed.html" title="Next document">6.3.3.8. <cite>psl2bed</cite></a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../../../../index.html">Home</a></li>
      <li>
        <a href="../../../reference.html">6. Reference</a>
      </li>
      <li>
        <a href="../../file-management.html">6.3. File management</a>
      </li>
      <li>
        <a href="../conversion.html">6.3.3. Data conversion</a>
      </li> 
    </ul>
  </nav>
  </div>
  <footer id="pagefooter">&copy; 2011-2019, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a>
      1.8.5
        with the <a href="http://github.com/irskep/sphinx-better-theme">
          better</a> theme.

  </footer>

  
  </body>
</html>