<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>6.3.3.4. psl2bed &mdash; BEDOPS 2.4.0 documentation</title>
    
    <link rel="stylesheet" href="../../../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../../',
        VERSION:     '2.4.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../../_static/sidebar.js"></script>
    <link rel="shortcut icon" href="../../../../_static/favicon.ico"/>
    <link rel="top" title="BEDOPS 2.4.0 documentation" href="../../../../index.html" />
    <link rel="up" title="6.3.3. Data conversion" href="../conversion.html" />
    <link rel="next" title="6.3.3.5. sam2bed" href="sam2bed.html" />
    <link rel="prev" title="6.3.3.3. gtf2bed" href="gtf2bed.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="sam2bed.html" title="6.3.3.5. sam2bed"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="gtf2bed.html" title="6.3.3.3. gtf2bed"
             accesskey="P">previous</a> |</li>
        <li><a href="../../../../index.html">BEDOPS 2.4.0 documentation</a> &raquo;</li>
          <li><a href="../../../reference.html" >6. Reference</a> &raquo;</li>
          <li><a href="../../file-management.html" >6.3. File management</a> &raquo;</li>
          <li><a href="../conversion.html" accesskey="U">6.3.3. Data conversion</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="psl2bed">
<span id="id1"></span><h1>6.3.3.4. <cite>psl2bed</cite><a class="headerlink" href="#psl2bed" title="Permalink to this headline">¶</a></h1>
<p>The <tt class="docutils literal"><span class="pre">psl2bed</span></tt> script converts 0-based, half-open <tt class="docutils literal"><span class="pre">[start-1,</span> <span class="pre">end)</span></tt> <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format2">Pattern Space Layout</a> (PSL) to unsorted, 0-based, half-open <tt class="docutils literal"><span class="pre">[start-1,</span> <span class="pre">end)</span></tt> extended BED-formatted data.</p>
<p>For convenience, we also offer <tt class="docutils literal"><span class="pre">psl2starch</span></tt>, which performs the extra step of creating a Starch-formatted archive.</p>
<div class="section" id="dependencies">
<h2>6.3.3.4.1. Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">psl2bed</span></tt> script requires Python, version 2.7 or greater (and less than Python3).</p>
<p>This script is also dependent on input that follows the <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format2">PSL specification</a>.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">Conversion of data which are PSL-like, but which do not follow the specification can cause <tt class="docutils literal"><span class="pre">IOError</span></tt> and other runtime exceptions. If you run into problems, please check that your input follows the PSL specification.</p>
</div>
</div>
<div class="section" id="source">
<h2>6.3.3.4.2. Source<a class="headerlink" href="#source" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">psl2bed</span></tt> and <tt class="docutils literal"><span class="pre">psl2starch</span></tt> conversion scripts are part of the binary and source downloads of BEDOPS. See the <a class="reference internal" href="../../../installation.html#installation"><em>Installation</em></a> documentation for more details.</p>
</div>
<div class="section" id="usage">
<h2>6.3.3.4.3. Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">psl2bed</span></tt> script parses PSL from standard input and prints sorted BED to standard output. The <tt class="docutils literal"><span class="pre">psl2starch</span></tt> script uses an extra step to parse GFF to a compressed BEDOPS <a class="reference internal" href="../compression/starch-specification.html#starch-specification"><em>Starch-formatted</em></a> archive, which is also directed to standard output.</p>
<p>The header data of a headered PSL file is usually discarded, unless you add the <tt class="docutils literal"><span class="pre">--keep-header</span></tt> option. In this case, BED elements are created from these data, using the chromosome name <tt class="docutils literal"><span class="pre">_header</span></tt> to denote content. Line numbers are specified in the start and stop coordinates, and unmodified header data are placed in the fourth column (ID field).</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">By default, all conversion scripts now output sorted BED data ready for use with BEDOPS utilities. If you do not want to sort converted output, use the <tt class="docutils literal"><span class="pre">--do-not-sort</span></tt> option. Run the script with the <tt class="docutils literal"><span class="pre">--help</span></tt> option for more details.</p>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">If you are sorting data larger than system memory, use the <tt class="docutils literal"><span class="pre">--max-mem</span></tt> option to limit sort memory usage to a reasonable fraction of available memory, <em>e.g.</em>, <tt class="docutils literal"><span class="pre">--max-mem</span> <span class="pre">2G</span></tt> or similar. See <tt class="docutils literal"><span class="pre">--help</span></tt> for more details.</p>
</div>
</div>
<div class="section" id="example">
<h2>6.3.3.4.4. Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>To demonstrate these scripts, we use a sample GFF input called <tt class="docutils literal"><span class="pre">foo.gff</span></tt> (see the <a class="reference internal" href="gff2bed.html#gff2bed-downloads"><em>Downloads</em></a> section to grab this file).</p>
<div class="highlight-python"><pre>psLayout version 3

match   mis-    rep.    N's     Q gap   Q gap   T gap   T gap   strand  Q               Q       Q       Q       T               T       T       T       block   blockSizes      qStarts  tStarts
        match   match           count   bases   count   bases           name            size    start   end     name            size    start   end     count
---------------------------------------------------------------------------------------------------------------------------------------------------------------
35      0       0       0       0       0       0       0       +       foo     50      15      50      chrX    155270560       40535836        40535871        1       35,     15,     40535836,
34      2       0       0       0       0       0       0       +       foo     50      14      50      chrX    155270560       68019028        68019064        1       36,     14,     68019028,
33      2       0       0       0       0       0       0       +       foo     50      14      49      chrX    155270560       43068135        43068170        1       35,     14,     43068135,
35      2       0       0       0       0       0       0       +       foo     50      13      50      chr8    146364022       131572122       131572159       1       37,     13,     131572122,
30      0       0       0       0       0       0       0       +       foo     50      14      44      chr6    171115067       127685756       127685786       1       30,     14,     127685756,
30      0       0       0       0       0       0       0       +       foo     50      14      44      chr6    171115067       93161871        93161901        1       30,     14,     93161871,
31      0       0       0       0       0       0       0       +       foo     50      13      44      chr5    180915260       119897315       119897346       1       31,     13,     119897315,
30      0       0       0       0       0       0       0       +       foo     50      14      44      chr5    180915260       123254725       123254755       1       30,     14,     123254725,
...</pre>
</div>
<p>We can convert it to sorted BED data in the following manner:</p>
<div class="highlight-python"><pre>$ psl2bed --headered &lt; foo.psl
chr1    30571100        30571135        foo     50      -       35      0       0       0       0       0       0       0       15      50      249250621       1       35,     0,      30571100,
chr1    69592160        69592195        foo     50      -       34      1       0       0       0       0       0       0       15      50      249250621       1       35,     0,      69592160,
chr1    107200050       107200100       foo     50      +       50      0       0       0       0       0       0       0       0       50      249250621       1       50,     0,      107200050,
chr11   12618347        12618389        foo     50      +       39      3       0       0       0       0       0       0       8       50      135006516       1       42,     8,      12618347,
chr11   32933028        32933063        foo     50      +       35      0       0       0       1       1       0       0       8       44      135006516       2       4,31,   8,13,   32933028,32933032,
chr11   80116421        80116457        foo     50      +       35      1       0       0       0       0       0       0       14      50      135006516       1       36,     14,     80116421,
chr11   133952291       133952327       foo     50      +       34      2       0       0       0       0       0       0       14      50      135006516       1       36,     14,     133952291,
chr13   99729482        99729523        foo     50      +       39      2       0       0       0       0       0       0       8       49      115169878       1       41,     8,      99729482,
chr13   111391852       111391888       foo     50      +       34      2       0       0       0       0       0       0       14      50      115169878       1       36,     14,     111391852,
chr16   8149657 8149694 foo     50      +       36      1       0       0       0       0       0       0       13      50      90354753        1       37,     13,     8149657,
...</pre>
</div>
<p>As you see here, the header data of a headered PSL file is discarded, unless you add the <tt class="docutils literal"><span class="pre">--keep-header</span></tt> option. In this case, BED elements are created from these data, using the chromosome name <tt class="docutils literal"><span class="pre">_header</span></tt> to denote content. Line numbers are specified in the start and stop coordinates, and unmodified header data are placed in the fourth column (ID field).</p>
<p>Here is a demonstration of conversion of the same headered input, adding the <tt class="docutils literal"><span class="pre">--keep-header</span></tt> option:</p>
<div class="highlight-python"><pre>$ psl2bed --headered --keep-header &lt; foo.psl
_header 0       1       psLayout version 3
_header 1       2
_header 2       3       match   mis-    rep.    N's     Q gap   Q gap   T gap   T gap   strand  Q               Q       Q       Q       T               T       T       T       block   blockSizes      qStarts  tStarts
_header 3       4       match   match           count   bases   count   bases           name            size    start   end     name            size    start   end     count
_header 4       5       ---------------------------------------------------------------------------------------------------------------------------------------------------------------
chr1    30571100        30571135        foo     50      -       35      0       0       0       0       0       0       0       15      50      249250621       1       35,     0,      30571100,
chr1    69592160        69592195        foo     50      -       34      1       0       0       0       0       0       0       15      50      249250621       1       35,     0,      69592160,
chr1    107200050       107200100       foo     50      +       50      0       0       0       0       0       0       0       0       50      249250621       1       50,     0,      107200050,
chr11   12618347        12618389        foo     50      +       39      3       0       0       0       0       0       0       8       50      135006516       1       42,     8,      12618347,
chr11   32933028        32933063        foo     50      +       35      0       0       0       1       1       0       0       8       44      135006516       2       4,31,   8,13,   32933028,32933032,
chr11   80116421        80116457        foo     50      +       35      1       0       0       0       0       0       0       14      50      135006516       1       36,     14,     80116421,
chr11   133952291       133952327       foo     50      +       34      2       0       0       0       0       0       0       14      50      135006516       1       36,     14,     133952291,
chr13   99729482        99729523        foo     50      +       39      2       0       0       0       0       0       0       8       49      115169878       1       41,     8,      99729482,
chr13   111391852       111391888       foo     50      +       34      2       0       0       0       0       0       0       14      50      115169878       1       36,     14,     111391852,
chr16   8149657 8149694 foo     50      +       36      1       0       0       0       0       0       0       13      50      90354753        1       37,     13,     8149657,
...</pre>
</div>
<p>With this option, the <tt class="docutils literal"><span class="pre">psl2bed</span></tt> and <tt class="docutils literal"><span class="pre">psl2starch</span></tt> scripts are completely &#8220;non-lossy&#8221;. Use of <tt class="docutils literal"><span class="pre">awk</span></tt> or other scripting tools can munge these data back into a PSL-formatted file.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">By default, the <tt class="docutils literal"><span class="pre">psl2bed</span></tt> and <tt class="docutils literal"><span class="pre">psl2starch</span></tt> scripts work with headerless PSL data. If you have headered PSL output, use the <tt class="docutils literal"><span class="pre">--headered</span></tt> operator with either conversion script, as shown in the example above.</p>
</div>
</div>
<div class="section" id="column-mapping">
<span id="psl2bed-column-mapping"></span><h2>6.3.3.4.5. Column mapping<a class="headerlink" href="#column-mapping" title="Permalink to this headline">¶</a></h2>
<p>In this section, we describe how PSL columns are mapped to BED columns. We start with the first six UCSC BED columns as follows:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="33%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">PSL field</th>
<th class="head">BED column index</th>
<th class="head">BED field</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>tName</td>
<td>1</td>
<td>chromosome</td>
</tr>
<tr class="row-odd"><td>tStart</td>
<td>2</td>
<td>start</td>
</tr>
<tr class="row-even"><td>tEnd</td>
<td>3</td>
<td>stop</td>
</tr>
<tr class="row-odd"><td>qName</td>
<td>4</td>
<td>id</td>
</tr>
<tr class="row-even"><td>qSize</td>
<td>5</td>
<td>score</td>
</tr>
<tr class="row-odd"><td>strand</td>
<td>6</td>
<td>strand</td>
</tr>
</tbody>
</table>
<p>The remaining PSL columns are mapped, in order, to the remaining columns of the BED output:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="33%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">PSL field</th>
<th class="head">BED column index</th>
<th class="head">BED field</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>matches</td>
<td>7</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>misMatches</td>
<td>8</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>repMatches</td>
<td>9</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>nCount</td>
<td>10</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>qNumInsert</td>
<td>11</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>qBaseInsert</td>
<td>12</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>tNumInsert</td>
<td>13</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>tBaseInsert</td>
<td>14</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>qStart</td>
<td>15</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>qEnd</td>
<td>16</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>tSize</td>
<td>17</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>blockCount</td>
<td>18</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>blockSizes</td>
<td>19</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>qStarts</td>
<td>20</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>tStarts</td>
<td>21</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
<p>This is a lossless mapping. Because we have mapped all columns, we can translate converted BED data back to headerless PSL with a simple <tt class="docutils literal"><span class="pre">awk</span></tt> statement that permutes columns to PSL-based ordering:</p>
<div class="highlight-python"><pre>$ awk 'BEGIN { OFS = "\t" } { print $7" "$8" "$9" "$10" "$11" "$12" "$13" "$14" "$6" "$4" "$5" "$15" "$16" "$1" "$17" "$2" "$3" "$18" "$19" "$20" "$21 }' converted.bed &gt; original.psl</pre>
</div>
</div>
<div class="section" id="downloads">
<span id="psl2bed-downloads"></span><h2>6.3.3.4.6. Downloads<a class="headerlink" href="#downloads" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Sample PSL dataset: <a class="reference download internal" href="../../../../_downloads/reference_psl2bed_foo.psl"><tt class="xref download docutils literal"><span class="pre">foo.psl</span></tt></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../../../index.html">
              <img class="logo" src="../../../../_static/logo_with_label_v2.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">6.3.3.4. <cite>psl2bed</cite></a><ul>
<li><a class="reference internal" href="#dependencies">6.3.3.4.1. Dependencies</a></li>
<li><a class="reference internal" href="#source">6.3.3.4.2. Source</a></li>
<li><a class="reference internal" href="#usage">6.3.3.4.3. Usage</a></li>
<li><a class="reference internal" href="#example">6.3.3.4.4. Example</a></li>
<li><a class="reference internal" href="#column-mapping">6.3.3.4.5. Column mapping</a></li>
<li><a class="reference internal" href="#downloads">6.3.3.4.6. Downloads</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="gtf2bed.html"
                        title="previous chapter">6.3.3.3. <cite>gtf2bed</cite></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="sam2bed.html"
                        title="next chapter">6.3.3.5. <cite>sam2bed</cite></a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../../../../_sources/content/reference/file-management/conversion/psl2bed.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="sam2bed.html" title="6.3.3.5. sam2bed"
             >next</a> |</li>
        <li class="right" >
          <a href="gtf2bed.html" title="6.3.3.3. gtf2bed"
             >previous</a> |</li>
        <li><a href="../../../../index.html">BEDOPS 2.4.0 documentation</a> &raquo;</li>
          <li><a href="../../../reference.html" >6. Reference</a> &raquo;</li>
          <li><a href="../../file-management.html" >6.3. File management</a> &raquo;</li>
          <li><a href="../conversion.html" >6.3.3. Data conversion</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.
    </div>
  </body>
</html>