<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>6.3.3.8. sam2bed &mdash; BEDOPS 2.4.4 documentation</title>
    
    <link rel="stylesheet" href="../../../../_static/better.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/test.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../../',
        VERSION:     '2.4.4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../../../../_static/favicon.ico"/>
    <link rel="top" title="BEDOPS 2.4.4 documentation" href="../../../../index.html" />
    <link rel="up" title="6.3.3. Data conversion" href="../conversion.html" />
    <link rel="next" title="6.3.3.9. vcf2bed" href="vcf2bed.html" />
    <link rel="prev" title="6.3.3.7. psl2bed" href="psl2bed.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link rel="stylesheet" href="../../../../_static/custom.css" type="text/css" />
  </head>
  <body>
    <header id="pageheader"><h1><a href="../../../../index.html ">
        BEDOPS 2.4.4 documentation
    </a></h1></header>
  <div class="related top">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="psl2bed.html" title="Previous document">6.3.3.7. <cite>psl2bed</cite></a>
        </li>
        <li>
          <a href="vcf2bed.html" title="Next document">6.3.3.9. <cite>vcf2bed</cite></a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../../../../index.html">BEDOPS 2.4.4 documentation</a></li>
      <li>
        <a href="../../../reference.html">6. Reference</a>
      </li>
      <li>
        <a href="../../file-management.html">6.3. File management</a>
      </li>
      <li>
        <a href="../conversion.html">6.3.3. Data conversion</a>
      </li> 
    </ul>
  </nav>
  </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="sam2bed">
<span id="id1"></span><h1>6.3.3.8. <cite>sam2bed</cite><a class="headerlink" href="#sam2bed" title="Permalink to this headline">¶</a></h1>
<p>The <tt class="docutils literal"><span class="pre">sam2bed</span></tt> script converts 1-based, closed <tt class="docutils literal"><span class="pre">[start,</span> <span class="pre">end]</span></tt> <a class="reference external" href="http://samtools.sourceforge.net/">Sequence Alignment/Map</a> (SAM) to sorted, 0-based, half-open <tt class="docutils literal"><span class="pre">[start-1,</span> <span class="pre">end)</span></tt> UCSC BED data.</p>
<p>For convenience, we also offer <tt class="docutils literal"><span class="pre">sam2starch</span></tt>, which performs the extra step of creating a <a class="reference internal" href="../compression/starch-specification.html#starch-specification"><em>Starch-formatted</em></a> archive.</p>
<p>The <tt class="docutils literal"><span class="pre">sam2bed</span></tt> script is &#8220;non-lossy&#8221; (with the use of specific options, described below). Other toolkits tend to throw out information from the original SAM input upon conversion; <tt class="docutils literal"><span class="pre">sam2bed</span></tt> retains everything, facilitating reuse of converted data and conversion to other formats.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">Doing the extra step of creating a <a class="reference internal" href="../compression/starch-specification.html#starch-specification"><em>Starch-formatted</em></a> archive can save a lot of space relative to the original SAM format, up to 33% of the original SAM dataset, while offering per-chromosome random access.</p>
</div>
<div class="section" id="dependencies">
<h2>6.3.3.8.1. Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">sam2bed</span></tt> wrapper script is dependent upon the installation of <a class="reference external" href="http://samtools.sourceforge.net/">SAMtools</a> and <a class="reference internal" href="convert2bed.html#convert2bed"><em>convert2bed</em></a>. The <tt class="docutils literal"><span class="pre">sam2starch</span></tt> wrapper script is further dependent on the installation of the <a class="reference internal" href="../compression/starch.html#starch"><em>starch</em></a> binary, part of a typical BEDOPS installation.</p>
</div>
<div class="section" id="source">
<h2>6.3.3.8.2. Source<a class="headerlink" href="#source" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">sam2bed</span></tt> and <tt class="docutils literal"><span class="pre">sam2starch</span></tt> conversion scripts are part of the binary and source downloads of BEDOPS. See the <a class="reference internal" href="../../../installation.html#installation"><em>Installation</em></a> documentation for more details.</p>
</div>
<div class="section" id="usage">
<h2>6.3.3.8.3. Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">sam2bed</span></tt> script parses SAM data from standard input and prints <a class="reference internal" href="../sorting/sort-bed.html#sort-bed"><em>sorted</em></a> BED to standard output. The <tt class="docutils literal"><span class="pre">sam2starch</span></tt> script uses an extra step to parse SAM to a compressed BEDOPS <a class="reference internal" href="../compression/starch-specification.html#starch-specification"><em>Starch-formatted</em></a> archive, which is also directed to standard output.</p>
<p>The header data of a SAM file is usually discarded, unless you add the <tt class="docutils literal"><span class="pre">--keep-header</span></tt> option. In this case, BED elements are created from these data, using the chromosome name <tt class="docutils literal"><span class="pre">_header</span></tt> to denote content. Line numbers are specified in the start and stop coordinates, and unmodified header data are placed in the fourth column (ID field).</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">If you work with RNA-seq data, you can use the <tt class="docutils literal"><span class="pre">--split</span></tt> option to process reads with <tt class="docutils literal"><span class="pre">N</span></tt>-CIGAR operations, splitting them into separate BED elements.</p>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">By default, all conversion scripts now output sorted BED data ready for use with BEDOPS utilities. If you do not want to sort converted output, use the <tt class="docutils literal"><span class="pre">--do-not-sort</span></tt> option. Run the script with the <tt class="docutils literal"><span class="pre">--help</span></tt> option for more details.</p>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">If sorting converted data larger than system memory, use the <tt class="docutils literal"><span class="pre">--max-mem</span></tt> option to limit sort memory usage to a reasonable fraction of available memory, <em>e.g.</em>, <tt class="docutils literal"><span class="pre">--max-mem</span> <span class="pre">2G</span></tt> or similar. See <tt class="docutils literal"><span class="pre">--help</span></tt> for more details.</p>
</div>
</div>
<div class="section" id="example">
<h2>6.3.3.8.4. Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>To demonstrate these scripts, we use a sample binary input called <tt class="docutils literal"><span class="pre">foo.sam</span></tt> (see the <a class="reference internal" href="#sam2bed-downloads"><em>Downloads</em></a> section to grab this file).</p>
<div class="highlight-python"><div class="highlight"><pre>@HD     VN:1.0 SO:coordinate
@SQ     SN:seq1 LN:5000
@SQ     SN:seq2 LN:5000
@CO     Example of SAM/BAM file format.
B7_591:4:96:693:509     73      seq1    1       99      36M     *       0       0       CACTAGTGGCTCATTGTAAATGTGTGGTTTAACTCG    &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;5&lt;&lt;&lt;&lt;&lt;;:&lt;;7    MF:i:18 Aq:i:73 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
EAS54_65:7:152:368:113  73      seq1    3       99      35M     *       0       0       CTAGTGGCTCATTGTAAATGTGTGGTTTAACTCGT     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;0&lt;&lt;&lt;&lt;655&lt;&lt;7&lt;&lt;&lt;:9&lt;&lt;3/:&lt;6):     MF:i:18 Aq:i:66 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
EAS51_64:8:5:734:57     137     seq1    5       99      35M     *       0       0       AGTGGCTCATTGTAAATGTGTGGTTTAACTCGTCC     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;7;71&lt;&lt;;&lt;;;&lt;7;&lt;&lt;3;);3*8/5     MF:i:18 Aq:i:66 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
...
</pre></div>
</div>
<p>We can convert it to sorted BED data in the following manner (omitting standard error messages):</p>
<div class="highlight-python"><div class="highlight"><pre>$ sam2bed &lt; foo.sam
seq1    0       36      B7_591:4:96:693:509     99      +       73      36M     *       0       0       CACTAGTGGCTCATTGTAAATGTGTGGTTTAACTCG    &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;5&lt;&lt;&lt;&lt;&lt;;:&lt;;7    MF:i:18 Aq:i:73 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
seq1    2       37      EAS54_65:7:152:368:113  99      +       73      35M     *       0       0       CTAGTGGCTCATTGTAAATGTGTGGTTTAACTCGT     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;0&lt;&lt;&lt;&lt;655&lt;&lt;7&lt;&lt;&lt;:9&lt;&lt;3/:&lt;6):     MF:i:18 Aq:i:66 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
seq1    4       39      EAS51_64:8:5:734:57     99      +       137      35M     *       0       0       AGTGGCTCATTGTAAATGTGTGGTTTAACTCGTCC     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;7;71&lt;&lt;;&lt;;;&lt;7;&lt;&lt;3;);3*8/5     MF:i:18 Aq:i:66 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
seq1    5       41      B7_591:1:289:587:906    63      +       137      36M     *       0       0       GTGGCTCATTGTAATTTTTTGTTTTAACTCTTCTCT    (-&amp;----,----)-)-),&#39;--)---&#39;,+-,),&#39;&#39;*,    MF:i:130        Aq:i:63 NM:i:5  UQ:i:38 H0:i:0  H1:i:0
...
</pre></div>
</div>
<p>Note also that we strip the header section from the output. If we want to keep this, the use of the <tt class="docutils literal"><span class="pre">--keep-header</span></tt> option will preserve the BAM file&#8217;s header, turning it into BED elements that use <tt class="docutils literal"><span class="pre">_header</span></tt> as a chromosome name.</p>
<p>Here&#8217;s an example:</p>
<div class="highlight-python"><div class="highlight"><pre>$ sam2bed --keep-header &lt; foo.sam
_header 0       1       @HD     VN:1.0 SO:coordinate
_header 1       2       @SQ     SN:seq1 LN:5000
_header 2       3       @SQ     SN:seq2 LN:5000
_header 3       4       @CO     Example of SAM/BAM file format.
seq1    0       36      B7_591:4:96:693:509     99      +       73      36M     *       0       0       CACTAGTGGCTCATTGTAAATGTGTGGTTTAACTCG    &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;5&lt;&lt;&lt;&lt;&lt;;:&lt;;7    MF:i:18 Aq:i:73 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
seq1    2       37      EAS54_65:7:152:368:113  99      +       73      35M     *       0       0       CTAGTGGCTCATTGTAAATGTGTGGTTTAACTCGT     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;0&lt;&lt;&lt;&lt;655&lt;&lt;7&lt;&lt;&lt;:9&lt;&lt;3/:&lt;6):     MF:i:18 Aq:i:66 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
seq1    4       39      EAS51_64:8:5:734:57     99      +       137      35M     *       0       0       AGTGGCTCATTGTAAATGTGTGGTTTAACTCGTCC     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;7;71&lt;&lt;;&lt;;;&lt;7;&lt;&lt;3;);3*8/5     MF:i:18 Aq:i:66 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
seq1    5       41      B7_591:1:289:587:906    63      +       137      36M     *       0       0       GTGGCTCATTGTAATTTTTTGTTTTAACTCTTCTCT    (-&amp;----,----)-)-),&#39;--)---&#39;,+-,),&#39;&#39;*,    MF:i:130        Aq:i:63 NM:i:5  UQ:i:38 H0:i:0  H1:i:0
...
</pre></div>
</div>
<p>With this option, the <tt class="docutils literal"><span class="pre">sam2bed</span></tt> and <tt class="docutils literal"><span class="pre">sam2starch</span></tt> scripts are completely &#8220;non-lossy&#8221; (with the exception of unmapped reads; see note below). Use of <tt class="docutils literal"><span class="pre">awk</span></tt> or other scripting tools can munge these data back into a SAM-formatted file.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The provided scripts <strong>strip out unmapped reads</strong> from the SAM file. We believe this makes sense under most circumstances. Add the <tt class="docutils literal"><span class="pre">--all-reads</span></tt> option if you need unmapped and mapped reads.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Note the conversion from 1- to 0-based coordinates. While BEDOPS fully supports 0- and 1-based coordinates, the coordinate change in BED is believed to be convenient to most end users.</p>
</div>
</div>
<div class="section" id="column-mapping">
<span id="sam2bed-column-mapping"></span><h2>6.3.3.8.5. Column mapping<a class="headerlink" href="#column-mapping" title="Permalink to this headline">¶</a></h2>
<p>In this section, we describe how SAM columns are mapped to BED columns. We start with the first six UCSC BED columns as follows:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="33%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">SAM field</th>
<th class="head">BED column index</th>
<th class="head">BED field</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>RNAME</td>
<td>1</td>
<td>chromosome</td>
</tr>
<tr class="row-odd"><td>POS - 1</td>
<td>2</td>
<td>start</td>
</tr>
<tr class="row-even"><td>POS + length(CIGAR) - 1</td>
<td>3</td>
<td>stop</td>
</tr>
<tr class="row-odd"><td>QNAME</td>
<td>4</td>
<td>id</td>
</tr>
<tr class="row-even"><td>MAPQ</td>
<td>5</td>
<td>score</td>
</tr>
<tr class="row-odd"><td>16 &amp; FLAG</td>
<td>6</td>
<td>strand</td>
</tr>
</tbody>
</table>
<p>The remaining SAM columns are mapped as-is, in same order, to adjacent BED columns:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="33%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">SAM field</th>
<th class="head">BED column index</th>
<th class="head">BED field</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>FLAG</td>
<td>7</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>CIGAR</td>
<td>8</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>RNEXT</td>
<td>9</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>PNEXT</td>
<td>10</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>TLEN</td>
<td>11</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>SEQ</td>
<td>12</td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>QUAL</td>
<td>13</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
<p>Because we have mapped all columns, we can translate converted BED data back to headered or headerless SAM reads with a simple <tt class="docutils literal"><span class="pre">awk</span></tt> statement (or other script) that reverts back to 1-based coordinates and permutes columns to SAM-based ordering.</p>
</div>
<div class="section" id="downloads">
<span id="sam2bed-downloads"></span><h2>6.3.3.8.6. Downloads<a class="headerlink" href="#downloads" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Sample SAM dataset: <a class="reference download internal" href="../../../../_downloads/reference_sam2bed_foo.sam"><tt class="xref download docutils literal"><span class="pre">foo.sam</span></tt></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../../../index.html">
              <img class="logo" src="../../../../_static/logo_with_label_v3.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">6.3.3.8. <cite>sam2bed</cite></a><ul>
<li><a class="reference internal" href="#dependencies">6.3.3.8.1. Dependencies</a></li>
<li><a class="reference internal" href="#source">6.3.3.8.2. Source</a></li>
<li><a class="reference internal" href="#usage">6.3.3.8.3. Usage</a></li>
<li><a class="reference internal" href="#example">6.3.3.8.4. Example</a></li>
<li><a class="reference internal" href="#column-mapping">6.3.3.8.5. Column mapping</a></li>
<li><a class="reference internal" href="#downloads">6.3.3.8.6. Downloads</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="psl2bed.html"
                        title="previous chapter">6.3.3.7. <cite>psl2bed</cite></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="vcf2bed.html"
                        title="next chapter">6.3.3.9. <cite>vcf2bed</cite></a></p>
<div class="sourcelink">
  <a href="../../../../_sources/content/reference/file-management/conversion/sam2bed.txt" rel="nofollow">
    Show page source
  </a>
</div>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>

        </div>
      </div>
      <div class="clearer"></div>
    </div>
  <div class="related bottom">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="psl2bed.html" title="Previous document">6.3.3.7. <cite>psl2bed</cite></a>
        </li>
        <li>
          <a href="vcf2bed.html" title="Next document">6.3.3.9. <cite>vcf2bed</cite></a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../../../../index.html">BEDOPS 2.4.4 documentation</a></li>
      <li>
        <a href="../../../reference.html">6. Reference</a>
      </li>
      <li>
        <a href="../../file-management.html">6.3. File management</a>
      </li>
      <li>
        <a href="../conversion.html">6.3.3. Data conversion</a>
      </li> 
    </ul>
  </nav>
  </div>
  <footer id="pagefooter">&copy; 2014, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a>
      1.2.3
        with the <a href="http://github.com/irskep/sphinx-better-theme">
          better</a> theme.

  </footer>

  
  </body>
</html>