
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>6.3.3.11. vcf2bed &#8212; BEDOPS v2.4.36</title>
    <link rel="stylesheet" href="../../../../_static/better.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/custom.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../../../" src="../../../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../../../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../../../../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="next" title="6.3.3.12. wig2bed" href="wig2bed.html" />
    <link rel="prev" title="6.3.3.10. sam2bed" href="sam2bed.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  </head><body>
    <header id="pageheader"><h1><a href="../../../../index.html ">
        BEDOPS v2.4.36
    </a></h1></header>
  <div class="related top">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="sam2bed.html" title="Previous document">6.3.3.10. <cite>sam2bed</cite></a>
        </li>
        <li>
          <a href="wig2bed.html" title="Next document">6.3.3.12. <cite>wig2bed</cite></a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../../../../index.html">Home</a></li>
      <li>
        <a href="../../../reference.html">6. Reference</a>
      </li>
      <li>
        <a href="../../file-management.html">6.3. File management</a>
      </li>
      <li>
        <a href="../conversion.html">6.3.3. Data conversion</a>
      </li> 
    </ul>
  </nav>
  </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="vcf2bed">
<span id="id1"></span><h1>6.3.3.11. <cite>vcf2bed</cite><a class="headerlink" href="#vcf2bed" title="Permalink to this headline">¶</a></h1>
<p>The <code class="docutils literal notranslate"><span class="pre">vcf2bed</span></code> script converts 1-based, closed <code class="docutils literal notranslate"><span class="pre">[start,</span> <span class="pre">end]</span></code> <a class="reference external" href="http://vcftools.sourceforge.net/specs.html">Variant Call Format v4.2</a> (VCF) to sorted, 0-based, half-open <code class="docutils literal notranslate"><span class="pre">[start-1,</span> <span class="pre">start)</span></code> extended BED data.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Note that this script converts from <code class="docutils literal notranslate"><span class="pre">[start,</span> <span class="pre">end]</span></code> to <code class="docutils literal notranslate"><span class="pre">[start-1,</span> <span class="pre">start)</span></code>. Unless the <code class="docutils literal notranslate"><span class="pre">--snvs</span></code>, <code class="docutils literal notranslate"><span class="pre">--insertions</span></code> or <code class="docutils literal notranslate"><span class="pre">--deletions</span></code> options are added, we perform the equivalent of a <em>single-base insertion</em> to make BED output that is guaranteed to work with BEDOPS, <strong>regardless of what the actual variant may be</strong>, to allow operations to be performed. The converted output contains additional columns which allow reconstruction of the original VCF data and associated variant parameters.</p>
</div>
<p>For convenience, we also offer <code class="docutils literal notranslate"><span class="pre">vcf2starch</span></code>, which performs the extra step of creating a <a class="reference internal" href="../compression/starch-specification.html#starch-specification"><span class="std std-ref">Starch-formatted</span></a> archive.</p>
<div class="section" id="dependencies">
<h2>6.3.3.11.1. Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">vcf2bed</span></code> script requires <a class="reference internal" href="convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a>. The <code class="docutils literal notranslate"><span class="pre">vcf2starch</span></code> script requires <a class="reference internal" href="../compression/starch.html#starch"><span class="std std-ref">starch</span></a>. Both dependencies are part of a typical BEDOPS installation.</p>
<p>This script is also dependent on input that follows the VCF v4.2 specification.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">Conversion of data which are VCF-like, but which do not follow the specification can cause parsing issues. If you run into problems, please check that your input follows the VCF specification using validation tools, such as those packaged with <a class="reference external" href="http://vcftools.sourceforge.net/perl_module.html#vcf-validator">VCFTools</a>.</p>
</div>
</div>
<div class="section" id="source">
<h2>6.3.3.11.2. Source<a class="headerlink" href="#source" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">vcf2bed</span></code> and <code class="docutils literal notranslate"><span class="pre">vcf2starch</span></code> conversion scripts are part of the binary and source downloads of BEDOPS. See the <a class="reference internal" href="../../../installation.html#installation"><span class="std std-ref">Installation</span></a> documentation for more details.</p>
</div>
<div class="section" id="usage">
<h2>6.3.3.11.3. Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">vcf2bed</span></code> script parses VCF from standard input and prints sorted BED to standard output. The <code class="docutils literal notranslate"><span class="pre">vcf2starch</span></code> script uses an extra step to parse VCF to a compressed BEDOPS <a class="reference internal" href="../compression/starch-specification.html#starch-specification"><span class="std std-ref">Starch-formatted</span></a> archive, which is also directed to standard output.</p>
<p>The header data of a VCF file is usually discarded, unless you add the <code class="docutils literal notranslate"><span class="pre">--keep-header</span></code> option. In this case, BED elements are created from these data, using the chromosome name <code class="docutils literal notranslate"><span class="pre">_header</span></code> to denote content. Line numbers are specified in the start and stop coordinates, and unmodified header data are placed in the fourth column (ID field).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">By default, multiple BED annotations are printed if there are multiple alternate alleles in a variant call. Use the <code class="docutils literal notranslate"><span class="pre">--do-not-split-alt-alleles</span></code> option to preserve the alternate allele string and print only one BED element for the variant call.</p>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">By default, all conversion scripts now output sorted BED data ready for use with BEDOPS utilities. If the converted BED output looks truncated or incomplete, and you are converting a VCF file that is larger than the capacity of your <code class="docutils literal notranslate"><span class="pre">/tmp</span></code> folder, then use the <code class="docutils literal notranslate"><span class="pre">--sort-tmpdir</span></code> option to specify an alternative directory to store intermediate sort results. Or, if you do not want to sort the converted output, use the <code class="docutils literal notranslate"><span class="pre">--do-not-sort</span></code> option. Run the script with the <code class="docutils literal notranslate"><span class="pre">--help</span></code> option for more details.</p>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">If you are sorting data larger than system memory, use the <code class="docutils literal notranslate"><span class="pre">--max-mem</span></code> option to limit sort memory usage to a reasonable fraction of available memory, <em>e.g.</em>, <code class="docutils literal notranslate"><span class="pre">--max-mem</span> <span class="pre">2G</span></code> or similar. See <code class="docutils literal notranslate"><span class="pre">--help</span></code> for more details.</p>
</div>
</div>
<div class="section" id="customized-variant-handling">
<span id="vcf2bed-custom-variants"></span><h2>6.3.3.11.4. Customized variant handling<a class="headerlink" href="#customized-variant-handling" title="Permalink to this headline">¶</a></h2>
<p>By default, the <code class="docutils literal notranslate"><span class="pre">vcf2bed</span></code> script translates all variants to single-base positions in the resulting BED output. Depending on the category of variant you are interested in, however, you may want more specific categories handled differently.</p>
<p>Based on the VCF v4.2 specification, we also provide three custom options for filtering input for each of the three types of variants listed: <code class="docutils literal notranslate"><span class="pre">--snvs</span></code>, <code class="docutils literal notranslate"><span class="pre">--insertions</span></code> and <code class="docutils literal notranslate"><span class="pre">--deletions</span></code>. In each case, we use the length of the reference and alternate alleles to determine which type of variant is being handled.</p>
<p>In addition, using any of these three custom options automatically results in processing of mixed variant records for a microsatellite, where present. For instance, the following record contains a mixture of a deletion and insertion variant (<code class="docutils literal notranslate"><span class="pre">GTC</span> <span class="pre">-&gt;</span> <span class="pre">G</span></code> and <code class="docutils literal notranslate"><span class="pre">GTC</span> <span class="pre">-&gt;</span> <span class="pre">GTCT</span></code>, respectively):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#CHROM POS     ID        REF    ALT     QUAL FILTER INFO                              FORMAT      NA00001        NA00002        NA00003</span>
<span class="mi">20</span>     <span class="mf">12.4</span><span class="o">.</span><span class="mi">367</span> <span class="n">microsat1</span> <span class="n">GTC</span>    <span class="n">G</span><span class="p">,</span><span class="n">GTCT</span>  <span class="mi">50</span>   <span class="n">PASS</span>   <span class="n">NS</span><span class="o">=</span><span class="mi">3</span><span class="p">;</span><span class="n">DP</span><span class="o">=</span><span class="mi">9</span><span class="p">;</span><span class="n">AA</span><span class="o">=</span><span class="n">G</span>                    <span class="n">GT</span><span class="p">:</span><span class="n">GQ</span><span class="p">:</span><span class="n">DP</span>    <span class="mi">0</span><span class="o">/</span><span class="mi">1</span><span class="p">:</span><span class="mi">35</span><span class="p">:</span><span class="mi">4</span>       <span class="mi">0</span><span class="o">/</span><span class="mi">2</span><span class="p">:</span><span class="mi">17</span><span class="p">:</span><span class="mi">2</span>       <span class="mi">1</span><span class="o">/</span><span class="mi">1</span><span class="p">:</span><span class="mi">40</span><span class="p">:</span><span class="mi">3</span>
</pre></div>
</div>
<p>When using <code class="docutils literal notranslate"><span class="pre">--snvs</span></code>, <code class="docutils literal notranslate"><span class="pre">--insertions</span></code> or <code class="docutils literal notranslate"><span class="pre">--deletions</span></code>, this record is split into two distinct BED records and filtered depending on which custom option was chosen. The <code class="docutils literal notranslate"><span class="pre">--insertions</span></code> option would only export the single-base position of the insertion in this mixed variant, while <code class="docutils literal notranslate"><span class="pre">--deletions</span></code> would show the deletion.</p>
<p>In this way, you can control what kinds of variants are translated into BED outputs—most importantly, there is also no confusion about what the length of the BED element signifies.</p>
</div>
<div class="section" id="example">
<h2>6.3.3.11.5. Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>To demonstrate these scripts, we use a sample VCF input called <code class="docutils literal notranslate"><span class="pre">foo.vcf</span></code> (see the <a class="reference internal" href="#vcf2bed-downloads"><span class="std std-ref">Downloads</span></a> section to grab this file).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This data is also publicly available from the <a class="reference external" href="http://www.broadinstitute.org/gsa/wiki/index.php/Understanding_the_Unified_Genotyper's_VCF_files">Broad Institute</a>.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">##fileformat=VCFv4.0</span>
<span class="c1">##FILTER=&lt;ID=LowQual,Description=&quot;QUAL &lt; 50.0&quot;&gt;</span>
<span class="c1">##FORMAT=&lt;ID=AD,Number=.,Type=Integer,Description=&quot;Allelic depths for the ref and alt alleles in the order listed&quot;&gt;</span>
<span class="c1">##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description=&quot;Read Depth (only filtered reads used for calling)&quot;&gt;</span>
<span class="c1">##FORMAT=&lt;ID=GQ,Number=1,Type=Float,Description=&quot;Genotype Quality&quot;&gt;</span>
<span class="c1">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;</span>
<span class="c1">##FORMAT=&lt;ID=PL,Number=3,Type=Float,Description=&quot;Normalized, Phred-scaled likelihoods for AA,AB,BB genotypes where A=ref and B=alt; not applicable if site is not biallelic&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=AC,Number=.,Type=Integer,Description=&quot;Allele count in genotypes, for each ALT allele, in the same order as listed&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=AF,Number=.,Type=Float,Description=&quot;Allele Frequency, for each ALT allele, in the same order as listed&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=AN,Number=1,Type=Integer,Description=&quot;Total number of alleles in called genotypes&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=DB,Number=0,Type=Flag,Description=&quot;dbSNP Membership&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=DP,Number=1,Type=Integer,Description=&quot;Total Depth&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=DS,Number=0,Type=Flag,Description=&quot;Were any of the samples downsampled?&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=Dels,Number=1,Type=Float,Description=&quot;Fraction of Reads Containing Spanning Deletions&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=HRun,Number=1,Type=Integer,Description=&quot;Largest Contiguous Homopolymer Run of Variant Allele In Either Direction&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=HaplotypeScore,Number=1,Type=Float,Description=&quot;Consistency of the site with two (and only two) segregating haplotypes&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=MQ,Number=1,Type=Float,Description=&quot;RMS Mapping Quality&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=MQ0,Number=1,Type=Integer,Description=&quot;Total Mapping Quality Zero Reads&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=QD,Number=1,Type=Float,Description=&quot;Variant Confidence/Quality by Depth&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=SB,Number=1,Type=Float,Description=&quot;Strand Bias&quot;&gt;</span>
<span class="c1">##INFO=&lt;ID=VQSLOD,Number=1,Type=Float,Description=&quot;log10-scaled probability of variant being true under the trained gaussian mixture model&quot;&gt;</span>
<span class="c1">##UnifiedGenotyperV2=&quot;analysis_type=UnifiedGenotyperV2 input_file=[TEXT CLIPPED FOR CLARITY]&quot;</span>
<span class="c1">#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  NA12878</span>
<span class="n">chr1</span>    <span class="mi">873762</span>  <span class="o">.</span>       <span class="n">T</span>       <span class="n">G</span>       <span class="mf">5231.78</span> <span class="n">PASS</span>    <span class="n">AC</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">AF</span><span class="o">=</span><span class="mf">0.50</span><span class="p">;</span><span class="n">AN</span><span class="o">=</span><span class="mi">2</span><span class="p">;</span><span class="n">DP</span><span class="o">=</span><span class="mi">315</span><span class="p">;</span><span class="n">Dels</span><span class="o">=</span><span class="mf">0.00</span><span class="p">;</span><span class="n">HRun</span><span class="o">=</span><span class="mi">2</span><span class="p">;</span><span class="n">HaplotypeScore</span><span class="o">=</span><span class="mf">15.11</span><span class="p">;</span><span class="n">MQ</span><span class="o">=</span><span class="mf">91.05</span><span class="p">;</span><span class="n">MQ0</span><span class="o">=</span><span class="mi">15</span><span class="p">;</span><span class="n">QD</span><span class="o">=</span><span class="mf">16.61</span><span class="p">;</span><span class="n">SB</span><span class="o">=-</span><span class="mf">1533.02</span><span class="p">;</span><span class="n">VQSLOD</span><span class="o">=-</span><span class="mf">1.5473</span>      <span class="n">GT</span><span class="p">:</span><span class="n">AD</span><span class="p">:</span><span class="n">DP</span><span class="p">:</span><span class="n">GQ</span><span class="p">:</span><span class="n">PL</span>  <span class="mi">0</span><span class="o">/</span><span class="mi">1</span><span class="p">:</span><span class="mi">173</span><span class="p">,</span><span class="mi">141</span><span class="p">:</span><span class="mi">282</span><span class="p">:</span><span class="mi">99</span><span class="p">:</span><span class="mi">255</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">255</span>
<span class="n">chr1</span>    <span class="mi">877664</span>  <span class="n">rs3828047</span>       <span class="n">A</span>       <span class="n">G</span>       <span class="mf">3931.66</span> <span class="n">PASS</span>    <span class="n">AC</span><span class="o">=</span><span class="mi">2</span><span class="p">;</span><span class="n">AF</span><span class="o">=</span><span class="mf">1.00</span><span class="p">;</span><span class="n">AN</span><span class="o">=</span><span class="mi">2</span><span class="p">;</span><span class="n">DB</span><span class="p">;</span><span class="n">DP</span><span class="o">=</span><span class="mi">105</span><span class="p">;</span><span class="n">Dels</span><span class="o">=</span><span class="mf">0.00</span><span class="p">;</span><span class="n">HRun</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">HaplotypeScore</span><span class="o">=</span><span class="mf">1.59</span><span class="p">;</span><span class="n">MQ</span><span class="o">=</span><span class="mf">92.52</span><span class="p">;</span><span class="n">MQ0</span><span class="o">=</span><span class="mi">4</span><span class="p">;</span><span class="n">QD</span><span class="o">=</span><span class="mf">37.44</span><span class="p">;</span><span class="n">SB</span><span class="o">=-</span><span class="mf">1152.13</span><span class="p">;</span><span class="n">VQSLOD</span><span class="o">=</span><span class="mf">0.1185</span>      <span class="n">GT</span><span class="p">:</span><span class="n">AD</span><span class="p">:</span><span class="n">DP</span><span class="p">:</span><span class="n">GQ</span><span class="p">:</span><span class="n">PL</span>  <span class="mi">1</span><span class="o">/</span><span class="mi">1</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span><span class="mi">105</span><span class="p">:</span><span class="mi">94</span><span class="p">:</span><span class="mi">99</span><span class="p">:</span><span class="mi">255</span><span class="p">,</span><span class="mi">255</span><span class="p">,</span><span class="mi">0</span>
<span class="n">chr1</span>    <span class="mi">899282</span>  <span class="n">rs28548431</span>      <span class="n">C</span>       <span class="n">T</span>       <span class="mf">71.77</span>   <span class="n">PASS</span>    <span class="n">AC</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">AF</span><span class="o">=</span><span class="mf">0.50</span><span class="p">;</span><span class="n">AN</span><span class="o">=</span><span class="mi">2</span><span class="p">;</span><span class="n">DB</span><span class="p">;</span><span class="n">DP</span><span class="o">=</span><span class="mi">4</span><span class="p">;</span><span class="n">Dels</span><span class="o">=</span><span class="mf">0.00</span><span class="p">;</span><span class="n">HRun</span><span class="o">=</span><span class="mi">0</span><span class="p">;</span><span class="n">HaplotypeScore</span><span class="o">=</span><span class="mf">0.00</span><span class="p">;</span><span class="n">MQ</span><span class="o">=</span><span class="mf">99.00</span><span class="p">;</span><span class="n">MQ0</span><span class="o">=</span><span class="mi">0</span><span class="p">;</span><span class="n">QD</span><span class="o">=</span><span class="mf">17.94</span><span class="p">;</span><span class="n">SB</span><span class="o">=-</span><span class="mf">46.55</span><span class="p">;</span><span class="n">VQSLOD</span><span class="o">=-</span><span class="mf">1.9148</span> <span class="n">GT</span><span class="p">:</span><span class="n">AD</span><span class="p">:</span><span class="n">DP</span><span class="p">:</span><span class="n">GQ</span><span class="p">:</span><span class="n">PL</span>  <span class="mi">0</span><span class="o">/</span><span class="mi">1</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="mf">25.92</span><span class="p">:</span><span class="mi">103</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">26</span>
<span class="n">chr1</span>    <span class="mi">974165</span>  <span class="n">rs9442391</span>       <span class="n">T</span>       <span class="n">C</span>       <span class="mf">29.84</span>   <span class="n">LowQual</span> <span class="n">AC</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">AF</span><span class="o">=</span><span class="mf">0.50</span><span class="p">;</span><span class="n">AN</span><span class="o">=</span><span class="mi">2</span><span class="p">;</span><span class="n">DB</span><span class="p">;</span><span class="n">DP</span><span class="o">=</span><span class="mi">18</span><span class="p">;</span><span class="n">Dels</span><span class="o">=</span><span class="mf">0.00</span><span class="p">;</span><span class="n">HRun</span><span class="o">=</span><span class="mi">1</span><span class="p">;</span><span class="n">HaplotypeScore</span><span class="o">=</span><span class="mf">0.16</span><span class="p">;</span><span class="n">MQ</span><span class="o">=</span><span class="mf">95.26</span><span class="p">;</span><span class="n">MQ0</span><span class="o">=</span><span class="mi">0</span><span class="p">;</span><span class="n">QD</span><span class="o">=</span><span class="mf">1.66</span><span class="p">;</span><span class="n">SB</span><span class="o">=-</span><span class="mf">0.98</span> <span class="n">GT</span><span class="p">:</span><span class="n">AD</span><span class="p">:</span><span class="n">DP</span><span class="p">:</span><span class="n">GQ</span><span class="p">:</span><span class="n">PL</span>  <span class="mi">0</span><span class="o">/</span><span class="mi">1</span><span class="p">:</span><span class="mi">14</span><span class="p">,</span><span class="mi">4</span><span class="p">:</span><span class="mi">14</span><span class="p">:</span><span class="mf">60.91</span><span class="p">:</span><span class="mi">61</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">255</span>
</pre></div>
</div>
<p>We can convert VCF to sorted BED data in the following manner:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ vcf2bed &lt; foo.vcf
chr1    873761  873762  .       5231.78 T       G       PASS    AC=1;AF=0.50;AN=2;DP=315;Dels=0.00;HRun=2;HaplotypeScore=15.11;MQ=91.05;MQ0=15;QD=16.61;SB=-1533.02;VQSLOD=-1.5473      GT:AD:DP:GQ:PL  0/1:173,141:282:99:255,0,255
chr1    877663  877664  rs3828047       3931.66 A       G       PASS    AC=2;AF=1.00;AN=2;DB;DP=105;Dels=0.00;HRun=1;HaplotypeScore=1.59;MQ=92.52;MQ0=4;QD=37.44;SB=-1152.13;VQSLOD=0.1185      GT:AD:DP:GQ:PL  1/1:0,105:94:99:255,255,0
chr1    899281  899282  rs28548431      71.77   C       T       PASS    AC=1;AF=0.50;AN=2;DB;DP=4;Dels=0.00;HRun=0;HaplotypeScore=0.00;MQ=99.00;MQ0=0;QD=17.94;SB=-46.55;VQSLOD=-1.9148 GT:AD:DP:GQ:PL  0/1:1,3:4:25.92:103,0,26
chr1    974164  974165  rs9442391       29.84   T       C       LowQual AC=1;AF=0.50;AN=2;DB;DP=18;Dels=0.00;HRun=1;HaplotypeScore=0.16;MQ=95.26;MQ0=0;QD=1.66;SB=-0.98 GT:AD:DP:GQ:PL  0/1:14,4:14:60.91:61,0,255
</pre></div>
</div>
<p>As you see here, the header data of the VCF file is discarded, unless you add the <code class="docutils literal notranslate"><span class="pre">--keep-header</span></code> option. In this case, BED elements are created from these data, using the chromosome name <code class="docutils literal notranslate"><span class="pre">_header</span></code> to denote content. Line numbers are specified in the start and stop coordinates, and unmodified header data are placed in the fourth column (ID field).</p>
<p>Here we use <code class="docutils literal notranslate"><span class="pre">--keep-header</span></code> with our example dataset:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ vcf2bed --keep-header &lt; foo.vcf
_header 0       1       ##fileformat=VCFv4.0
_header 1       2       ##FILTER=&lt;ID=LowQual,Description=&quot;QUAL &lt; 50.0&quot;&gt;
_header 2       3       ##FORMAT=&lt;ID=AD,Number=.,Type=Integer,Description=&quot;Allelic depths for the ref and alt alleles in the order listed&quot;&gt;
_header 3       4       ##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description=&quot;Read Depth (only filtered reads used for calling)&quot;&gt;
_header 4       5       ##FORMAT=&lt;ID=GQ,Number=1,Type=Float,Description=&quot;Genotype Quality&quot;&gt;
_header 5       6       ##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;
_header 6       7       ##FORMAT=&lt;ID=PL,Number=3,Type=Float,Description=&quot;Normalized, Phred-scaled likelihoods for AA,AB,BB genotypes where A=ref and B=alt; not applicable if site is not biallelic&quot;&gt;
_header 7       8       ##INFO=&lt;ID=AC,Number=.,Type=Integer,Description=&quot;Allele count in genotypes, for each ALT allele, in the same order as listed&quot;&gt;
_header 8       9       ##INFO=&lt;ID=AF,Number=.,Type=Float,Description=&quot;Allele Frequency, for each ALT allele, in the same order as listed&quot;&gt;
_header 9       10      ##INFO=&lt;ID=AN,Number=1,Type=Integer,Description=&quot;Total number of alleles in called genotypes&quot;&gt;
_header 10      11      ##INFO=&lt;ID=DB,Number=0,Type=Flag,Description=&quot;dbSNP Membership&quot;&gt;
_header 11      12      ##INFO=&lt;ID=DP,Number=1,Type=Integer,Description=&quot;Total Depth&quot;&gt;
_header 12      13      ##INFO=&lt;ID=DS,Number=0,Type=Flag,Description=&quot;Were any of the samples downsampled?&quot;&gt;
_header 13      14      ##INFO=&lt;ID=Dels,Number=1,Type=Float,Description=&quot;Fraction of Reads Containing Spanning Deletions&quot;&gt;
_header 14      15      ##INFO=&lt;ID=HRun,Number=1,Type=Integer,Description=&quot;Largest Contiguous Homopolymer Run of Variant Allele In Either Direction&quot;&gt;
_header 15      16      ##INFO=&lt;ID=HaplotypeScore,Number=1,Type=Float,Description=&quot;Consistency of the site with two (and only two) segregating haplotypes&quot;&gt;
_header 16      17      ##INFO=&lt;ID=MQ,Number=1,Type=Float,Description=&quot;RMS Mapping Quality&quot;&gt;
_header 17      18      ##INFO=&lt;ID=MQ0,Number=1,Type=Integer,Description=&quot;Total Mapping Quality Zero Reads&quot;&gt;
_header 18      19      ##INFO=&lt;ID=QD,Number=1,Type=Float,Description=&quot;Variant Confidence/Quality by Depth&quot;&gt;
_header 19      20      ##INFO=&lt;ID=SB,Number=1,Type=Float,Description=&quot;Strand Bias&quot;&gt;
_header 20      21      ##INFO=&lt;ID=VQSLOD,Number=1,Type=Float,Description=&quot;log10-scaled probability of variant being true under the trained gaussian mixture model&quot;&gt;
_header 21      22      ##UnifiedGenotyperV2=&quot;analysis_type=UnifiedGenotyperV2 input_file=[TEXT CLIPPED FOR CLARITY]&quot;
_header 22      23      #CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  NA12878
chr1    873761  873762  .       5231.78 T       G       PASS    AC=1;AF=0.50;AN=2;DP=315;Dels=0.00;HRun=2;HaplotypeScore=15.11;MQ=91.05;MQ0=15;QD=16.61;SB=-1533.02;VQSLOD=-1.5473      GT:AD:DP:GQ:PL  0/1:173,141:282:99:255,0,255
chr1    877663  877664  rs3828047       3931.66 A       G       PASS    AC=2;AF=1.00;AN=2;DB;DP=105;Dels=0.00;HRun=1;HaplotypeScore=1.59;MQ=92.52;MQ0=4;QD=37.44;SB=-1152.13;VQSLOD=0.1185      GT:AD:DP:GQ:PL  1/1:0,105:94:99:255,255,0
chr1    899281  899282  rs28548431      71.77   C       T       PASS    AC=1;AF=0.50;AN=2;DB;DP=4;Dels=0.00;HRun=0;HaplotypeScore=0.00;MQ=99.00;MQ0=0;QD=17.94;SB=-46.55;VQSLOD=-1.9148 GT:AD:DP:GQ:PL  0/1:1,3:4:25.92:103,0,26
chr1    974164  974165  rs9442391       29.84   T       C       LowQual AC=1;AF=0.50;AN=2;DB;DP=18;Dels=0.00;HRun=1;HaplotypeScore=0.16;MQ=95.26;MQ0=0;QD=1.66;SB=-0.98 GT:AD:DP:GQ:PL  0/1:14,4:14:60.91:61,0,255
</pre></div>
</div>
<p>With this option, the <code class="docutils literal notranslate"><span class="pre">vcf2*</span></code> scripts are completely “non-lossy”. Use of <code class="docutils literal notranslate"><span class="pre">awk</span></code> or other scripting tools can munge these data back into a VCF-formatted file.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Note the conversion from 1- to 0-based coordinate indexing, in the transition from VCF to BED. While BEDOPS supports 0- and 1-based coordinate indexing, the coordinate change made here is believed to be convenient for most end users.</p>
</div>
</div>
<div class="section" id="column-mapping">
<span id="vcf2bed-column-mapping"></span><h2>6.3.3.11.6. Column mapping<a class="headerlink" href="#column-mapping" title="Permalink to this headline">¶</a></h2>
<p>In this section, we describe how VCF v4.2 columns are mapped to BED columns. We start with the first five UCSC BED columns as follows:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="33%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">VCF v4.2 field</th>
<th class="head">BED column index</th>
<th class="head">BED field</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>#CHROM</td>
<td>1</td>
<td>chromosome</td>
</tr>
<tr class="row-odd"><td>POS - 1</td>
<td>2</td>
<td>start</td>
</tr>
<tr class="row-even"><td>POS (*)</td>
<td>3</td>
<td>stop</td>
</tr>
<tr class="row-odd"><td>ID</td>
<td>4</td>
<td>id</td>
</tr>
<tr class="row-even"><td>QUAL</td>
<td>5</td>
<td>score</td>
</tr>
</tbody>
</table>
<p>The remaining columns are mapped as follows:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="33%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">VCF v4.2 field</th>
<th class="head">BED column index</th>
<th class="head">BED field</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>REF</td>
<td>6</td>
<td>&#160;</td>
</tr>
<tr class="row-odd"><td>ALT</td>
<td>7</td>
<td>&#160;</td>
</tr>
<tr class="row-even"><td>FILTER</td>
<td>8</td>
<td>&#160;</td>
</tr>
<tr class="row-odd"><td>INFO</td>
<td>9</td>
<td>&#160;</td>
</tr>
</tbody>
</table>
<p>If present in the VCF v4.2 input, the following columns are also mapped:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="33%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">VCF v4.2 field</th>
<th class="head">BED column index</th>
<th class="head">BED field</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>FORMAT</td>
<td>10</td>
<td>&#160;</td>
</tr>
<tr class="row-odd"><td>Sample ID 1</td>
<td>11</td>
<td>&#160;</td>
</tr>
<tr class="row-even"><td>Sample ID 2</td>
<td>12</td>
<td>&#160;</td>
</tr>
<tr class="row-odd"><td>…</td>
<td>13, 14, etc.</td>
<td>&#160;</td>
</tr>
</tbody>
</table>
<p>When using <code class="docutils literal notranslate"><span class="pre">--deletions</span></code>, the stop value of the BED output is determined by the length difference between ALT and REF alleles. Use of <code class="docutils literal notranslate"><span class="pre">--insertions</span></code> or <code class="docutils literal notranslate"><span class="pre">--snvs</span></code> yields a one-base BED element.</p>
<p>If the ALT field contains more than one allele, multiple BED records will be printed. Use the <code class="docutils literal notranslate"><span class="pre">--do-not-split</span></code> option if you only want one BED record per variant call.</p>
<p>The “meta-information” (starting with <code class="docutils literal notranslate"><span class="pre">##</span></code>) and “header” lines (starting with <code class="docutils literal notranslate"><span class="pre">#</span></code>) are discarded, unless the <code class="docutils literal notranslate"><span class="pre">--keep-headers</span></code> options is specified.</p>
</div>
<div class="section" id="downloads">
<span id="vcf2bed-downloads"></span><h2>6.3.3.11.7. Downloads<a class="headerlink" href="#downloads" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Sample VCF dataset: <a class="reference download internal" download="" href="../../../../_downloads/3a101aa02f5fb31d9786257067cec836/reference_vcf2bed_foo.vcf"><code class="xref download docutils literal notranslate"><span class="pre">foo.vcf</span></code></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../../../index.html">
              <img class="logo" src="../../../../_static/logo_with_label_v3.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../../../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">6.3.3.11. <cite>vcf2bed</cite></a><ul>
<li><a class="reference internal" href="#dependencies">6.3.3.11.1. Dependencies</a></li>
<li><a class="reference internal" href="#source">6.3.3.11.2. Source</a></li>
<li><a class="reference internal" href="#usage">6.3.3.11.3. Usage</a></li>
<li><a class="reference internal" href="#customized-variant-handling">6.3.3.11.4. Customized variant handling</a></li>
<li><a class="reference internal" href="#example">6.3.3.11.5. Example</a></li>
<li><a class="reference internal" href="#column-mapping">6.3.3.11.6. Column mapping</a></li>
<li><a class="reference internal" href="#downloads">6.3.3.11.7. Downloads</a></li>
</ul>
</li>
</ul>

<form class="search" action="../../../../search.html" method="get">
  <input type="text" name="q"
   placeholder="type to search" />
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
  <div class="related bottom">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="sam2bed.html" title="Previous document">6.3.3.10. <cite>sam2bed</cite></a>
        </li>
        <li>
          <a href="wig2bed.html" title="Next document">6.3.3.12. <cite>wig2bed</cite></a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../../../../index.html">Home</a></li>
      <li>
        <a href="../../../reference.html">6. Reference</a>
      </li>
      <li>
        <a href="../../file-management.html">6.3. File management</a>
      </li>
      <li>
        <a href="../conversion.html">6.3.3. Data conversion</a>
      </li> 
    </ul>
  </nav>
  </div>
  <footer id="pagefooter">&copy; 2011-2019, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a>
      1.8.5
        with the <a href="http://github.com/irskep/sphinx-better-theme">
          better</a> theme.

  </footer>

  
  </body>
</html>