<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>3. Revision history &mdash; BEDOPS v2.4.4</title>
    
    <link rel="stylesheet" href="../_static/better.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.4.4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://c328740.ssl.cf1.rackcdn.com/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="BEDOPS v2.4.4" href="../index.html" />
    <link rel="next" title="4. Usage examples" href="usage-examples.html" />
    <link rel="prev" title="2. Installation" href="installation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  </head>
  <body role="document">
    <header id="pageheader"><h1><a href="../index.html ">
        BEDOPS v2.4.4
    </a></h1></header>
  <div class="related top">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="installation.html" title="Previous document">2. Installation</a>
        </li>
        <li>
          <a href="usage-examples.html" title="Next document">4. Usage examples</a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../index.html">Home</a></li> 
    </ul>
  </nav>
  </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="revision-history">
<span id="id1"></span><h1>3. Revision history<a class="headerlink" href="#revision-history" title="Permalink to this headline">¶</a></h1>
<p>This page summarizes some of the more important changes between releases.</p>
<div class="section" id="current-version">
<span id="revision-history-of-current-version"></span><h2>3.1. Current version<a class="headerlink" href="#current-version" title="Permalink to this headline">¶</a></h2>
<div class="section" id="v2-4-3">
<h3>3.1.1. v2.4.3<a class="headerlink" href="#v2-4-3" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>December 18, 2014</strong></p>
<ul class="simple">
<li>Compilation improvements<ul>
<li>Shane Neph put in a great deal of work to enable parallel builds (<em>e.g.</em>, <code class="docutils literal"><span class="pre">make</span> <span class="pre">-j</span> <span class="pre">N</span></code> to build various targets in parallel). Depending on the end user&#8217;s environment, this can speed up compilation time by a factor of 2, 4 or more.</li>
<li>Fixed numerous compilation warnings of debug builds of <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a> toolkit under RHEL6/GCC and OS X 10.10.1/LLVM.</li>
</ul>
</li>
<li>New <a class="reference internal" href="reference/set-operations/bedops.html#bedops"><em>bedops</em></a> features<ul>
<li>Added <code class="docutils literal"><span class="pre">--chop</span></code> and <code class="docutils literal"><span class="pre">--stagger</span></code> options to &#8220;melt&#8221; inputs into contiguous or staggered disjoint regions of equivalent size.</li>
<li>For less confusion, arguments for <code class="docutils literal"><span class="pre">--element-of</span></code>, <code class="docutils literal"><span class="pre">--chop</span></code> and other <code class="docutils literal"><span class="pre">bedops</span></code> operations that take numerical modifiers no longer require a leading hyphen character. For instance, <code class="docutils literal"><span class="pre">--element-of</span> <span class="pre">1</span></code> is now equivalent to the former usage of <code class="docutils literal"><span class="pre">--element-of</span> <span class="pre">-1</span></code>.</li>
</ul>
</li>
<li>New <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a> features<ul>
<li>The <code class="docutils literal"><span class="pre">--sweep-all</span></code> option reads through the entire map file without early termination and can help deal with <code class="docutils literal"><span class="pre">SIGPIPE</span></code> errors. It adds to execution time, but the penalty is not as severe as with the use of <code class="docutils literal"><span class="pre">--ec</span></code>. Using <code class="docutils literal"><span class="pre">--ec</span></code> alone will enable error checking, but will now no longer read through the entire map file. The <code class="docutils literal"><span class="pre">--ec</span></code> option can be used in conjunction with <code class="docutils literal"><span class="pre">--sweep-all</span></code>, with the associated time penalties. (Another method for dealing with issue this is to override how <code class="docutils literal"><span class="pre">SIGPIPE</span></code> errors are caught by the interpreter (bash, Python, etc.) and retrapping them or ignoring them. However, it may not a good idea to do this as other situations may arise in production pipelines where it is ideal to trap and handle all I/O errors in a default manner.)</li>
<li>New <code class="docutils literal"><span class="pre">--echo-ref-size</span></code> and <code class="docutils literal"><span class="pre">--echo-ref-name</span></code> operations report genomic length of reference element, and rename the reference element in <code class="docutils literal"><span class="pre">chrom:start-end</span></code> (useful for labeling rows for input for <code class="docutils literal"><span class="pre">matrix2png</span></code> or <code class="docutils literal"><span class="pre">R</span></code> or other applications).</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/bedextract.html#bedextract"><em>bedextract</em></a><ul>
<li>Fixed upper bound bug that would cause incorrect output in some cases</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a><ul>
<li>Brand new C99 binary called <a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><em>convert2bed</em></a>, which wrapper scripts (<code class="docutils literal"><span class="pre">bam2bed</span></code>, etc.) now call. No more Python version dependencies, and the C-based rewrite offers massive performance improvements over old Python-based scripts.</li>
<li>Added <a class="reference internal" href="reference/file-management/conversion/parallel_bam2starch.html#parallel-bam2starch"><em>Parallel bam2starch</em></a> script, which parallelizes creation of <a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><em>Starch</em></a> archive from very large BAM files in SGE environments.</li>
<li>Added bug fix for missing code in <a class="reference internal" href="usage-examples/starchcluster.html#starchcluster"><em>starchcluster.gnu_parallel</em></a> script, where the final collation step was missing.</li>
<li>The <a class="reference internal" href="reference/file-management/conversion/vcf2bed.html#vcf2bed"><em>vcf2bed</em></a> script now accepts the <code class="docutils literal"><span class="pre">--do-not-split</span></code> option, which prints one BED element for all alternate alleles.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><em>Starch</em></a> archival format and compression/extraction tools<ul>
<li>Added duplicate- and <a class="reference internal" href="reference/set-operations/nested-elements.html#nested-elements"><em>nested-element</em></a> flags in v2.1 of Starch metadata, which denote if a chromosome contains one or more duplicate and/or nested elements. BED files compressed with <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a> v2.5 or greater, or Starch archives updated with <a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><em>starchcat</em></a> v2.5 or greater will include these values in the archive metadata. The <a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><em>unstarch</em></a> extraction tool offers <code class="docutils literal"><span class="pre">--has-duplicate</span></code> and <code class="docutils literal"><span class="pre">--has-nested</span></code> options to retrieve these flag values for a specified chromosome (or for all chromosomes).</li>
<li>Added <code class="docutils literal"><span class="pre">--is-starch</span></code> option to <a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><em>unstarch</em></a> to test if specified input file is a Starch v1 or v2 archive.</li>
<li>Added bug fix for compressing BED files with <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a>, where the archive would not include the last element of the BED input, if the BED input lacked a trailing newline. The compression tools now include a routine for capturing the last line, if there is no newline.</li>
</ul>
</li>
<li>Documentation improvements<ul>
<li>Remade some image assets throughout the documents to support Retina-grade displays</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="previous-versions">
<h2>3.2. Previous versions<a class="headerlink" href="#previous-versions" title="Permalink to this headline">¶</a></h2>
<div class="section" id="v2-4-2">
<h3>3.2.1. v2.4.2<a class="headerlink" href="#v2-4-2" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 10, 2014</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a><ul>
<li>Added support for <a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a> <code class="docutils literal"><span class="pre">--tmpdir</span></code> option to conversion scripts, to allow specification of alternative temporary directory for sorted results when used in conjunction with <code class="docutils literal"><span class="pre">--max-mem</span></code> option.</li>
<li>Added support for GFF3 files which include a FASTA directive in <code class="docutils literal"><span class="pre">gff2bed</span></code> and <code class="docutils literal"><span class="pre">gff2starch</span></code> (thanks to Keith Hughitt).</li>
<li>Extended support for Python-based conversion scripts to support use with Python v2.6.2 and forwards, except for <code class="docutils literal"><span class="pre">sam2bed</span></code> and <code class="docutils literal"><span class="pre">sam2starch</span></code>, which still require Python v2.7 or greater (and under Python3).</li>
<li>Fixed <code class="docutils literal"><span class="pre">--insertions</span></code> option in <a class="reference internal" href="reference/file-management/conversion/vcf2bed.html#vcf2bed"><em>vcf2bed</em></a> to now report a single-base BED element (thanks to Matt Maurano).</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-1">
<h3>3.2.2. v2.4.1<a class="headerlink" href="#v2-4-1" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 26, 2014</strong></p>
<ul>
<li><p class="first"><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a></p>
<ul class="simple">
<li>Added <code class="docutils literal"><span class="pre">--fraction-both</span></code> and <code class="docutils literal"><span class="pre">--exact</span></code> (<code class="docutils literal"><span class="pre">--fraction-both</span> <span class="pre">1</span></code>) to list of compatible overlap options with <code class="docutils literal"><span class="pre">--faster</span></code>.</li>
<li>Added 5% performance improvement with <cite>bedmap</cite> operations without <code class="docutils literal"><span class="pre">--faster</span></code>.</li>
<li>Fixed scenario that can yield incorrect results (cf. <a class="reference external" href="https://github.com/bedops/bedops/issues/43">Issue 43</a>).</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a></p>
<ul class="simple">
<li>Added <code class="docutils literal"><span class="pre">--tmpdir</span></code> option to allow specification of an alternative temporary directory, when used in conjunction with <code class="docutils literal"><span class="pre">--max-mem</span></code> option. This is useful if the host operating system&#8217;s standard temporary directory (<em>e.g.</em>, <code class="docutils literal"><span class="pre">/tmp</span></code> on Linux or OS X) does not have sufficient space to hold intermediate results.</li>
</ul>
</li>
<li><p class="first">All <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a></p>
<ul class="simple">
<li>Improvements to error handling in Python-based conversion scripts, in the case where no input is specified.</li>
<li>Fixed typos in <a class="reference internal" href="reference/file-management/conversion/gff2bed.html#gff2bed"><em>gff2bed</em></a> and <a class="reference internal" href="reference/file-management/conversion/psl2bed.html#psl2bed"><em>psl2bed</em></a> documentation (cf. <a class="reference external" href="https://github.com/bedops/bedops/commit/a091e18">commit a091e18</a>).</li>
</ul>
</li>
<li><p class="first">OS X compilation improvements</p>
<ul>
<li><p class="first">We have completed changes to the OS X build process for the remaining half of the BEDOPS binaries, which now allows direct, full compilation with Clang/LLVM (part of the Apple Xcode distribution).</p>
<p>All OS X BEDOPS binaries now use Apple&#8217;s system-level C++ library, instead of GNU&#8217;s <code class="docutils literal"><span class="pre">libstdc++</span></code>. It is no longer required (or recommended) to use GNU gcc to compile BEDOPS on OS X.</p>
<p>Compilation is faster and simpler, and we can reduce the size and complexity of Mac OS X builds and installer packages. By using Apple&#8217;s C++ library, we also eliminate the likelihood of missing library errors.</p>
<p>In the longer term, this gets us closer to moving BEDOPS to using the CMake build system, to further abstract and simplify the build process.</p>
</li>
</ul>
</li>
<li><p class="first">Cleaned up various compilation warnings found with <code class="docutils literal"><span class="pre">clang</span></code>/<code class="docutils literal"><span class="pre">clang++</span></code> and GCC kits.</p>
</li>
</ul>
</div>
<div class="section" id="v2-4-0">
<h3>3.2.3. v2.4.0<a class="headerlink" href="#v2-4-0" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 9, 2014</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a><ul>
<li>Added new <code class="docutils literal"><span class="pre">--echo-map-size</span></code> and <code class="docutils literal"><span class="pre">--echo-overlap-size</span></code> options to calculate sizes of mapped elements and overlaps between mapped and reference elements.</li>
<li>Improved performance for all <code class="docutils literal"><span class="pre">--echo-map-*</span></code> operations.</li>
<li>Updated documentation.</li>
</ul>
</li>
<li>Major enhancements and fixes to <a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a>:<ul>
<li>Improved performance.</li>
<li>Fixed memory leak.</li>
<li>Added support for millions of distinct chromosomes.</li>
<li>Improved internal estimation of memory usage with <code class="docutils literal"><span class="pre">--max-mem</span></code> option.</li>
</ul>
</li>
<li>Added support for compilation on Cygwin (64-bit). Refer to the <a class="reference internal" href="installation.html#installation-via-source-code-on-cygwin"><em>installation documentation</em></a> for build instructions.</li>
<li><a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><em>starchcat</em></a><ul>
<li>Fixed embarassing buffer overflow condition that caused segmentation faults on Ubuntu 13.</li>
</ul>
</li>
<li>All <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a><ul>
<li>Python-based scripts no longer use temporary files, which reduces file I/O and improves performance. This change also reduces the need for large amounts of free space in a user&#8217;s <code class="docutils literal"><span class="pre">/tmp</span></code> folder, particularly relevant for users converting multi-GB BAM files.</li>
<li>We now test for ability to locate <code class="docutils literal"><span class="pre">starch</span></code>, <code class="docutils literal"><span class="pre">sort-bed</span></code>, <code class="docutils literal"><span class="pre">wig2bed_bin</span></code> and <code class="docutils literal"><span class="pre">samtools</span></code> in user environment, quitting with the appropriate error state if the dependencies cannot be found.</li>
<li>Improved documentation. In particular, we have added descriptive tables to each script&#8217;s documentation page which describe how columns map from original data input to BED output.</li>
<li><a class="reference internal" href="reference/file-management/conversion/bam2bed.html#bam2bed"><em>bam2bed</em></a> and <a class="reference internal" href="reference/file-management/conversion/sam2bed.html#sam2bed"><em>sam2bed</em></a><ul>
<li>Added <code class="docutils literal"><span class="pre">--custom-tags</span> <span class="pre">&lt;value&gt;</span></code> command-line option to support a comma-separated list of custom tags (cf. <a class="reference external" href="http://www.biostars.org/p/87062/">Biostars discussion</a>), <em>i.e.</em>, tags which are not part of the original SAMtools specification.</li>
<li>Added <code class="docutils literal"><span class="pre">--keep-header</span></code> option to preserve header and metadata as BED elements that use <code class="docutils literal"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully &#8220;non-lossy&#8221;.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/vcf2bed.html#vcf2bed"><em>vcf2bed</em></a><ul>
<li>Added new <code class="docutils literal"><span class="pre">--snvs</span></code>, <code class="docutils literal"><span class="pre">--insertions</span></code> and <code class="docutils literal"><span class="pre">--deletions</span></code> options that filter VCF variants into three separate subcategories.</li>
<li>Added <code class="docutils literal"><span class="pre">--keep-header</span></code> option to preserve header and metadata as BED elements that use <code class="docutils literal"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully &#8220;non-lossy&#8221;.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/gff2bed.html#gff2bed"><em>gff2bed</em></a><ul>
<li>Added <code class="docutils literal"><span class="pre">--keep-header</span></code> option to preserve header and metadata as BED elements that use <code class="docutils literal"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully &#8220;non-lossy&#8221;.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/psl2bed.html#psl2bed"><em>psl2bed</em></a><ul>
<li>Added <code class="docutils literal"><span class="pre">--keep-header</span></code> option to preserve header and metadata as BED elements that use <code class="docutils literal"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully &#8220;non-lossy&#8221;.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/wig2bed.html#wig2bed"><em>wig2bed</em></a><ul>
<li>Added <code class="docutils literal"><span class="pre">--keep-header</span></code> option to <a class="reference internal" href="reference/file-management/conversion/wig2bed.html#wig2bed"><em>wig2bed</em></a> binary and <code class="docutils literal"><span class="pre">wig2bed</span></code>/<code class="docutils literal"><span class="pre">wig2starch</span></code> wrapper scripts, to preserve header and metadata as BED elements that use <code class="docutils literal"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully &#8220;non-lossy&#8221;.</li>
</ul>
</li>
</ul>
</li>
<li>Added OS X uninstaller project to allow end user to more easily remove BEDOPS tools from this platform.</li>
<li>Cleaned up various compilation warnings found with <code class="docutils literal"><span class="pre">clang</span></code>/<code class="docutils literal"><span class="pre">clang++</span></code> and GCC kits.</li>
</ul>
</div>
<div class="section" id="v2-3-0">
<h3>3.2.4. v2.3.0<a class="headerlink" href="#v2-3-0" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>October 2, 2013</strong></p>
<ul>
<li><p class="first">Migration of BEDOPS code and documentation from Google Code to Github.</p>
<ul class="simple">
<li>Due to changes with Google Code hosting policies at the end of the year, we have decided to change our process for distributing code, packages and documentation. While most of the work is done, we appreciate feedback on any problems you may encounter. Please email us at <a class="reference external" href="mailto:bedops&#37;&#52;&#48;stamlab&#46;org">bedops<span>&#64;</span>stamlab<span>&#46;</span>org</a> with details.</li>
<li>Migration to Github should facilitate requests for code by those who are familiar with <code class="docutils literal"><span class="pre">git</span></code> and want to fork our project to submit <a class="reference external" href="https://help.github.com/articles/using-pull-requests">pull requests</a>.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><em>bedops</em></a></p>
<ul class="simple">
<li>General <code class="docutils literal"><span class="pre">--ec</span></code> performance improvements.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a></p>
<ul class="simple">
<li>Adds support for the new <code class="docutils literal"><span class="pre">--skip-unmapped</span></code> option, which filters out reference elements which do not have mapped elements associated with them. See the end of the <a class="reference internal" href="reference/statistics/bedmap.html#bedmap-score-operations"><em>score operations</em></a> section of the <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a> documentation for more detail.</li>
<li>General <code class="docutils literal"><span class="pre">--ec</span></code> performance improvements.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a></p>
<ul>
<li><p class="first">Fixed bug with <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a> where zero-byte BED input (<em>i.e.</em>, an &#8220;empty set&#8221;) created a truncated and unusable archive. We now put in a &#8220;dummy&#8221; chromosome for zero-byte input, which <a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><em>unstarch</em></a> can now unpack.</p>
<p>This should simplify error handling with certain pipelines, specifically where set or other BEDOPS operations yield an &#8220;empty set&#8221; BED file that is subsequently compressed with <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a>.</p>
</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><em>unstarch</em></a></p>
<ul class="simple">
<li>Can now unpack zero-byte (&#8220;empty set&#8221;) compressed <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a> archive (see above).</li>
<li>Changed <code class="docutils literal"><span class="pre">unstarch</span> <span class="pre">--list</span></code> option to print to <code class="docutils literal"><span class="pre">stdout</span></code> stream (this was previously sent to <code class="docutils literal"><span class="pre">stderr</span></code>).</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a> metadata library</p>
<ul class="simple">
<li>Fixed array overflow bug with BEDOPS tools that take <a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><em>starch</em></a> archives as inputs, which affected use of archives as inputs to <a class="reference internal" href="reference/set-operations/closest-features.html#closest-features"><em>closest-features</em></a>, <a class="reference internal" href="reference/set-operations/bedops.html#bedops"><em>bedops</em></a> and <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a>.</li>
</ul>
</li>
<li><p class="first">All <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a></p>
<ul class="simple">
<li>Python scripts require v2.7+ or greater.</li>
<li>Improved (more &#8220;Pythonic&#8221;) error code handling.</li>
<li>Disabled support for <code class="docutils literal"><span class="pre">--max-mem</span></code> sort parameter until <a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a> <a class="reference external" href="https://github.com/bedops/bedops/issues/1">issue</a> is resolved. Scripts will continue to sort, but they will be limited to available system memory. If you are processing files larger than system memory, please contact us at <a class="reference external" href="mailto:bedops&#37;&#52;&#48;stamlab&#46;org">bedops<span>&#64;</span>stamlab<span>&#46;</span>org</a> for details of a temporary workaround.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/conversion/gff2bed.html#gff2bed"><em>gff2bed</em></a> conversion script</p>
<ul class="simple">
<li>Resolved <code class="docutils literal"><span class="pre">IndexError</span></code> exceptions by fixing header support, bringing script in line with <a class="reference external" href="http://www.sequenceontology.org/gff3.shtml">v1.21 GFF3 spec</a>.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/conversion/bam2bed.html#bam2bed"><em>bam2bed</em></a> and <a class="reference internal" href="reference/file-management/conversion/sam2bed.html#sam2bed"><em>sam2bed</em></a> conversion scripts</p>
<ul class="simple">
<li>Rewritten <code class="docutils literal"><span class="pre">bam2*</span></code> and <code class="docutils literal"><span class="pre">sam2*</span></code> scripts from <code class="docutils literal"><span class="pre">bash</span></code> into Python (v2.7+ support).</li>
<li>Improved BAM and SAM input validation against the <a class="reference external" href="http://samtools.sourceforge.net/SAMv1.pdf">v1.4 SAM spec</a>.</li>
<li>New <code class="docutils literal"><span class="pre">--split</span></code> option prints reads with <code class="docutils literal"><span class="pre">N</span></code> CIGAR operations as separated BED elements.</li>
<li>New <code class="docutils literal"><span class="pre">--all-reads</span></code> option prints all reads, mapped and unmapped.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/set-operations/bedextract.html#bedextract"><em>bedextract</em></a></p>
<ul class="simple">
<li>Fixed <code class="docutils literal"><span class="pre">stdin</span></code> bug with <a class="reference internal" href="reference/set-operations/bedextract.html#bedextract"><em>bedextract</em></a>.</li>
</ul>
</li>
<li><p class="first">New documentation via <a class="reference external" href="readthedocs.org">readthedocs.org</a>.</p>
<ul class="simple">
<li>Documentation is now part of the BEDOPS distribution, instead of being a separate download.</li>
<li>We use <a class="reference external" href="readthedocs.org">readthedocs.org</a> to host indexed and searchable HTML.</li>
<li><a class="reference external" href="https://readthedocs.org/projects/bedops/downloads/">PDF and eBook</a> documents are also available for download.</li>
<li>Documentation is refreshed and simplified, with new installation and compilation guides.</li>
</ul>
</li>
<li><p class="first">OS X compilation improvements</p>
<ul>
<li><p class="first">We have made changes to the OS X build process for half of the BEDOPS binaries, which allows direct compilation with Clang/LLVM (part of the Apple Xcode distribution). Those binaries now use Apple&#8217;s system-level C++ library, instead of GNU&#8217;s <code class="docutils literal"><span class="pre">libstdc++</span></code>.</p>
<p>This change means that we require Mac OS X 10.7 (&#8220;Lion&#8221;) or greater—we do not support 10.6 at this time.</p>
<p>Compilation is faster and simpler, and we can reduce the size and complexity of Mac OS X builds and installer packages. By using Apple&#8217;s C++ library, we also reduce the likelihood of missing library errors. When this process is completed for the remaining binaries, it will no longer be necessary to install GCC 4.7+ (by way of MacPorts or other package managers) in order to build BEDOPS on OS X, nor will we have to bundle <code class="docutils literal"><span class="pre">libstdc++</span></code> with the installer.</p>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-2-0b">
<h3>3.2.5. v2.2.0b<a class="headerlink" href="#v2-2-0b" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Fixed bug with OS X installer&#8217;s post-installation scripts.</li>
</ul>
</div>
<div class="section" id="v2-2-0">
<h3>3.2.6. v2.2.0<a class="headerlink" href="#v2-2-0" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>May 22, 2013</strong></p>
<ul class="simple">
<li>Updated packages<ul>
<li>Precompiled packages are now available for Linux (32- and 64-bit) and Mac OS X 10.6-10.8 (32- and 64-bit) hosts.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><em>Starch v2 test suite</em></a><ul>
<li>We have added a test suite for the Starch archive toolkit with the source download. Test inputs include randomized BED data generated from chromosome and bounds data stored on UCSC servers as well as static FIMO search results. Tests put <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a>, <a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><em>unstarch</em></a> and <a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><em>starchcat</em></a> through various usage scenarios. Please refer to the Starch-specific Makefiles and the test target and subfolder&#8217;s <cite>README</cite> doc for more information.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><em>starchcat</em></a><ul>
<li>Resolves bug with <code class="docutils literal"><span class="pre">--gzip</span></code> option, allowing updates of <code class="docutils literal"><span class="pre">gzip</span></code> -backed v1.2 and v1.5 archives to the <a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><em>v2 Starch format</em></a> (either <code class="docutils literal"><span class="pre">bzip2</span></code> - or <code class="docutils literal"><span class="pre">gzip</span></code> -backed).</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><em>unstarch</em></a><ul>
<li>Resolves bug with extraction of <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>Starch</em></a> archive made from BED files with four or more columns. A condition where the total length of additional columns exceeds a certain number of characters would result in extracted data in those columns being cut off. As an example, this could affect Starch archives made from the raw, uncut output of GTF- and GFF- <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a>.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a><ul>
<li>We have partially reverted <a class="reference internal" href="reference/file-management/conversion/wig2bed.html#wig2bed"><em>wig2bed</em></a>, providing a Bash shell wrapper to the original C binary. This preserves consistency of command-line options across the conversion suite, while making use of the C binary to recover performance lost from the Python-based v2.1 revision of <a class="reference internal" href="reference/file-management/conversion/wig2bed.html#wig2bed"><em>wig2bed</em></a> (which at this time is no longer supported). (Thanks to Matt Maurano for reporting this issue.)</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-1-1">
<h3>3.2.7. v2.1.1<a class="headerlink" href="#v2-1-1" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>May 3, 2013</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a><ul>
<li>Major performance improvements made in v2.1.1, such that current <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a> now operates as fast or faster than the v1.2.5 version of <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a>!</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><em>bedops</em></a><ul>
<li>Resolves bug with <code class="docutils literal"><span class="pre">--partition</span></code> option.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a><ul>
<li>All v2.1.0 Python-based scripts now include fix for <code class="docutils literal"><span class="pre">SIGPIPE</span></code> handling, such that use of <code class="docutils literal"><span class="pre">head</span></code> or other common UNIX utilities to process buffered standard output no longer yields <code class="docutils literal"><span class="pre">IOError</span></code> exceptions. (Thanks to Matt Maurano for reporting this bug.)</li>
</ul>
</li>
<li>32-bit Linux binary support<ul>
<li>Pre-built Linux binaries are now available for end users with 32-bit workstations.</li>
</ul>
</li>
</ul>
<p>Other issues fixed:</p>
<ul class="simple">
<li>Jansson tarball no longer includes already-compiled libraries that could potentially interfere with 32-bit builds.</li>
<li>Minor changes to conversion script test suite to exit with useful error code on successful completion of test.</li>
</ul>
</div>
<div class="section" id="v2-1-0">
<h3>3.2.8. v2.1.0<a class="headerlink" href="#v2-1-0" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 22, 2013</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><em>bedops</em></a><ul>
<li>New <code class="docutils literal"><span class="pre">--partition</span></code> operator efficiently generates disjoint segments made from genomic boundaries of all overlapping inputs.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a><ul>
<li>All scripts now use <a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a> behind the scenes to output sorted BED output, ready for use with BEDOPS utilities. It is no longer necessary to pipe data to or otherwise post-process converted data with <a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a>.</li>
<li>New <a class="reference internal" href="reference/file-management/conversion/psl2bed.html#psl2bed"><em>psl2bed</em></a> conversion script, converting <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format2">PSL-formatted UCSC BLAT output</a> to BED.</li>
<li>New <a class="reference internal" href="reference/file-management/conversion/wig2bed.html#wig2bed"><em>wig2bed</em></a> conversion script written in Python.</li>
<li>New <code class="docutils literal"><span class="pre">*2starch</span></code> <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a> offered for all <code class="docutils literal"><span class="pre">*2bed</span></code> scripts, which output Starch v2 archives.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/closest-features.html#closest-features"><em>closest-features</em></a><ul>
<li>Replaced <code class="docutils literal"><span class="pre">--shortest</span></code> option name with <code class="docutils literal"><span class="pre">--closest</span></code>, for clarity. (Old scripts which use <code class="docutils literal"><span class="pre">--shortest</span></code> will continue to work with the deprecated option name for now. We advise editing pipelines, as needed.)</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a><ul>
<li>Improved error checking for interleaved records. This also makes use of <code class="docutils literal"><span class="pre">*2starch</span></code> conversion scripts with the <code class="docutils literal"><span class="pre">--do-not-sort</span></code> option safer.</li>
</ul>
</li>
<li>Improved Mac OS X support<ul>
<li>New Mac OS X package installer makes installation of BEDOPS binaries and scripts very easy for OS X 10.6 - 10.8 hosts.</li>
<li>Installer resolves fatal library errors seen by some end users of older OS X BEDOPS releases.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-0-0b">
<h3>3.2.9. v2.0.0b<a class="headerlink" href="#v2-0-0b" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 19, 2013</strong></p>
<ul class="simple">
<li>Added <a class="reference internal" href="usage-examples/starchcluster.html#starchcluster"><em>Efficiently creating Starch-formatted archives with a cluster</em></a> script variant which supports task distribution with <a class="reference external" href="http://www.gnu.org/software/parallel/">GNU Parallel</a>.</li>
<li>Fixed minor problem with <a class="reference internal" href="reference/file-management/conversion/bam2bed.html#bam2bed"><em>bam2bed</em></a> and <a class="reference internal" href="reference/file-management/conversion/sam2bed.html#sam2bed"><em>sam2bed</em></a> conversion scripts.</li>
</ul>
</div>
<div class="section" id="v2-0-0a">
<h3>3.2.10. v2.0.0a<a class="headerlink" href="#v2-0-0a" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 7, 2013</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a><ul>
<li>Takes in Starch-formatted archives as input, as well as raw BED (i.e., it is no longer required to extract a Starch archive to an intermediate, temporary file or named pipe before applying operations).</li>
<li>New <code class="docutils literal"><span class="pre">--chrom</span></code> operator jumps to and operates on information for specified chromosome only.</li>
<li>New <code class="docutils literal"><span class="pre">--echo-map-id-uniq</span></code> operator lists unique IDs from overlapping mapping elements.</li>
<li>New <code class="docutils literal"><span class="pre">--max-element</span></code> and <code class="docutils literal"><span class="pre">--min-element</span></code> operators return the highest or lowest scoring overlapping map element.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><em>bedops</em></a><ul>
<li>Takes in Starch-formatted archives as input, as well as raw BED.</li>
<li>New <code class="docutils literal"><span class="pre">--chrom</span></code> operator jumps to and operates on information for specified chromosome only.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/closest-features.html#closest-features"><em>closest-features</em></a><ul>
<li>Takes in Starch-formatted archives as input, as well as raw BED.</li>
<li>New <code class="docutils literal"><span class="pre">--chrom</span></code> operator jumps to and operates on information for specified chromosome only.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a> and <code class="docutils literal"><span class="pre">bbms</span></code><ul>
<li>New <code class="docutils literal"><span class="pre">--max-mem</span></code> option to limit system memory on large BED inputs.</li>
<li>Incorporated <code class="docutils literal"><span class="pre">bbms</span></code> functionality into <a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a> with use of <code class="docutils literal"><span class="pre">--max-mem</span></code> operator.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a>, <a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><em>starchcat</em></a> and <a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><em>unstarch</em></a><ul>
<li>New metadata enhancements to Starch-format archival and extraction, including: <code class="docutils literal"><span class="pre">--note</span></code>, <code class="docutils literal"><span class="pre">--elements</span></code>, <code class="docutils literal"><span class="pre">--bases</span></code>, <code class="docutils literal"><span class="pre">--bases-uniq</span></code>, <code class="docutils literal"><span class="pre">--list-chromosomes</span></code>, <code class="docutils literal"><span class="pre">--archive-timestamp</span></code>, <code class="docutils literal"><span class="pre">--archive-type</span></code> and <code class="docutils literal"><span class="pre">--archive-version</span></code> (see <code class="docutils literal"><span class="pre">--help</span></code> to <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a>, <a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><em>starchcat</em></a> and <a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><em>unstarch</em></a> binaries, or view the documentation for these applications for more detail).</li>
<li>Adds 20-35% performance boost to creating Starch archives with <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><em>starch</em></a> utility.</li>
<li>New documentation with technical overview of the Starch format specification.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><em>conversion scripts</em></a><ul>
<li>New <a class="reference internal" href="reference/file-management/conversion/gtf2bed.html#gtf2bed"><em>gtf2bed</em></a> conversion script, converting GTF (v2.2) to BED.</li>
</ul>
</li>
<li>Scripts are now part of main download; it is no longer necessary to download the BEDOPS companion separately.</li>
</ul>
</div>
<div class="section" id="v1-2-5b">
<h3>3.2.11. v1.2.5b<a class="headerlink" href="#v1-2-5b" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 14, 2013</strong></p>
<ul class="simple">
<li>Adds support for Apple 32- and 64-bit Intel hardware running OS X 10.5 through 10.8.</li>
<li>Adds <code class="docutils literal"><span class="pre">README</span></code> for companion download.</li>
<li>Removes some obsolete code.</li>
</ul>
</div>
<div class="section" id="v1-2-5">
<h3>3.2.12. v1.2.5<a class="headerlink" href="#v1-2-5" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>October 13, 2012</strong></p>
<ul class="simple">
<li>Fixed unusual bug with <a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><em>unstarch</em></a>, where an extra (and incorrect) line of BED data can potentially be extracted from an archive.</li>
<li>Updated companion download with updated <a class="reference internal" href="reference/file-management/conversion/bam2bed.html#bam2bed"><em>bam2bed</em></a> and <a class="reference internal" href="reference/file-management/conversion/sam2bed.html#sam2bed"><em>sam2bed</em></a> conversion scripts to address 0-indexing error with previous revisions.</li>
</ul>
</div>
<div class="section" id="v1-2-3">
<h3>3.2.13. v1.2.3<a class="headerlink" href="#v1-2-3" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>August 17, 2012</strong></p>
<ul class="simple">
<li>Added <code class="docutils literal"><span class="pre">--indicator</span></code> option to <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><em>bedmap</em></a>.</li>
<li>Assorted changes to conversion scripts and associated companion download.</li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/logo_with_label_v3.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">3. Revision history</a><ul>
<li><a class="reference internal" href="#current-version">3.1. Current version</a><ul>
<li><a class="reference internal" href="#v2-4-3">3.1.1. v2.4.3</a></li>
</ul>
</li>
<li><a class="reference internal" href="#previous-versions">3.2. Previous versions</a><ul>
<li><a class="reference internal" href="#v2-4-2">3.2.1. v2.4.2</a></li>
<li><a class="reference internal" href="#v2-4-1">3.2.2. v2.4.1</a></li>
<li><a class="reference internal" href="#v2-4-0">3.2.3. v2.4.0</a></li>
<li><a class="reference internal" href="#v2-3-0">3.2.4. v2.3.0</a></li>
<li><a class="reference internal" href="#v2-2-0b">3.2.5. v2.2.0b</a></li>
<li><a class="reference internal" href="#v2-2-0">3.2.6. v2.2.0</a></li>
<li><a class="reference internal" href="#v2-1-1">3.2.7. v2.1.1</a></li>
<li><a class="reference internal" href="#v2-1-0">3.2.8. v2.1.0</a></li>
<li><a class="reference internal" href="#v2-0-0b">3.2.9. v2.0.0b</a></li>
<li><a class="reference internal" href="#v2-0-0a">3.2.10. v2.0.0a</a></li>
<li><a class="reference internal" href="#v1-2-5b">3.2.11. v1.2.5b</a></li>
<li><a class="reference internal" href="#v1-2-5">3.2.12. v1.2.5</a></li>
<li><a class="reference internal" href="#v1-2-3">3.2.13. v1.2.3</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="installation.html"
                        title="previous chapter">2. Installation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="usage-examples.html"
                        title="next chapter">4. Usage examples</a></p>
<div class="sourcelink">
  <a href="../_sources/content/revision-history.txt" rel="nofollow">
    Show page source
  </a>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>

        </div>
      </div>
      <div class="clearer"></div>
    </div>
  <div class="related bottom">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="installation.html" title="Previous document">2. Installation</a>
        </li>
        <li>
          <a href="usage-examples.html" title="Next document">4. Usage examples</a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../index.html">Home</a></li> 
    </ul>
  </nav>
  </div>
  <footer id="pagefooter">&copy; 2011-2015, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a>
      1.3b2
        with the <a href="http://github.com/irskep/sphinx-better-theme">
          better</a> theme.

  </footer>

  
  </body>
</html>