
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>3. Revision history &#8212; BEDOPS v2.4.36</title>
    <link rel="stylesheet" href="../_static/better.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="4. Usage examples" href="usage-examples.html" />
    <link rel="prev" title="2. Installation" href="installation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  </head><body>
    <header id="pageheader"><h1><a href="../index.html ">
        BEDOPS v2.4.36
    </a></h1></header>
  <div class="related top">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="installation.html" title="Previous document">2. Installation</a>
        </li>
        <li>
          <a href="usage-examples.html" title="Next document">4. Usage examples</a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../index.html">Home</a></li> 
    </ul>
  </nav>
  </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="revision-history">
<span id="id1"></span><h1>3. Revision history<a class="headerlink" href="#revision-history" title="Permalink to this headline">¶</a></h1>
<p>This page summarizes some of the more important changes between releases.</p>
<div class="section" id="current-version">
<span id="revision-history-of-current-version"></span><h2>3.1. Current version<a class="headerlink" href="#current-version" title="Permalink to this headline">¶</a></h2>
<div class="section" id="v2-4-36">
<h3>3.1.1. v2.4.36<a class="headerlink" href="#v2-4-36" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>May 2, 2019</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a><ul>
<li>Resolved an issue preventing use of a <code class="code docutils literal notranslate"><span class="pre">bash</span></code> process substitution or Unix named pipe in the reference position: <em>i.e.</em>, <code class="code docutils literal notranslate"><span class="pre">bedmap</span> <span class="pre">--options</span> <span class="pre">&lt;(processToGenerateReferenceElements)</span> <span class="pre">map.bed</span></code> and similar would issue incorrect output. Thanks to Wouter Meuleman and others for reports and test input.</li>
<li>To avoid mapping problems, map elements should not contain spaces in the ID or subsequent non-interval fields. Use of the <code class="code docutils literal notranslate"><span class="pre">--ec</span></code> can help identify problems in map input, at the cost of a longer runtime. The documentation is clarified to warn users about avoiding spaces in map input. Thanks to Wouter Meuleman for the report and providing test input.</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--unmapped-val</span> <span class="pre">&lt;val&gt;</span></code> option, where <code class="code docutils literal notranslate"><span class="pre">&lt;val&gt;</span></code> replaces the empty string output of <code class="code docutils literal notranslate"><span class="pre">--echo-map*</span></code> operations when there are no mapped elements. The <code class="code docutils literal notranslate"><span class="pre">--min/max-element</span></code> operators will give results as before (the empty string).</li>
</ul>
</li>
<li>:ref:General<ul>
<li>Reduced <code class="code docutils literal notranslate"><span class="pre">warning:</span> <span class="pre">zero</span> <span class="pre">as</span> <span class="pre">null</span> <span class="pre">pointer</span> <span class="pre">constant</span> <span class="pre">[-Wzero-as-null-pointer-constant]</span></code> compiler warnings via Clang.</li>
<li>Begun work on a comprehensive test suite for BEDOPS applications. Tests are available via source distribution in <code class="code docutils literal notranslate"><span class="pre">${root}/tests</span></code> and can be run by entering <code class="code docutils literal notranslate"><span class="pre">make</span></code> in this directory.</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="previous-versions">
<h2>3.2. Previous versions<a class="headerlink" href="#previous-versions" title="Permalink to this headline">¶</a></h2>
<div class="section" id="v2-4-35">
<h3>3.2.1. v2.4.35<a class="headerlink" href="#v2-4-35" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>May 2, 2018</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a><ul>
<li>When compressing records, if the last interval in the former chromosome is identical to the first interval of the next chromosome, then a test on the sort order of the remainder string of that interval is applied (incorrectly). This is patched to test that chromosome names are identical before applying sort order rules. Thanks to Andrew Nishida for the report and for providing test input.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-34">
<h3>3.2.2. v2.4.34<a class="headerlink" href="#v2-4-34" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 26, 2018</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>In <a class="reference external" href="https://github.com/bedops/bedops/issues/208">Issue 208</a> builds of <a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a> would exit with an error state when converting SAM input with newline-delimited records longer than the 5 MB per-thread I/O buffer. The <code class="code docutils literal notranslate"><span class="pre">C2B_THREAD_IO_BUFFER_SIZE</span></code> constant is now set to the suite-wide <code class="code docutils literal notranslate"><span class="pre">TOKENS_MAX_LENGTH</span></code> value, which should make it easier to compile custom builds of BEDOPS that support very-long line lengths. Thanks to Erich Schwarz for the initial report.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starchstrip.html#starchstrip"><span class="std std-ref">starchstrip</span></a><ul>
<li>When <cite>starchstrip</cite> is compiled with a C compiler, <code class="code docutils literal notranslate"><span class="pre">qsort</span></code> uses a comparator that works correctly on the input chromosome list. When compiled with a C++ compiler (such as when building the larger BEDOPS toolkit), a different comparator is used that does not make variables of the correct type, and so the <code class="code docutils literal notranslate"><span class="pre">qsort</span></code> result is garbage, leading to missing chromosomes. Thanks to Jemma Nelson for the initial report.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-33">
<h3>3.2.3. v2.4.33<a class="headerlink" href="#v2-4-33" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 9, 2018</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Resolved <a class="reference external" href="https://github.com/bedops/bedops/issues/207">Issue 207</a> where a megarow build of <a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a> would raise segmentation faults when converting SAM input. This build uses constants that define a longer BED line length. Arrays ended up using more stack than available, resulting in segmentation faults. This issue could potentially affect conversion of any data with the megarow build of <a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a>. Arrays using megarow-constants were replaced with heap- or dynamically-allocated pointers. Thanks to Erich Schwarz for the initial report.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-32">
<h3>3.2.4. v2.4.32<a class="headerlink" href="#v2-4-32" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>March 14, 2018</strong></p>
<ul class="simple">
<li>New build type (128-bit precision floating point arithmetic, <code class="code docutils literal notranslate"><span class="pre">float128</span></code>)<ul>
<li>A new build type adds support for <code class="code docutils literal notranslate"><span class="pre">long</span> <span class="pre">double</span></code> or 128-bit floating point operations on measurement values in <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a>, such as is used with score operators like: <code class="code docutils literal notranslate"><span class="pre">--min</span></code>, <code class="code docutils literal notranslate"><span class="pre">--max</span></code>, <code class="code docutils literal notranslate"><span class="pre">--min-element</span></code>, <code class="code docutils literal notranslate"><span class="pre">--max-element</span></code>, <code class="code docutils literal notranslate"><span class="pre">--mean</span></code>, and so on.</li>
<li>This build includes support for measurements on values ranging from approximately ±6.48e−4966 to ±6.48e4966 (<a class="reference external" href="https://en.wikipedia.org/wiki/Denormal_number">subnormal</a>), or ±1.19e-4932 to ±1.19e4932 (normal), which enables <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a> to handle, for example, lower p-values without log- or other transformation preprocessing steps. The article on <a class="reference external" href="https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format">quadruple precision</a> can be useful for technical review.</li>
<li>For comparison, the current “non-float128” typical and megarow builds allow measurements on values from approximately ±5e−324 to ±5e324 (subnormal) or ±2.23e-308 to ±2.23e308 (normal). Please refer to the article on <a class="reference external" href="https://en.wikipedia.org/wiki/Double-precision_floating-point_format">double precision</a> for more technical detail.</li>
<li>Please use <code class="code docutils literal notranslate"><span class="pre">make</span> <span class="pre">float128</span> <span class="pre">&amp;&amp;</span> <span class="pre">make</span> <span class="pre">install_float128</span></code> to install this build type.</li>
<li>This build type combines support for quadruple, 128-bit precision floats with the <code class="code docutils literal notranslate"><span class="pre">typical</span></code> build type for handling “typical” BED4+ style line lengths. At this time, “megarow” support is not enabled with higher precision floats.</li>
<li>This build will use more memory to store floating-point values with higher precision, and processing those data will require more computation time. It is recommended that this build be used only if analyses require a higher level of precision than what the <code class="code docutils literal notranslate"><span class="pre">double</span></code> type allows.</li>
</ul>
</li>
<li>OS X (Darwin) megarow build<ul>
<li>Some applications packaged in the OS X installer or compiled via the OS X command-line developer toolkit lacked <a class="reference external" href="http://bedops.readthedocs.io/en/latest/content/revision-history.html#v2-4-27">megarow</a> build support, despite those flags being specified in the parent Makefile. These applications specifically were affected: <code class="code docutils literal notranslate"><span class="pre">bedextract</span></code>, <code class="code docutils literal notranslate"><span class="pre">bedmap</span></code>, and <code class="code docutils literal notranslate"><span class="pre">convert2bed</span></code>. These binaries rely on wider suite-wide constants and data types that the megarow build variety specifies. The Darwin-specific Makefiles have been fixed to resolve this build issue, so that all OS X BEDOPS binaries should now be able to compile in the correct megarow-specific settings.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-31">
<h3>3.2.5. v2.4.31<a class="headerlink" href="#v2-4-31" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>March 8, 2018</strong></p>
<ul class="simple">
<li>User forum<ul>
<li>BEDOPS user forum moved domains from <a class="reference external" href="http://bedops.stamlab.org">http://bedops.stamlab.org</a> to <a class="reference external" href="https://bedops.altius.org">https://bedops.altius.org</a></li>
<li>Testing out administrator approval requirement for new forum accounts, to help try to reduce visits from spammers.</li>
</ul>
</li>
<li>Documentation<ul>
<li>Updated Homebrew installation instructions per <a class="reference external" href="https://github.com/bedops/bedops/issues/202">issue 202</a> (thanks to user EricFromCanada).</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/wig2bed.html#wig2bed"><span class="std std-ref">wig2bed</span></a><ul>
<li>Increased maximum length of chromosome name buffer to suite-wide <code class="code docutils literal notranslate"><span class="pre">TOKEN_CHR_MAX_LENGTH</span></code> value, to reduce likelihood of segmentation faults (thanks to user ma-diroma).</li>
</ul>
</li>
<li>General<ul>
<li>Updated copyright dates in source and headers.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-30">
<h3>3.2.6. v2.4.30<a class="headerlink" href="#v2-4-30" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>November 25, 2017</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a><ul>
<li>Errors are no longer reported when error checking is enabled and running in non-fast mode, when a fully-nested element is detected. This follows up on <a class="reference external" href="https://github.com/bedops/bedops/issues/199">issue 199</a>.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a><ul>
<li>Previously, a chromosome record in a Starch archive would result in corrupted metadata, if the chromosome is larger than <code class="code docutils literal notranslate"><span class="pre">UINT32_MAX</span></code> bytes (~4.3GB) in size when compressed. This limitation is now removed, and a single chromosome (when compressed in a Starch archive) can be up to <code class="code docutils literal notranslate"><span class="pre">UINT64_MAX</span></code> bytes in size.</li>
<li>The <code class="code docutils literal notranslate"><span class="pre">starch</span></code> binary does more stringent input checks for the character lengths of ID and remainder strings, which must be no larger than 2<sup>ID_EXPONENT</sup> - 1 and 2<sup>REST_EXPONENT</sup> - 1 characters in length. (These constants are specific to the build-time settings in the Makefile and in the app-wide constants.) This follows up on <a class="reference external" href="https://github.com/bedops/bedops/issues/195">issue 195</a>.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><span class="std std-ref">starchcat</span></a><ul>
<li>Previously, a chromosome record in a Starch archive would result in corrupted metadata, if the chromosome is larger than <code class="code docutils literal notranslate"><span class="pre">UINT32_MAX</span></code> bytes (~4.3GB) in size when compressed. This limitation is now removed, and a single chromosome (when compressed in a Starch archive) can be up to <code class="code docutils literal notranslate"><span class="pre">UINT64_MAX</span></code> bytes in size.</li>
<li>More stringent memory management and stricter adherance to BEDOPS-wide constants, to help reduce likelihood of pointer access out of bounds and incidence of segfaults.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><span class="std std-ref">unstarch</span></a><ul>
<li>The <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code> binary implements the character length constants of ID and remainder strings, specific to the build-time settings in the Makefile and in the app-wide constants. This follows up on <a class="reference external" href="https://github.com/bedops/bedops/issues/195">issue 195</a>.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a><ul>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--unique</span></code> (<code class="code docutils literal notranslate"><span class="pre">-u</span></code>) and <code class="code docutils literal notranslate"><span class="pre">--duplicates</span></code> (<code class="code docutils literal notranslate"><span class="pre">-d</span></code>) options to only print unique and duplicate in sorted output, to mimic the behavior of <code class="code docutils literal notranslate"><span class="pre">sort</span> <span class="pre">-u</span></code> and <code class="code docutils literal notranslate"><span class="pre">uniq</span> <span class="pre">-d</span></code> Unix tools. This follows up on <a class="reference external" href="https://github.com/bedops/bedops/issues/196">issue 196</a>.</li>
<li>Switched compile-time, stack-allocated <code class="code docutils literal notranslate"><span class="pre">char</span></code> arrays to runtime, heap-based pointers. Timing tests on shuffled FIMO datasets suggest the impact from having to allocate space for buffers at runtime is very minimal. Moving from stack to heap will help avoid segfaults from running into OS-level stack limits, when BEDOPS-level constants change the maximum line length to something larger than the stack.</li>
</ul>
</li>
<li>Revision testing<ul>
<li>Starch suite tests were updated for v2.2 archives created via v2.4.30 binaries (Linux, libc 2.22).</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-29">
<h3>3.2.7. v2.4.29<a class="headerlink" href="#v2-4-29" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>September 26, 2017</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a><ul>
<li>Increased megarow build ID length up to 2<sup>18</sup>.</li>
<li>Changed behavior of mapping to return mapped items in sort order provided in inputs. This follows up on <a class="reference external" href="https://github.com/bedops/bedops/issues/198">issue 198</a>.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><span class="std std-ref">unstarch</span></a><ul>
<li>Changed behavior of <code class="code docutils literal notranslate"><span class="pre">--is-starch</span></code> option to always return a successful exit code of <code class="code docutils literal notranslate"><span class="pre">0</span></code> whether or not the input file is a Starch archive. It will now be up to the person running this option to test the 0 (false) or 1 (true) value printed to the standard output stream. This follows up on <a class="reference external" href="https://github.com/bedops/bedops/issues/197">issue 197</a>.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-28">
<h3>3.2.8. v2.4.28<a class="headerlink" href="#v2-4-28" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>August 18, 2017</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a><ul>
<li>Patched <a class="reference external" href="https://github.com/bedops/bedops/issues/191">issue 191</a> where <code class="code docutils literal notranslate"><span class="pre">--wmean</span></code> option was not recognized.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/bedextract.html#bedextract"><span class="std std-ref">bedextract</span></a><ul>
<li>Updated documentation with fixed usage statement.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a><ul>
<li>Patched typo in <code class="code docutils literal notranslate"><span class="pre">update-sort-bed-starch-slurm.py</span></code> script.</li>
<li>Fixed bug with <code class="code docutils literal notranslate"><span class="pre">--max-mem</span></code> on properly ordering things on fourth and subsequent columns, when the genomic intervals are the same.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a><ul>
<li>Updated Makefiles to remove <cite>lib</cite> on <cite>clean</cite> target and to help prevent <code class="code docutils literal notranslate"><span class="pre">ARCH</span></code> variable from getting clobbered by third-party package managers.</li>
</ul>
</li>
<li>Build process<ul>
<li>Updated the OS X installer XML to resolve missing asset links.</li>
<li>Updated the <code class="code docutils literal notranslate"><span class="pre">module_binaries</span></code> target to copy over <code class="code docutils literal notranslate"><span class="pre">starchcluster_*</span></code> and <code class="code docutils literal notranslate"><span class="pre">starch-diff</span></code> assets for <code class="code docutils literal notranslate"><span class="pre">modules</span></code> distributions.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-27">
<h3>3.2.9. v2.4.27<a class="headerlink" href="#v2-4-27" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>July 17, 2017</strong></p>
<p>This revision of BEDOPS includes significant performance improvements for core tools: <code class="code docutils literal notranslate"><span class="pre">bedops</span></code>, <code class="code docutils literal notranslate"><span class="pre">bedmap</span></code>, and <code class="code docutils literal notranslate"><span class="pre">closest-features</span></code>. Performance tests were done with whole-genome TRANSFAC FIMO scans, with cache purges in between trials.</p>
<p>Pre-built binaries for Darwin and GNU/Linux targets include both the default <code class="code docutils literal notranslate"><span class="pre">typical</span></code> and <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> builds of BEDOPS. The program names that you are accustomed to will remain as-is, but the binaries will exist as symbolic links pointing to the <code class="code docutils literal notranslate"><span class="pre">typical</span></code> builds. These links can be repointed to the <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> builds by calling <code class="code docutils literal notranslate"><span class="pre">switch-BEDOPS-binary-type</span> <span class="pre">--megarow</span></code>, which will set the usual BEDOPS binaries to link to the <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> builds. One can run <code class="code docutils literal notranslate"><span class="pre">switch-BEDOPS-binary-type</span> <span class="pre">--typical</span></code> at any time to revert to the default (<code class="code docutils literal notranslate"><span class="pre">typical</span></code>) builds.</p>
<p>The top-level Makefile includes some new variables for those who choose to build from source. The <code class="code docutils literal notranslate"><span class="pre">JPARALLEL</span></code> variable sets the number of CPUs to use in parallel when compiling BEDOPS, which can speed compilation time dramatically. The <code class="code docutils literal notranslate"><span class="pre">MASSIVE_REST_EXP</span></code>, <code class="code docutils literal notranslate"><span class="pre">MASSIVE_ID_EXP</span></code>, and <code class="code docutils literal notranslate"><span class="pre">MASSIVE_CHROM_EXP</span></code> are used when building the <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> to support any required row lengths (build using <code class="code docutils literal notranslate"><span class="pre">make</span> <span class="pre">megarow</span></code>).  These are the exponents (the <em>n</em> in 2<sup>n</sup>) for holding all characters after chromosome, start, and stop fields, the ID field (column 4, typically), and the chromosome field (column 1).</p>
<p>To simplify distribution and support, we have removed pre-built 32-bit program versions in this release. These can be built from source by specifying the correct <code class="code docutils literal notranslate"><span class="pre">ARCH</span></code> value in the top-level Makefile. For OS X, our package installer now requires OS X version 10.10 or greater.</p>
<p>Application-level notes follow:</p>
<ul>
<li><p class="first"><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a></p>
<ul class="simple">
<li>Performance of <code class="code docutils literal notranslate"><span class="pre">bedops</span></code> tool improved, doing typical work in <strong>76.5%</strong> of the time of all previous versions.</li>
<li>Performance of <code class="code docutils literal notranslate"><span class="pre">-u</span></code>/<code class="code docutils literal notranslate"><span class="pre">--everything</span></code> has improved, doing the same work in only <strong>55.6%</strong> of the time of previous versions when given a large number of input files.</li>
<li>The <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> build of this application handles input files with very long rows (4M+ characters). Such input might arise from conversion of very-long-read BAM files to BED via <code class="code docutils literal notranslate"><span class="pre">bam2bed</span></code>, such as those that may come from Nanopore or PacBio MinION platforms. This build requires more runtime memory than the default (<code class="code docutils literal notranslate"><span class="pre">typical</span></code>) build. Pertinent variables for <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> execution can be modified through the make system without changing source.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a></p>
<ul class="simple">
<li>Performance of <code class="code docutils literal notranslate"><span class="pre">bedmap</span></code> tool improved, doing the same work in <strong>86.7%</strong> of the time of all previous versions.</li>
<li>Automatically use <code class="code docutils literal notranslate"><span class="pre">--faster</span></code> option when <code class="code docutils literal notranslate"><span class="pre">--exact</span></code> is used as the overlap criterion, or if the input files are formatted as Starch archives, no fully-nested elements exist in the archives, and the overlap criterion supports <code class="code docutils literal notranslate"><span class="pre">--faster</span></code> (such as <code class="code docutils literal notranslate"><span class="pre">--bp-ovr</span></code>, <code class="code docutils literal notranslate"><span class="pre">--exact</span></code>, and <code class="code docutils literal notranslate"><span class="pre">--range</span></code>).</li>
<li>The <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> build target handles input files with very long rows (4M+ characters). Such input might arise from conversion of very-long-read BAM files to BED via <code class="code docutils literal notranslate"><span class="pre">bam2bed</span></code>, such as those that may come from Nanopore or PacBio MinION platforms. This build requires more runtime memory than the default (<code class="code docutils literal notranslate"><span class="pre">typical</span></code>) build. Pertinent variables for <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> execution can be modified through the make system without changing source.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">--min-memory</span></code> option for use when the reference file has very large regions, and the map file has many small regions that fall within those larger regions. One example is when <code class="code docutils literal notranslate"><span class="pre">--range</span> <span class="pre">100000</span></code> is used and the map file consists of whole-genome motif scan hits.  Memory overhead can be reduced to that used by all previous versions, up to and including v2.4.26.</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--faster</span></code> automatically when <code class="code docutils literal notranslate"><span class="pre">--exact</span></code> is used, which is robust even when nested elements exist in inputs.  Similarly, <code class="code docutils literal notranslate"><span class="pre">--faster</span></code> is used automatically when inputs are Starch-formatted archives, none of which have nested elements (see <code class="code docutils literal notranslate"><span class="pre">unstarch</span> <span class="pre">--has-nested</span></code>) when the overlap criterion allows for <code class="code docutils literal notranslate"><span class="pre">--faster</span></code>.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/set-operations/closest-features.html#closest-features"><span class="std std-ref">closest-features</span></a></p>
<ul class="simple">
<li>Performance of <code class="code docutils literal notranslate"><span class="pre">closest-features</span></code> tool has been improved, doing the same work in <strong>87.7%</strong> of the time of all previous versions.</li>
<li>The <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> build target is available to compile a version of the program that can handle input files with very long rows (4M+ characters).  This requires more runtime memory than the default build.  Pertinent variables can be modified through the make system without editing source.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a></p>
<p>Numerous internal changes, including giving line functors the ability to resize the destination (write) buffer in mid-stream, along with increased integration with BEDOPS-wide constants. Destination buffer resizing is particularly useful when converting very-long-read BAM files containing numerous D (deletion) operations, such as when used with the new <code class="code docutils literal notranslate"><span class="pre">--split-with-deletions</span></code> option.</p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/psl2bed.html#psl2bed"><span class="std std-ref">psl2bed</span></a><ul>
<li>Migrated storage of PSL conversion state from stack to heap, which helps address segmentation faults on OS X (thanks to <a class="reference external" href="mailto:rmartson&#37;&#52;&#48;Biostars">rmartson<span>&#64;</span>Biostars</a> for the bug report).</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/bam2bed.html#bam2bed"><span class="std std-ref">bam2bed</span></a> and <a class="reference internal" href="reference/file-management/conversion/sam2bed.html#sam2bed"><span class="std std-ref">sam2bed</span></a><ul>
<li>Increased thread I/O heap buffer size to reduce likelihood of overflows while parsing reads from Nanopore and PacBio platforms.</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--split-with-deletions</span></code> option to split spliced junctions by <code class="code docutils literal notranslate"><span class="pre">N</span></code> and <code class="code docutils literal notranslate"><span class="pre">D</span></code> CIGAR operations. The <code class="code docutils literal notranslate"><span class="pre">--split</span></code> option now splits only on <code class="code docutils literal notranslate"><span class="pre">N</span></code> operations.</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--reduced</span></code> option to print first six columns of BED data to standard output.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/gff2bed.html#gff2bed"><span class="std std-ref">gff2bed</span></a><ul>
<li>Resolved issue parsing GFF input with <code class="code docutils literal notranslate"><span class="pre">##FASTA</span></code> directive.</li>
</ul>
</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a></p>
<ul class="simple">
<li>The <code class="code docutils literal notranslate"><span class="pre">megarow</span></code> build target is available to compile a version of the program that can handle input files with very long rows (4M+ characters).  This requires more runtime memory than the default build.  The pertinent variables can be modified through the make system without changing source.  This is useful for converting ultra-long reads from Nanopore and PacBio sequencing platforms to BED via <code class="code docutils literal notranslate"><span class="pre">bam2bed</span></code> / <code class="code docutils literal notranslate"><span class="pre">convert2bed</span></code>.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a></p>
<ul class="simple">
<li>Fixed a potential segmentation fault result with <code class="code docutils literal notranslate"><span class="pre">--header</span></code> usage.</li>
</ul>
</li>
<li><p class="first">Starch C++ API</p>
<ul class="simple">
<li>Fixed output from <code class="code docutils literal notranslate"><span class="pre">bedops</span> <span class="pre">-u</span></code> (<code class="code docutils literal notranslate"><span class="pre">--everything</span></code>, or multiset union) on two or more Starch archives, where the remainder string was not being cleared correctly.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/starch-diff.html#starch-diff"><span class="std std-ref">starch-diff</span></a></p>
<ul class="simple">
<li>Improved usage statement to clarify output (cf. <a class="reference external" href="https://github.com/bedops/bedops/issues/180">Issue 180</a>).</li>
</ul>
</li>
<li><p class="first">Clang warnings</p>
<ul class="simple">
<li>Resolved compilation warnings for several binaries.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-26">
<h3>3.2.10. v2.4.26<a class="headerlink" href="#v2-4-26" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>March 14, 2017</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/compression/starchstrip.html#starchstrip"><span class="std std-ref">starchstrip</span></a><ul>
<li>New utility to efficiently filter a Starch archive, including or excluding records by specified chromosome names, without doing expensive extraction and recompression. This follows up on <a class="reference external" href="https://stamlab.slack.com/archives/bedops/p1487878245000103">internal discussion</a> on the Altius Slack channel.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch-diff.html#starch-diff"><span class="std std-ref">starch-diff</span></a><ul>
<li>Fixed testing logic in <code class="code docutils literal notranslate"><span class="pre">starch-diff</span></code> for certain archives. Thanks to Shane Neph for the report.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><span class="std std-ref">starchcat</span></a><ul>
<li>Fixed possible condition where too many variables on the stack can cause a stack overload on some platforms, leading to a fatal segmentation fault. Improved logic for updating v2.1 to v2.2 Starch archives.</li>
</ul>
</li>
<li>Starch C++ API<ul>
<li>Patched gzip-backed Starch archive extraction issue. Thanks to Matt Maurano for the bug report.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">update-sort-bed-migrate-candidates</span></a><ul>
<li>Added detailed logging via <code class="code docutils literal notranslate"><span class="pre">--debug</span></code> option.</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--bedops-root-dir</span></code> option to allow specifying where all BEDOPS binaries are stored. This setting can be overruled on a per-binary basis by adding <code class="code docutils literal notranslate"><span class="pre">--bedextract-path</span></code>, <code class="code docutils literal notranslate"><span class="pre">--sort-bed-path</span></code>, etc.</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--non-recursive-search</span></code> option to restrict search for BED and Starch candidates to the top-level of the specified parent directory <code class="code docutils literal notranslate"><span class="pre">--parent-dir</span></code> option.</li>
<li>Further simplification and customization of parameters sent to <code class="code docutils literal notranslate"><span class="pre">update-sort-bed-slurm</span></code> and <code class="code docutils literal notranslate"><span class="pre">update-sort-bed-starch-slurm</span></code> cluster scripts, as well as logging and variable name improvements to those two scripts.</li>
<li>Thanks again to Matt Maurano for ongoing feedback and suggestions on functionality and fixes.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/gtf2bed.html#gtf2bed"><span class="std std-ref">gtf2bed</span></a><ul>
<li>Resolved segmentation fault with certain inputs, in follow-up to <a class="reference external" href="http://bedops.uwencode.org/forum/index.php?topic=136.0">this BEDOPS Forum post</a>. Thanks to zebasilio for the report and feedback.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-25">
<h3>3.2.11. v2.4.25<a class="headerlink" href="#v2-4-25" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 15, 2017</strong></p>
<ul>
<li><p class="first"><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a></p>
<ul class="simple">
<li>Patch for RepeatMasker inputs with blank lines that have no spaces. This follows up on <a class="reference external" href="https://github.com/bedops/bedops/issues/173">Issue 173</a>. Thanks to saketkc for the bug report.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">update-sort-bed-migrate-candidates</span></a></p>
<p>The <code class="code docutils literal notranslate"><span class="pre">update-sort-bed-migrate-candidates</span></code> utility recursively searches into the specified directory for BED and Starch files which fail a <code class="code docutils literal notranslate"><span class="pre">sort-bed</span> <span class="pre">--check-sort</span></code> test. Those files which fail this test can have their paths written to a text file for further downstream processing, or the end user can decide to apply an immediate resort on those files, either locally or via a SLURM-managed cluster. Grateful thanks to Matt Maurano for input and testing.</p>
<p>See <code class="code docutils literal notranslate"><span class="pre">update-sort-bed-migrate-candidates</span> <span class="pre">--help</span></code> for more information, or review the <a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a> documentation.</p>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">update-sort-bed-starch-slurm</span></a></p>
<p>This is an adjunct to the <code class="code docutils literal notranslate"><span class="pre">update-sort-bed-slurm</span></code> utility, which resorts the provided Starch file and writes a new file. (The <code class="code docutils literal notranslate"><span class="pre">update-sort-bed-slurm</span></code> utility only takes in BED files as input and writes BED as output.)</p>
</li>
</ul>
</div>
<div class="section" id="v2-4-24">
<h3>3.2.12. v2.4.24<a class="headerlink" href="#v2-4-24" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 6, 2017</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/compression/starch-diff.html#starch-diff"><span class="std std-ref">starch-diff</span></a><ul>
<li>The <code class="code docutils literal notranslate"><span class="pre">starch-diff</span></code> utility compares signatures of two or more v2.2+ Starch archives. This tool tests all chromosomes or one specified chromosome. It returns a zero exit code, if the signature(s) are identical, or a non-zero error exit code, if one or more signature(s) are dissimilar.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">update-sort-bed-slurm</span></a><ul>
<li>The <code class="code docutils literal notranslate"><span class="pre">update-sort-bed-slurm</span></code> convenience utility provides a parallelized update of the sort order on BED files sorted with pre-v2.4.20 sort-bed, for users with a SLURM job scheduler and associated cluster. See <code class="code docutils literal notranslate"><span class="pre">update-sort-bed-slurm</span> <span class="pre">--help</span></code> for more details.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Patched a memory leak in VCF conversion. Thanks to ehsueh for the bug report.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-23">
<h3>3.2.13. v2.4.23<a class="headerlink" href="#v2-4-23" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 30, 2017</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><span class="std std-ref">unstarch</span></a><ul>
<li>Fixed bug where missing signature from pre-v2.2 Starch archives would cause a fatal metadata error. Thanks to Shane Neph and Eric Rynes for the bug report.</li>
<li>Improved logic reporting signature mismatches when input v2.2 archive lacks signature (<em>e.g.</em>, for a v2.2 archive made with <code class="code docutils literal notranslate"><span class="pre">--omit-signature</span></code>).</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a> and <a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><span class="std std-ref">starchcat</span></a><ul>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--omit-signature</span></code> option to compress without creating a per-chromosome data integrity signature. While this reduces compression time, this eliminates the verification benefits of the data integrity signature.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-22">
<h3>3.2.14. v2.4.22<a class="headerlink" href="#v2-4-22" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 25, 2017</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Fixed heap corruption in GFF conversion. Thanks to J. Miguel Mendez (ObjectiveTruth) for the bug report.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-21">
<h3>3.2.15. v2.4.21<a class="headerlink" href="#v2-4-21" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 23, 2017</strong></p>
<ul>
<li><p class="first"><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a></p>
<ul class="simple">
<li>New <code class="code docutils literal notranslate"><span class="pre">--wmean</span></code> operation offers a weighted mean calculation. The “weight” is derived from the proportion of the reference element covered by overlapping map elements: <em>i.e.</em>, a map element that covers more of the reference element has its signal given a larger weight or greater impact than another map element with a shorter overlap.</li>
<li>Measurement values in <code class="code docutils literal notranslate"><span class="pre">bedmap</span></code> did not allow <code class="code docutils literal notranslate"><span class="pre">+</span></code> in the exponent (both <code class="code docutils literal notranslate"><span class="pre">-</span></code> worked and no <code class="code docutils literal notranslate"><span class="pre">+</span></code> for a positive value.  Similarly, out in front of the number, <code class="code docutils literal notranslate"><span class="pre">+</span></code> was previously not allowed. Shane Neph posted the report and the fix.</li>
<li>The <code class="code docutils literal notranslate"><span class="pre">--min-element</span></code> and <code class="code docutils literal notranslate"><span class="pre">--max-element</span></code> operations in <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a> now process elements in unambiguous order. Former behavior is moved to the operations <code class="code docutils literal notranslate"><span class="pre">--min-element-rand</span></code> and <code class="code docutils literal notranslate"><span class="pre">--max-element-rand</span></code>, respectively.</li>
<li>Fixed issue with use of <code class="code docutils literal notranslate"><span class="pre">--echo-overlap-size</span></code> with <code class="code docutils literal notranslate"><span class="pre">--multidelim</span></code> (cf. <a class="reference external" href="https://github.com/bedops/bedops/issues/165">Issue 165</a>). Shane Neph posted the fix. Thanks to Jeff Vierstra for the bug report!</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a></p>
<ul class="simple">
<li>Fixed issue with <code class="code docutils literal notranslate"><span class="pre">--chop</span></code> where complement operation could potentially be included. Shane Neph posted the fix.</li>
<li>The <code class="code docutils literal notranslate"><span class="pre">bedops</span> <span class="pre">--everything</span></code> or <code class="code docutils literal notranslate"><span class="pre">bedops</span> <span class="pre">-u</span></code> (union) operation now writes elements to standard output in unambiguous sort order. If any data are contained in fourth or subsequent fields, a lexicographical sort on that data is applied for resolving order of interval matches.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a></p>
<ul>
<li><p class="first">Improved sort times from replacing quicksort (<code class="code docutils literal notranslate"><span class="pre">std::qsort</span></code>) with inlined C++ <code class="code docutils literal notranslate"><span class="pre">std::sort</span></code>.</p>
</li>
<li><p class="first">Sorting of BED input now leads to unambiguous result when two or more elements have the same genomic interval (chromosome name and start and stop position), but different content in remaining columns (ID, score, etc.).</p>
<p>Formerly, elements with the same genomic interval that have different content in fourth and subsequent columns could be printed in a non-consistent ordering on repeated sorts. A deterministic sort order facilitates the use of data integrity functions on sorted BED and Starch data.</p>
</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="usage-examples/starchcluster.html#starchcluster"><span class="std std-ref">starchcluster</span></a></p>
<ul class="simple">
<li>A SLURM-ready version of the <code class="code docutils literal notranslate"><span class="pre">starchcluster</span></code> script was added to help SLURM job scheduler users with parallelizing the creation of Starch archives.</li>
</ul>
</li>
<li><p class="first">Parallel <a class="reference internal" href="reference/file-management/conversion/parallel_bam2bed.html#parallel-bam2bed"><span class="std std-ref">bam2bed</span></a> and <a class="reference internal" href="reference/file-management/conversion/parallel_bam2starch.html#parallel-bam2starch"><span class="std std-ref">bam2starch</span></a></p>
<ul class="simple">
<li>SLURM-ready versions of these scripts were added to help parallelize the conversion of BAM to BED files (<code class="code docutils literal notranslate"><span class="pre">bam2bed_slurm</span></code>) or to Starch archives (<code class="code docutils literal notranslate"><span class="pre">bam2starch_slurm</span></code>).</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><span class="std std-ref">unstarch</span></a></p>
<ul>
<li><p class="first">Added <code class="code docutils literal notranslate"><span class="pre">--signature</span></code> option to report the Base64-encoded SHA-1 data integrity signature of the Starch-transformed bytes of a specified chromosome, or to report the signature of the metadata string as well as the signatures of all chromosomes, if unspecified.</p>
</li>
<li><p class="first">Added <code class="code docutils literal notranslate"><span class="pre">--verify-signature</span></code> option to compare the “expected” Base64-encoded SHA-1 data integrity signature stored within the archive’s metadata with the “observed” data integrity signature generated from extracting the specified chromosome.</p>
<p>If the observed and expected signatures differ, then this suggests that the chromosome record may be corrupted in some way; <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code> will exit with a non-zero error code. If the signatures agree, the archive data should be intact and <cite>unstarch</cite> will exit with a helpful notice and a zero error code.</p>
<p>If no chromosome is specified, <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code> will loop through all chromosomes in the archive metadata, comparing observed and expected values for each chromosome record. Upon completion, error and progress messages will be reported to the standard error stream, and <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code> will exit with a zero error code, if all signatures match, or a non-zero exit state, if one or more signatures do not agree.</p>
</li>
<li><p class="first">The output from the <code class="code docutils literal notranslate"><span class="pre">--list</span></code> option includes a <code class="code docutils literal notranslate"><span class="pre">signature</span></code> column to report the data integrity signature of all Starch-transformed chromosome data.</p>
</li>
<li><p class="first">The output from the <code class="code docutils literal notranslate"><span class="pre">--list-json</span></code> option includes a <code class="code docutils literal notranslate"><span class="pre">signature</span></code> key in each chromosome record in the archive metadata, reporting the same information.</p>
</li>
<li><p class="first">The <code class="code docutils literal notranslate"><span class="pre">--is-starch</span></code> option now quits with a non-zero exit code, if the specified input file is not a Starch archive.</p>
</li>
<li><p class="first">The <code class="code docutils literal notranslate"><span class="pre">--elements-max-string-length</span></code> option reports the length of the longest string within the specified chromosome, or the longest string over all chromosomes (if no chromosome name is specified).</p>
</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a></p>
<ul>
<li><p class="first">Added <code class="code docutils literal notranslate"><span class="pre">--report-progress=N</span></code> option to (optionally) report compression of the Nth element of the current chromosome to standard error stream.</p>
</li>
<li><p class="first">As a chromosome is compressed, the input Starch-transform bytes are continually run through a SHA-1 hash function. The resulting data integrity signature is stored as a Base64-encoded string in the output archive’s metadata. Signatures can be compared between and within archives to help better ensure the data integrity of the archive.</p>
</li>
<li><p class="first">Fixed <code class="code docutils literal notranslate"><span class="pre">--header</span></code> transform bug reported in <a class="reference external" href="https://github.com/bedops/bedops/issues/161">Issue 161</a>. Thanks to Shane Neph for the bug report!</p>
</li>
<li><p class="first">Added chromosome name and “remainder” order tests to <code class="code docutils literal notranslate"><span class="pre">STARCH2_transformHeaderlessBEDInput</span></code> and <code class="code docutils literal notranslate"><span class="pre">STARCH2_transformHeaderedBEDInput</span></code> functions.</p>
<p>Compression with <code class="code docutils literal notranslate"><span class="pre">starch</span></code> ends with a fatal error, should any of the following comparison tests fail:</p>
<ol class="arabic simple">
<li>The chromosome names are not lexicographically ordered (<em>e.g.</em>, <code class="code docutils literal notranslate"><span class="pre">chr1</span></code> records coming after <code class="code docutils literal notranslate"><span class="pre">chr2</span></code> records indicates the data are not correctly sorted).</li>
<li>The start position of an input element is less than the start position of a previous input element on the same chromosome (<em>e.g.</em>, <code class="code docutils literal notranslate"><span class="pre">chr1:1000-1234</span></code> coming after <code class="code docutils literal notranslate"><span class="pre">chr1:2000-2345</span></code> is not correctly sorted).</li>
<li>The stop positions of two or more input elements are not in ascending order when their start positions are equal (<em>e.g.</em>, <code class="code docutils literal notranslate"><span class="pre">chr1:1000-1234</span></code> coming after <code class="code docutils literal notranslate"><span class="pre">chr1:1000-2345</span></code> is not correctly sorted).</li>
<li>The start and stop positions of two or more input elements are equivalent, and their “remainders” (fourth and subsequent columns) are not in ascending order (<em>e.g.</em>, <code class="code docutils literal notranslate"><span class="pre">chr1:1000-1234:id-0</span></code> coming after <code class="code docutils literal notranslate"><span class="pre">chr1:1000-1234:id-1</span></code> is not correctly sorted).</li>
</ol>
<p>If the sort order of the input data is unknown or uncertain, simply use <code class="code docutils literal notranslate"><span class="pre">sort-bed</span></code> to generate the correct ordering and pipe the output from that to <code class="code docutils literal notranslate"><span class="pre">starch</span></code>, <em>e.g.</em> <code class="code docutils literal notranslate"><span class="pre">$</span> <span class="pre">cat</span> <span class="pre">elements.bed</span> <span class="pre">|</span> <span class="pre">sort-bed</span> <span class="pre">-</span> <span class="pre">|</span> <span class="pre">starch</span> <span class="pre">-</span> <span class="pre">&gt;</span> <span class="pre">elements.starch</span></code>.</p>
</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><span class="std std-ref">starchcat</span></a></p>
<ul class="simple">
<li>Added <code class="code docutils literal notranslate"><span class="pre">--report-progress=N</span></code> option to (optionally) report compression of the <em>N</em> th element of the current chromosome to standard error stream.</li>
<li>As in <code class="code docutils literal notranslate"><span class="pre">starch</span></code>, at the conclusion of compressing a chromosome made from one or more input Starch archives, the input Starch-transform bytes are continually run through a SHA-1 hash function. The resulting data integrity signature is stored as a Base64-encoded string in the chromosome’s entry in the new archive’s metadata.</li>
<li>As in <code class="code docutils literal notranslate"><span class="pre">starch</span></code>, if data should need to be extracted and recompressed, the output is written so that the order is unambiguous: ascending lexicographic ordering on chromosome names, numerical ordering on start positions, the same ordering on stop positions where start positions match, and ascending lexicographic ordering on the remainder of the BED element (fourth and subsequent columns, where present).</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a></p>
<ul>
<li><p class="first">Improvements in support for BAM/SAM inputs with larger-sized reads, as would come from alignments made from data collected from third-generation sequencers. Simulated read datasets were generated using <a class="reference external" href="https://bitbucket.org/genomeinformatics/simlord/">SimLoRD</a>. Tests have been performed on simulated hg19 data up to 100kb read lengths.</p>
<p>Improvements allow:</p>
<ul class="simple">
<li>conversion of dynamic number of CIGAR operations (up to system memory)</li>
<li>conversion of dynamically-sized read fields (up to system memory and inter-thread buffer allocations)</li>
</ul>
<p>These patches follow up on bug reports in <a class="reference external" href="https://github.com/bedops/bedops/issues/157">Issue 157</a>.</p>
</li>
<li><p class="first">Improvements in support for VCF inputs, to allow aribtrary-sized fields (up to system memory and inter-thread buffer allocations), which should reduce or eliminate segmentation faults from buffer overruns on fields larger than former stack defaults.</p>
</li>
<li><p class="first">Improvements in support for GFF inputs, to allow aribtrary-sized fields (up to system memory and inter-thread buffer allocations), which should reduce or eliminate segmentation faults from buffer overruns on fields larger than former stack defaults.</p>
</li>
<li><p class="first">Improvements in support for GTF inputs, to allow aribtrary-sized fields (up to system memory and inter-thread buffer allocations), which should reduce or eliminate segmentation faults from buffer overruns on fields larger than former stack defaults.</p>
</li>
</ul>
</li>
<li><p class="first">Testing</p>
<ul class="simple">
<li>Our use of Travis CI to automate testing of builds now includes Clang on <a class="reference external" href="https://docs.travis-ci.com/user/osx-ci-environment/">their OS X environment</a>.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-20">
<h3>3.2.16. v2.4.20<a class="headerlink" href="#v2-4-20" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>July 27, 2016</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Increased memory allocation for maximum number of per-read CIGAR operations in BAM and SAM conversion to help improve stability. Thanks to Adam Freedman for the report!</li>
<li>Improved reliability of gene ID parsing from GTF input, where <code class="code docutils literal notranslate"><span class="pre">gene_id</span></code> field may be positioned at start, middle, or end of attributes string, or may be empty. Thanks to blaiseli for the report!</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-19">
<h3>3.2.17. v2.4.19<a class="headerlink" href="#v2-4-19" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>May 9, 2016</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Fixed bug in BAM and SAM parallel conversion scripts (<code class="code docutils literal notranslate"><span class="pre">*_gnuParallel</span></code> and <code class="code docutils literal notranslate"><span class="pre">*_sge</span></code>) with inputs containing chromosome names without <code class="code docutils literal notranslate"><span class="pre">chr</span></code> prefix. Thanks to Eric Haugen for the bug report!</li>
</ul>
</li>
<li>Starch C++ API<ul>
<li>Fixed bug with extraction of bzip2- and gzip-backed archives with all other non-primary Starch tools (all tools except <code class="code docutils literal notranslate"><span class="pre">starch</span></code>, <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code>, <code class="code docutils literal notranslate"><span class="pre">starchcat</span></code>, and <code class="code docutils literal notranslate"><span class="pre">sort-bed</span></code>). Thanks to Eric Haugen for the bug report!</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-18">
<h3>3.2.18. v2.4.18<a class="headerlink" href="#v2-4-18" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 28, 2016</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Fixed compile warnings.</li>
<li>Fixed bug in BAM and SAM conversion with optional field line overflow. Thanks to Jemma Nelson for the bug report!</li>
</ul>
</li>
<li>General documentation improvements<ul>
<li>Updated OS X Installer and Github release instructions</li>
<li>Added thank-you to Feng Tian for bug report</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-17">
<h3>3.2.19. v2.4.17<a class="headerlink" href="#v2-4-17" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 26, 2016</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/bam2bed.html#bam2bed"><span class="std std-ref">bam2bed</span></a> and <a class="reference internal" href="reference/file-management/conversion/sam2bed.html#sam2bed"><span class="std std-ref">sam2bed</span></a><ul>
<li>Improved parsing of non-split BAM and SAM inputs.</li>
</ul>
</li>
<li>Docker container build target added for Debian<ul>
<li>Thanks to Leo Comitale (Poldo) for writing a Makefile target and spec for creating a BEDOPS Docker container for the Debian target.</li>
</ul>
</li>
<li>Starch C++ API<ul>
<li>Fixed bug with extraction of bzip2- and gzip-backed archives with all other non-primary Starch tools (all tools except <code class="code docutils literal notranslate"><span class="pre">starch</span></code>, <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code>, <code class="code docutils literal notranslate"><span class="pre">starchcat</span></code>, and <code class="code docutils literal notranslate"><span class="pre">sort-bed</span></code>). Thanks to Feng Tian for reports.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-16">
<h3>3.2.20. v2.4.16<a class="headerlink" href="#v2-4-16" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 5, 2016</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a><ul>
<li>Added new <code class="code docutils literal notranslate"><span class="pre">--echo-ref-row-id</span></code> option to report reference row ID elements.</li>
</ul>
</li>
<li>Starch C++ API<ul>
<li>Fixed bug with extraction of archives made with <code class="code docutils literal notranslate"><span class="pre">starch</span> <span class="pre">--gzip</span></code> (thanks to Brad Gulko for the bug report and Paul Verhoeven and Peter Weir for compile and testing assistance).</li>
</ul>
</li>
<li>General improvements<ul>
<li>Small improvements to build cleanup targets.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-15">
<h3>3.2.21. v2.4.15<a class="headerlink" href="#v2-4-15" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 21, 2016</strong></p>
<ul class="simple">
<li>Docker container build target added for CentOS 7<ul>
<li>Thanks to Leo Comitale (Poldo) for writing a Makefile target and spec for creating a BEDOPS Docker container for CentOS 7.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Fixed buffer overflows in <code class="code docutils literal notranslate"><span class="pre">convert2bed</span></code> to improve conversion reliability for VCF files (thanks to Jared Andrews and Kousik Kundu for bug reports).</li>
</ul>
</li>
<li>General improvements<ul>
<li>Improved OS X 10.11 build process.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-14">
<h3>3.2.22. v2.4.14<a class="headerlink" href="#v2-4-14" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 21, 2015</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Fixed missing <code class="code docutils literal notranslate"><span class="pre">samtools</span></code> variable references in cluster conversion scripts (thanks to Brad Gulko for the bug report).</li>
</ul>
</li>
<li>General suite-wide improvements<ul>
<li>Fixed exception error message for <code class="code docutils literal notranslate"><span class="pre">stdin</span></code> check (thanks to Brad Gulko for the bug report).</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-13">
<h3>3.2.23. v2.4.13<a class="headerlink" href="#v2-4-13" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 20, 2015</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a><ul>
<li>Resolved issue in using <code class="code docutils literal notranslate"><span class="pre">--ec</span></code> with <code class="code docutils literal notranslate"><span class="pre">bedops</span></code> when reading from <code class="code docutils literal notranslate"><span class="pre">stdin</span></code> (thanks to Brad Gulko for the bug report).</li>
</ul>
</li>
<li>General suite-wide improvements<ul>
<li>Addressed inconsistency with constants defined for the suite at the extreme end of the limits we allow for coordinate values (thanks again to Brad Gulko for the report).</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-12">
<h3>3.2.24. v2.4.12<a class="headerlink" href="#v2-4-12" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>March 13, 2015</strong></p>
<ul>
<li><p class="first"><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a></p>
<ul>
<li><p class="first">Checks have been added to determine if an integer argument is a file in the current working directory, before interpreting that argument as an overlap criterion for <code class="code docutils literal notranslate"><span class="pre">-e</span></code> and <code class="code docutils literal notranslate"><span class="pre">-n</span></code> options.</p>
<p>To reduce ambiguity, if an integer is used as a file input, <code class="code docutils literal notranslate"><span class="pre">bedops</span></code> issues a warning of the interpretation and provides guidance on how to force that value to instead be used as an overlap specification, if desired (thanks to E. Rynes for the pointer).</p>
</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a></p>
<ul class="simple">
<li>Added support for <code class="code docutils literal notranslate"><span class="pre">--prec</span></code> / <code class="code docutils literal notranslate"><span class="pre">--sci</span></code> with <code class="code docutils literal notranslate"><span class="pre">--min-element</span></code> and <code class="code docutils literal notranslate"><span class="pre">--max-element</span></code> operations (thanks to E. Rynes for the pointer).</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a> | <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a> | <a class="reference internal" href="reference/set-operations/closest-features.html#closest-features"><span class="std std-ref">closest-features</span></a></p>
<ul class="simple">
<li>Added support for <code class="code docutils literal notranslate"><span class="pre">bash</span></code> process substitution/named pipes with specification of <code class="code docutils literal notranslate"><span class="pre">--chrom</span></code> and/or <code class="code docutils literal notranslate"><span class="pre">--ec</span></code> options (thanks to B. Gulko for the bug report).</li>
<li>Fixed code that extracts <code class="code docutils literal notranslate"><span class="pre">gzip</span></code>-backed Starch archives from <code class="code docutils literal notranslate"><span class="pre">bedops</span></code> and other core tools (thanks again to B. Gulko for the bug report).</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a></p>
<ul class="simple">
<li>Switched <code class="code docutils literal notranslate"><span class="pre">matches</span></code> and <code class="code docutils literal notranslate"><span class="pre">qSize</span></code> fields in order of <code class="code docutils literal notranslate"><span class="pre">psl2bed</span></code> output. Refer to documentation for new field order.</li>
<li>Added null sentinel to GTF ID value.</li>
<li>To help reduce the chance of buffer overflows, the <code class="code docutils literal notranslate"><span class="pre">convert2bed</span></code> tool increases the maximum field length from 8191 to 24575 characters to allow parsing of inputs with longer field length, such as very long attributes from <a class="reference external" href="https://www.vectorbase.org/download/aedes-aegypti-liverpoolbasefeaturesaaegl33gff3gz">mosquito GFF3</a> data (thanks to T. Karginov for the bug report).</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-11">
<h3>3.2.25. v2.4.11<a class="headerlink" href="#v2-4-11" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 24, 2015</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Fixed bug in <code class="code docutils literal notranslate"><span class="pre">psl2bed</span></code> where <code class="code docutils literal notranslate"><span class="pre">matches</span></code> column value was truncated by one character. Updated unit tests. Thanks to M. Wirthlin for the bug report.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-10">
<h3>3.2.26. v2.4.10<a class="headerlink" href="#v2-4-10" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 23, 2015</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a><ul>
<li>In addition to checking chromosome interleaving, the <code class="code docutils literal notranslate"><span class="pre">starch</span></code> tool now enforces <code class="code docutils literal notranslate"><span class="pre">sort-bed</span></code> sort ordering on BED input and exits with an <code class="code docutils literal notranslate"><span class="pre">EINVAL</span></code> POSIX error code if the data are not sorted correctly.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a><ul>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--zero-indexed</span></code> option to <code class="code docutils literal notranslate"><span class="pre">wig2bed</span></code> and <code class="code docutils literal notranslate"><span class="pre">wig2starch</span></code> wrappers and <code class="code docutils literal notranslate"><span class="pre">convert2bed</span></code> binary, which converts WIG data that are zero-indexed without any coordinate adjustments. This is useful for WIG data sourced from the UCSC Kent tool <code class="code docutils literal notranslate"><span class="pre">bigWigToWig</span></code>, where the <code class="code docutils literal notranslate"><span class="pre">bigWig</span></code> data can potentially be sourced from 0-indexed BAM- or bedGraph-formatted data.</li>
<li>If the WIG input contains any element with a start coordinate of 0, the default use of <code class="code docutils literal notranslate"><span class="pre">wig2bed</span></code>, <code class="code docutils literal notranslate"><span class="pre">wig2starch</span></code> and <code class="code docutils literal notranslate"><span class="pre">convert2bed</span></code> will exit early with an error condition, suggesting the use of <code class="code docutils literal notranslate"><span class="pre">--zero-indexed</span></code>.</li>
<li>Updated copyright date range of wrapper scripts</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-9">
<h3>3.2.27. v2.4.9<a class="headerlink" href="#v2-4-9" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 17, 2015</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a><ul>
<li>Added support for <code class="code docutils literal notranslate"><span class="pre">--check-sort</span></code> to report if input is sorted (or not)</li>
</ul>
</li>
<li>Starch<ul>
<li>Improved support for <code class="code docutils literal notranslate"><span class="pre">starch</span> <span class="pre">--header</span></code>, where header contains tab-delimited fields</li>
</ul>
</li>
<li>Starch C++ API<ul>
<li>Fixed bug with <code class="code docutils literal notranslate"><span class="pre">starch</span> <span class="pre">--header</span></code> functionality, such that BEDOPS core tools (<code class="code docutils literal notranslate"><span class="pre">bedops</span></code>, etc.) would be unable to extract correct data from headered Starch archive</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-8">
<h3>3.2.28. v2.4.8<a class="headerlink" href="#v2-4-8" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 7, 2015</strong></p>
<ul class="simple">
<li>Mac OS X packaging<ul>
<li>Installer signed with <a class="reference external" href="https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man1/productsign.1.html#//apple_ref/doc/man/1/productsign">productsign</a> to pass <a class="reference external" href="http://support.apple.com/en-us/HT202491">OS X Gatekeeper</a></li>
</ul>
</li>
<li>Linux packaging<ul>
<li>SHA1 hashes of each tarball are now part of the <a class="reference external" href="https://github.com/bedops/bedops/releases/">BEDOPS Releases</a> description page, going forwards</li>
</ul>
</li>
<li>Updated copyright dates in source code</li>
</ul>
</div>
<div class="section" id="v2-4-7">
<h3>3.2.29. v2.4.7<a class="headerlink" href="#v2-4-7" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 2, 2015</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a> fixes and improvements<ul>
<li>Fixed <code class="code docutils literal notranslate"><span class="pre">--split</span></code> support in <code class="code docutils literal notranslate"><span class="pre">psl2bed</span></code> (thanks to Marco A.)</li>
<li>Fixed compilation warning regarding comparison of signed and unsigned values</li>
<li>Fixed corrupted <code class="code docutils literal notranslate"><span class="pre">psl2bed</span></code> test inputs</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-6">
<h3>3.2.30. v2.4.6<a class="headerlink" href="#v2-4-6" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 30, 2015</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a> fixes and improvements<ul>
<li>Added support for conversion of the <a class="reference external" href="http://www.sequenceontology.org/resources/gvf.html#summary">GVF file format</a>, including wrapper scripts and unit tests. Refer to the <code class="code docutils literal notranslate"><span class="pre">gvf2bed</span></code> documentation for more information.</li>
<li>Fixed bug in string copy of zero-length element attribute for <code class="code docutils literal notranslate"><span class="pre">gff2bed</span></code> and <code class="code docutils literal notranslate"><span class="pre">gtf2bed</span></code> (GFF and GTF) formats</li>
</ul>
</li>
<li>General fixes and improvements<ul>
<li>Fixed possibly corrupt bzip2, Jansson and zlib tarballs (thanks to rekado, Shane N. and Richard S.)</li>
<li>Fixed typo in <code class="code docutils literal notranslate"><span class="pre">bedextract</span></code> documentation</li>
<li>Fixed broken image in <a class="reference internal" href="overview.html#overview"><span class="std std-ref">Overview</span></a></li>
<li>Removed 19 MB <code class="code docutils literal notranslate"><span class="pre">_build</span></code> intermediate result directory (which should improve overall <code class="code docutils literal notranslate"><span class="pre">git</span> <span class="pre">clone</span></code> time considerably!)</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-5">
<h3>3.2.31. v2.4.5<a class="headerlink" href="#v2-4-5" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 28, 2015</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion/convert2bed.html#convert2bed"><span class="std std-ref">convert2bed</span></a> improvements<ul>
<li>Addition of RepeatMasker annotation output (<code class="code docutils literal notranslate"><span class="pre">.out</span></code>) file conversion support, <code class="code docutils literal notranslate"><span class="pre">rmsk2bed</span></code> and <code class="code docutils literal notranslate"><span class="pre">rmsk2starch</span></code> wrappers, and unit tests</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-4">
<h3>3.2.32. v2.4.4<a class="headerlink" href="#v2-4-4" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 25, 2015</strong></p>
<ul class="simple">
<li>Documentation improvements<ul>
<li>Implemented substantial style changes via <a class="reference external" href="http://github.com/irskep/sphinx-better-theme">A Better Sphinx Theme</a> and various customizations. We also include responsive web style elements to help improve browsing on mobile devices.</li>
<li>Fixes to typos in conversion and other documents.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-3">
<h3>3.2.33. v2.4.3<a class="headerlink" href="#v2-4-3" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>December 18, 2014</strong></p>
<ul class="simple">
<li>Compilation improvements<ul>
<li>Shane Neph put in a great deal of work to enable parallel builds (<em>e.g.</em>, <code class="code docutils literal notranslate"><span class="pre">make</span> <span class="pre">-j</span> <span class="pre">N</span></code> to build various targets in parallel). Depending on the end user’s environment, this can speed up compilation time by a factor of 2, 4 or more.</li>
<li>Fixed numerous compilation warnings of debug builds of <code class="code docutils literal notranslate"><span class="pre">starch</span></code> toolkit under RHEL6/GCC and OS X 10.10.1/LLVM.</li>
</ul>
</li>
<li>New <a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a> features<ul>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--chop</span></code> and <code class="code docutils literal notranslate"><span class="pre">--stagger</span></code> options to “melt” inputs into contiguous or staggered disjoint regions of equivalent size.</li>
<li>For less confusion, arguments for <code class="code docutils literal notranslate"><span class="pre">--element-of</span></code>, <code class="code docutils literal notranslate"><span class="pre">--chop</span></code> and other <code class="code docutils literal notranslate"><span class="pre">bedops</span></code> operations that take numerical modifiers no longer require a leading hyphen character. For instance, <code class="code docutils literal notranslate"><span class="pre">--element-of</span> <span class="pre">1</span></code> is now equivalent to the former usage of <code class="code docutils literal notranslate"><span class="pre">--element-of</span> <span class="pre">-1</span></code>.</li>
</ul>
</li>
<li>New <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a> features<ul>
<li>The <code class="code docutils literal notranslate"><span class="pre">--sweep-all</span></code> option reads through the entire map file without early termination and can help deal with <code class="code docutils literal notranslate"><span class="pre">SIGPIPE</span></code> errors. It adds to execution time, but the penalty is not as severe as with the use of <code class="code docutils literal notranslate"><span class="pre">--ec</span></code>. Using <code class="code docutils literal notranslate"><span class="pre">--ec</span></code> alone will enable error checking, but will now no longer read through the entire map file. The <code class="code docutils literal notranslate"><span class="pre">--ec</span></code> option can be used in conjunction with <code class="code docutils literal notranslate"><span class="pre">--sweep-all</span></code>, with the associated time penalties. (Another method for dealing with issue this is to override how <code class="code docutils literal notranslate"><span class="pre">SIGPIPE</span></code> errors are caught by the interpreter (<code class="code docutils literal notranslate"><span class="pre">bash</span></code>, Python, <em>etc.</em>) and retrapping them or ignoring them. However, it may not a good idea to do this as other situations may arise in production pipelines where it is ideal to trap and handle all I/O errors in a default manner.)</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">--echo-ref-size</span></code> and <code class="code docutils literal notranslate"><span class="pre">--echo-ref-name</span></code> operations report genomic length of reference element, and rename the reference element in <code class="code docutils literal notranslate"><span class="pre">chrom:start-end</span></code> (useful for labeling rows for input for <code class="code docutils literal notranslate"><span class="pre">matrix2png</span></code> or <code class="code docutils literal notranslate"><span class="pre">R</span></code> or other applications).</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/bedextract.html#bedextract"><span class="std std-ref">bedextract</span></a><ul>
<li>Fixed upper bound bug that would cause incorrect output in some cases</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a><ul>
<li>Brand new C99 binary called <code class="code docutils literal notranslate"><span class="pre">convert2bed</span></code>, which wrapper scripts (<code class="code docutils literal notranslate"><span class="pre">bam2bed</span></code>, <em>etc.</em>) now call. No more Python version dependencies, and the C-based rewrite offers massive performance improvements over old Python-based scripts.</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">parallel_bam2starch</span></code> script, which parallelizes creation of <a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><span class="std std-ref">Starch</span></a> archive from very large BAM files in SGE environments.</li>
<li>Added bug fix for missing code in <a class="reference internal" href="usage-examples/starchcluster.html#starchcluster"><span class="std std-ref">starchcluster.gnu_parallel</span></a> script, where the final collation step was missing.</li>
<li>The <code class="code docutils literal notranslate"><span class="pre">vcf2bed</span></code> script now accepts the <code class="code docutils literal notranslate"><span class="pre">--do-not-split</span></code> option, which prints one BED element for all alternate alleles.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><span class="std std-ref">Starch</span></a> archival format and compression/extraction tools<ul>
<li>Added duplicate- and <a class="reference internal" href="reference/set-operations/nested-elements.html#nested-elements"><span class="std std-ref">nested-element</span></a> flags in v2.1 of Starch metadata, which denote if a chromosome contains one or more duplicate and/or nested elements. BED files compressed with <code class="code docutils literal notranslate"><span class="pre">starch</span></code> v2.5 or greater, or Starch archives updated with <code class="code docutils literal notranslate"><span class="pre">starchcat</span></code> v2.5 or greater will include these values in the archive metadata. The <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code> extraction tool offers <code class="code docutils literal notranslate"><span class="pre">--has-duplicate</span></code> and <code class="code docutils literal notranslate"><span class="pre">--has-nested</span></code> options to retrieve these flag values for a specified chromosome (or for all chromosomes).</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--is-starch</span></code> option to <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code> to test if specified input file is a Starch v1 or v2 archive.</li>
<li>Added bug fix for compressing BED files with <code class="code docutils literal notranslate"><span class="pre">starch</span></code>, where the archive would not include the last element of the BED input, if the BED input lacked a trailing newline. The compression tools now include a routine for capturing the last line, if there is no newline.</li>
</ul>
</li>
<li>Documentation improvements<ul>
<li>Remade some image assets throughout the documents to support Retina-grade displays</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-2">
<h3>3.2.34. v2.4.2<a class="headerlink" href="#v2-4-2" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 10, 2014</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a><ul>
<li>Added support for <code class="code docutils literal notranslate"><span class="pre">sort-bed</span> <span class="pre">--tmpdir</span></code> option to conversion scripts, to allow specification of alternative temporary directory for sorted results when used in conjunction with <code class="code docutils literal notranslate"><span class="pre">--max-mem</span></code> option.</li>
<li>Added support for GFF3 files which include a FASTA directive in <code class="code docutils literal notranslate"><span class="pre">gff2bed</span></code> and <code class="code docutils literal notranslate"><span class="pre">gff2starch</span></code> (thanks to Keith Hughitt).</li>
<li>Extended support for Python-based conversion scripts to support use with Python v2.6.2 and forwards, except for <code class="code docutils literal notranslate"><span class="pre">sam2bed</span></code> and <code class="code docutils literal notranslate"><span class="pre">sam2starch</span></code>, which still require Python v2.7 or greater (and under Python3).</li>
<li>Fixed <code class="code docutils literal notranslate"><span class="pre">--insertions</span></code> option in <code class="code docutils literal notranslate"><span class="pre">vcf2bed</span></code> to now report a single-base BED element (thanks to Matt Maurano).</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-4-1">
<h3>3.2.35. v2.4.1<a class="headerlink" href="#v2-4-1" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 26, 2014</strong></p>
<ul>
<li><p class="first"><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a></p>
<ul class="simple">
<li>Added <code class="code docutils literal notranslate"><span class="pre">--fraction-both</span></code> and <code class="code docutils literal notranslate"><span class="pre">--exact</span></code> (<code class="code docutils literal notranslate"><span class="pre">--fraction-both</span> <span class="pre">1</span></code>) to list of compatible overlap options with <code class="code docutils literal notranslate"><span class="pre">--faster</span></code>.</li>
<li>Added 5% performance improvement with <code class="code docutils literal notranslate"><span class="pre">bedmap</span></code> operations without <code class="code docutils literal notranslate"><span class="pre">--faster</span></code>.</li>
<li>Fixed scenario that can yield incorrect results (cf. <a class="reference external" href="https://github.com/bedops/bedops/issues/43">Issue 43</a>).</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a></p>
<ul class="simple">
<li>Added <code class="code docutils literal notranslate"><span class="pre">--tmpdir</span></code> option to allow specification of an alternative temporary directory, when used in conjunction with <code class="code docutils literal notranslate"><span class="pre">--max-mem</span></code> option. This is useful if the host operating system’s standard temporary directory (<em>e.g.</em>, <code class="code docutils literal notranslate"><span class="pre">/tmp</span></code> on Linux or OS X) does not have sufficient space to hold intermediate results.</li>
</ul>
</li>
<li><p class="first">All <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a></p>
<ul class="simple">
<li>Improvements to error handling in Python-based conversion scripts, in the case where no input is specified.</li>
<li>Fixed typos in <code class="code docutils literal notranslate"><span class="pre">gff2bed</span></code> and <code class="code docutils literal notranslate"><span class="pre">psl2bed</span></code> documentation (cf. <a class="reference external" href="https://github.com/bedops/bedops/commit/a091e18">commit a091e18</a>).</li>
</ul>
</li>
<li><p class="first">OS X compilation improvements</p>
<ul>
<li><p class="first">We have completed changes to the OS X build process for the remaining half of the BEDOPS binaries, which now allows direct, full compilation with Clang/LLVM (part of the Apple Xcode distribution).</p>
<p>All OS X BEDOPS binaries now use Apple’s system-level C++ library, instead of GNU’s <code class="code docutils literal notranslate"><span class="pre">libstdc++</span></code>. It is no longer required (or recommended) to use GNU <code class="code docutils literal notranslate"><span class="pre">gcc</span></code> to compile BEDOPS on OS X.</p>
<p>Compilation is faster and simpler, and we can reduce the size and complexity of Mac OS X builds and installer packages. By using Apple’s C++ library, we also eliminate the likelihood of missing library errors.</p>
<p>In the longer term, this gets us closer to moving BEDOPS to using the CMake build system, to further abstract and simplify the build process.</p>
</li>
</ul>
</li>
<li><p class="first">Cleaned up various compilation warnings found with <code class="code docutils literal notranslate"><span class="pre">clang</span></code> / <code class="code docutils literal notranslate"><span class="pre">clang++</span></code> and GCC kits.</p>
</li>
</ul>
</div>
<div class="section" id="v2-4-0">
<h3>3.2.36. v2.4.0<a class="headerlink" href="#v2-4-0" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 9, 2014</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a><ul>
<li>Added new <code class="code docutils literal notranslate"><span class="pre">--echo-map-size</span></code> and <code class="code docutils literal notranslate"><span class="pre">--echo-overlap-size</span></code> options to calculate sizes of mapped elements and overlaps between mapped and reference elements.</li>
<li>Improved performance for all <code class="code docutils literal notranslate"><span class="pre">--echo-map-*</span></code> operations.</li>
<li>Updated documentation.</li>
</ul>
</li>
<li>Major enhancements and fixes to <a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a>:<ul>
<li>Improved performance.</li>
<li>Fixed memory leak.</li>
<li>Added support for millions of distinct chromosomes.</li>
<li>Improved internal estimation of memory usage with <code class="code docutils literal notranslate"><span class="pre">--max-mem</span></code> option.</li>
</ul>
</li>
<li>Added support for compilation on Cygwin (64-bit). Refer to the <a class="reference internal" href="installation.html#installation-via-source-code-on-cygwin"><span class="std std-ref">installation documentation</span></a> for build instructions.</li>
<li><a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><span class="std std-ref">starchcat</span></a><ul>
<li>Fixed embarassing buffer overflow condition that caused segmentation faults on Ubuntu 13.</li>
</ul>
</li>
<li>All <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a><ul>
<li>Python-based scripts no longer use temporary files, which reduces file I/O and improves performance. This change also reduces the need for large amounts of free space in a user’s <code class="code docutils literal notranslate"><span class="pre">/tmp</span></code> folder, particularly relevant for users converting multi-GB BAM files.</li>
<li>We now test for ability to locate <code class="code docutils literal notranslate"><span class="pre">starch</span></code>, <code class="code docutils literal notranslate"><span class="pre">sort-bed</span></code>, <code class="code docutils literal notranslate"><span class="pre">wig2bed_bin</span></code> and <code class="code docutils literal notranslate"><span class="pre">samtools</span></code> in user environment, quitting with the appropriate error state if the dependencies cannot be found.</li>
<li>Improved documentation. In particular, we have added descriptive tables to each script’s documentation page which describe how columns map from original data input to BED output.</li>
<li><a class="reference internal" href="reference/file-management/conversion/bam2bed.html#bam2bed"><span class="std std-ref">bam2bed</span></a> and <a class="reference internal" href="reference/file-management/conversion/sam2bed.html#sam2bed"><span class="std std-ref">sam2bed</span></a><ul>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--custom-tags</span> <span class="pre">&lt;value&gt;</span></code> command-line option to support a comma-separated list of custom tags (cf. <a class="reference external" href="http://www.biostars.org/p/87062/">Biostars discussion</a>), <em>i.e.</em>, tags which are not part of the original SAMtools specification.</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--keep-header</span></code> option to preserve header and metadata as BED elements that use <code class="code docutils literal notranslate"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully “non-lossy”.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/vcf2bed.html#vcf2bed"><span class="std std-ref">vcf2bed</span></a><ul>
<li>Added new <code class="code docutils literal notranslate"><span class="pre">--snvs</span></code>, <code class="code docutils literal notranslate"><span class="pre">--insertions</span></code> and <code class="code docutils literal notranslate"><span class="pre">--deletions</span></code> options that filter VCF variants into three separate subcategories.</li>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--keep-header</span></code> option to preserve header and metadata as BED elements that use <code class="code docutils literal notranslate"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully “non-lossy”.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/gff2bed.html#gff2bed"><span class="std std-ref">gff2bed</span></a><ul>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--keep-header</span></code> option to preserve header and metadata as BED elements that use <code class="code docutils literal notranslate"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully “non-lossy”.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/psl2bed.html#psl2bed"><span class="std std-ref">psl2bed</span></a><ul>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--keep-header</span></code> option to preserve header and metadata as BED elements that use <code class="code docutils literal notranslate"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully “non-lossy”.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion/wig2bed.html#wig2bed"><span class="std std-ref">wig2bed</span></a><ul>
<li>Added <code class="code docutils literal notranslate"><span class="pre">--keep-header</span></code> option to <code class="code docutils literal notranslate"><span class="pre">wig2bed</span></code> binary and <code class="code docutils literal notranslate"><span class="pre">wig2bed</span></code> / <code class="code docutils literal notranslate"><span class="pre">wig2starch</span></code> wrapper scripts, to preserve header and metadata as BED elements that use <code class="code docutils literal notranslate"><span class="pre">_header</span></code> as the chromosome name. This now makes these conversion scripts fully “non-lossy”.</li>
</ul>
</li>
</ul>
</li>
<li>Added OS X uninstaller project to allow end user to more easily remove BEDOPS tools from this platform.</li>
<li>Cleaned up various compilation warnings found with <code class="code docutils literal notranslate"><span class="pre">clang</span></code> / <code class="code docutils literal notranslate"><span class="pre">clang++</span></code> and GCC kits.</li>
</ul>
</div>
<div class="section" id="v2-3-0">
<h3>3.2.37. v2.3.0<a class="headerlink" href="#v2-3-0" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>October 2, 2013</strong></p>
<ul>
<li><p class="first">Migration of BEDOPS code and documentation from Google Code to Github.</p>
<ul class="simple">
<li>Due to changes with Google Code hosting policies at the end of the year, we have decided to change our process for distributing code, packages and documentation. While most of the work is done, we appreciate feedback on any problems you may encounter. Please email us at <a class="reference external" href="mailto:bedops&#37;&#52;&#48;stamlab&#46;org">bedops<span>&#64;</span>stamlab<span>&#46;</span>org</a> with details.</li>
<li>Migration to Github should facilitate requests for code by those who are familiar with <code class="code docutils literal notranslate"><span class="pre">git</span></code> and want to fork our project to submit <a class="reference external" href="https://help.github.com/articles/using-pull-requests">pull requests</a>.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a></p>
<ul class="simple">
<li>General <code class="code docutils literal notranslate"><span class="pre">--ec</span></code> performance improvements.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a></p>
<ul class="simple">
<li>Adds support for the new <code class="code docutils literal notranslate"><span class="pre">--skip-unmapped</span></code> option, which filters out reference elements which do not have mapped elements associated with them. See the end of the <a class="reference internal" href="reference/statistics/bedmap.html#bedmap-score-operations"><span class="std std-ref">score operations</span></a> section of the <a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a> documentation for more detail.</li>
<li>General <code class="code docutils literal notranslate"><span class="pre">--ec</span></code> performance improvements.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a></p>
<ul>
<li><p class="first">Fixed bug with <code class="code docutils literal notranslate"><span class="pre">starch</span></code> where zero-byte BED input (<em>i.e.</em>, an “empty set”) created a truncated and unusable archive. We now put in a “dummy” chromosome for zero-byte input, which <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code> can now unpack.</p>
<p>This should simplify error handling with certain pipelines, specifically where set or other BEDOPS operations yield an “empty set” BED file that is subsequently compressed with <code class="code docutils literal notranslate"><span class="pre">starch</span></code>.</p>
</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><span class="std std-ref">unstarch</span></a></p>
<ul class="simple">
<li>Can now unpack zero-byte (“empty set”) compressed <code class="code docutils literal notranslate"><span class="pre">starch</span></code> archive (see above).</li>
<li>Changed <code class="code docutils literal notranslate"><span class="pre">unstarch</span> <span class="pre">--list</span></code> option to print to <code class="code docutils literal notranslate"><span class="pre">stdout</span></code> stream (this was previously sent to <code class="code docutils literal notranslate"><span class="pre">stderr</span></code>).</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a> metadata library</p>
<ul class="simple">
<li>Fixed array overflow bug with BEDOPS tools that take <a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><span class="std std-ref">starch</span></a> archives as inputs, which affected use of archives as inputs to <code class="code docutils literal notranslate"><span class="pre">closest-features</span></code>, <code class="code docutils literal notranslate"><span class="pre">bedops</span></code> and <code class="code docutils literal notranslate"><span class="pre">bedmap</span></code>.</li>
</ul>
</li>
<li><p class="first">All <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a></p>
<ul class="simple">
<li>Python scripts require v2.7+ or greater.</li>
<li>Improved (more “Pythonic”) error code handling.</li>
<li>Disabled support for <code class="code docutils literal notranslate"><span class="pre">--max-mem</span></code> sort parameter until <a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a> <a class="reference external" href="https://github.com/bedops/bedops/issues/1">issue</a> is resolved. Scripts will continue to sort, but they will be limited to available system memory. If you are processing files larger than system memory, please contact us at <a class="reference external" href="mailto:bedops&#37;&#52;&#48;stamlab&#46;org">bedops<span>&#64;</span>stamlab<span>&#46;</span>org</a> for details of a temporary workaround.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/conversion/gff2bed.html#gff2bed"><span class="std std-ref">gff2bed</span></a> conversion script</p>
<ul class="simple">
<li>Resolved <code class="code docutils literal notranslate"><span class="pre">IndexError</span></code> exceptions by fixing header support, bringing script in line with <a class="reference external" href="http://www.sequenceontology.org/gff3.shtml">v1.21 GFF3 spec</a>.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/file-management/conversion/bam2bed.html#bam2bed"><span class="std std-ref">bam2bed</span></a> and <a class="reference internal" href="reference/file-management/conversion/sam2bed.html#sam2bed"><span class="std std-ref">sam2bed</span></a> conversion scripts</p>
<ul class="simple">
<li>Rewritten <code class="code docutils literal notranslate"><span class="pre">bam2*</span></code> and <code class="code docutils literal notranslate"><span class="pre">sam2*</span></code> scripts from <code class="code docutils literal notranslate"><span class="pre">bash</span></code> into Python (v2.7+ support).</li>
<li>Improved BAM and SAM input validation against the <a class="reference external" href="http://samtools.sourceforge.net/SAMv1.pdf">v1.4 SAM spec</a>.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">--split</span></code> option prints reads with <code class="code docutils literal notranslate"><span class="pre">N</span></code> CIGAR operations as separated BED elements.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">--all-reads</span></code> option prints all reads, mapped and unmapped.</li>
</ul>
</li>
<li><p class="first"><a class="reference internal" href="reference/set-operations/bedextract.html#bedextract"><span class="std std-ref">bedextract</span></a></p>
<ul class="simple">
<li>Fixed <code class="code docutils literal notranslate"><span class="pre">stdin</span></code> bug with <code class="code docutils literal notranslate"><span class="pre">bedextract</span></code>.</li>
</ul>
</li>
<li><p class="first">New documentation via <a class="reference external" href="readthedocs.org">readthedocs.org</a>.</p>
<ul class="simple">
<li>Documentation is now part of the BEDOPS distribution, instead of being a separate download.</li>
<li>We use <a class="reference external" href="readthedocs.org">readthedocs.org</a> to host indexed and searchable HTML.</li>
<li><a class="reference external" href="https://readthedocs.org/projects/bedops/downloads/">PDF and eBook</a> documents are also available for download.</li>
<li>Documentation is refreshed and simplified, with new installation and compilation guides.</li>
</ul>
</li>
<li><p class="first">OS X compilation improvements</p>
<ul>
<li><p class="first">We have made changes to the OS X build process for half of the BEDOPS binaries, which allows direct compilation with Clang/LLVM (part of the Apple Xcode distribution). Those binaries now use Apple’s system-level C++ library, instead of GNU’s <code class="code docutils literal notranslate"><span class="pre">libstdc++</span></code>.</p>
<p>This change means that we require Mac OS X 10.7 (“Lion”) or greater—we do not support 10.6 at this time.</p>
<p>Compilation is faster and simpler, and we can reduce the size and complexity of Mac OS X builds and installer packages. By using Apple’s C++ library, we also reduce the likelihood of missing library errors. When this process is completed for the remaining binaries, it will no longer be necessary to install GCC 4.7+ (by way of MacPorts or other package managers) in order to build BEDOPS on OS X, nor will we have to bundle <code class="code docutils literal notranslate"><span class="pre">libstdc++</span></code> with the installer.</p>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-2-0b">
<h3>3.2.38. v2.2.0b<a class="headerlink" href="#v2-2-0b" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Fixed bug with OS X installer’s post-installation scripts.</li>
</ul>
</div>
<div class="section" id="v2-2-0">
<h3>3.2.39. v2.2.0<a class="headerlink" href="#v2-2-0" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>May 22, 2013</strong></p>
<ul class="simple">
<li>Updated packages<ul>
<li>Precompiled packages are now available for Linux (32- and 64-bit) and Mac OS X 10.6-10.8 (32- and 64-bit) hosts.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><span class="std std-ref">Starch v2 test suite</span></a><ul>
<li>We have added a test suite for the Starch archive toolkit with the source download. Test inputs include randomized BED data generated from chromosome and bounds data stored on UCSC servers as well as static FIMO search results. Tests put <code class="code docutils literal notranslate"><span class="pre">starch</span></code>, <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code> and <code class="code docutils literal notranslate"><span class="pre">starchcat</span></code> through various usage scenarios. Please refer to the Starch-specific Makefiles and the test target and subfolder’s <cite>README</cite> doc for more information.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><span class="std std-ref">starchcat</span></a><ul>
<li>Resolves bug with <code class="code docutils literal notranslate"><span class="pre">--gzip</span></code> option, allowing updates of <code class="code docutils literal notranslate"><span class="pre">gzip</span></code> -backed v1.2 and v1.5 archives to the <a class="reference internal" href="reference/file-management/compression/starch-specification.html#starch-specification"><span class="std std-ref">v2 Starch format</span></a> (either <code class="code docutils literal notranslate"><span class="pre">bzip2</span></code> - or <code class="code docutils literal notranslate"><span class="pre">gzip</span></code> -backed).</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><span class="std std-ref">unstarch</span></a><ul>
<li>Resolves bug with extraction of <a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">Starch</span></a> archive made from BED files with four or more columns. A condition where the total length of additional columns exceeds a certain number of characters would result in extracted data in those columns being cut off. As an example, this could affect Starch archives made from the raw, uncut output of GTF- and GFF- <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a>.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a><ul>
<li>We have partially reverted <code class="code docutils literal notranslate"><span class="pre">wig2bed</span></code>, providing a Bash shell wrapper to the original C binary. This preserves consistency of command-line options across the conversion suite, while making use of the C binary to recover performance lost from the Python-based v2.1 revision of <code class="code docutils literal notranslate"><span class="pre">wig2bed</span></code> (which at this time is no longer supported). (Thanks to Matt Maurano for reporting this issue.)</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-1-1">
<h3>3.2.40. v2.1.1<a class="headerlink" href="#v2-1-1" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>May 3, 2013</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a><ul>
<li>Major performance improvements made in v2.1.1, such that current <code class="code docutils literal notranslate"><span class="pre">bedmap</span></code> now operates as fast or faster than the v1.2.5 version of <code class="code docutils literal notranslate"><span class="pre">bedmap</span></code>!</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a><ul>
<li>Resolves bug with <code class="code docutils literal notranslate"><span class="pre">--partition</span></code> option.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a><ul>
<li>All v2.1.0 Python-based scripts now include fix for <code class="code docutils literal notranslate"><span class="pre">SIGPIPE</span></code> handling, such that use of <code class="code docutils literal notranslate"><span class="pre">head</span></code> or other common UNIX utilities to process buffered standard output no longer yields <code class="code docutils literal notranslate"><span class="pre">IOError</span></code> exceptions. (Thanks to Matt Maurano for reporting this bug.)</li>
</ul>
</li>
<li>32-bit Linux binary support<ul>
<li>Pre-built Linux binaries are now available for end users with 32-bit workstations.</li>
</ul>
</li>
</ul>
<p>Other issues fixed:</p>
<ul class="simple">
<li>Jansson tarball no longer includes already-compiled libraries that could potentially interfere with 32-bit builds.</li>
<li>Minor changes to conversion script test suite to exit with useful error code on successful completion of test.</li>
</ul>
</div>
<div class="section" id="v2-1-0">
<h3>3.2.41. v2.1.0<a class="headerlink" href="#v2-1-0" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>April 22, 2013</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a><ul>
<li>New <code class="code docutils literal notranslate"><span class="pre">--partition</span></code> operator efficiently generates disjoint segments made from genomic boundaries of all overlapping inputs.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a><ul>
<li>All scripts now use <code class="code docutils literal notranslate"><span class="pre">sort-bed</span></code> behind the scenes to output sorted BED output, ready for use with BEDOPS utilities. It is no longer necessary to pipe data to or otherwise post-process converted data with <code class="code docutils literal notranslate"><span class="pre">sort-bed</span></code>.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">psl2bed</span></code> conversion script, converting <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format2">PSL-formatted UCSC BLAT output</a> to BED.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">wig2bed</span></code> conversion script written in Python.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">*2starch</span></code> <a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a> offered for all <code class="code docutils literal notranslate"><span class="pre">*2bed</span></code> scripts, which output Starch v2 archives.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/closest-features.html#closest-features"><span class="std std-ref">closest-features</span></a><ul>
<li>Replaced <code class="code docutils literal notranslate"><span class="pre">--shortest</span></code> option name with <code class="code docutils literal notranslate"><span class="pre">--closest</span></code>, for clarity. (Old scripts which use <code class="code docutils literal notranslate"><span class="pre">--shortest</span></code> will continue to work with the deprecated option name for now. We advise editing pipelines, as needed.)</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a><ul>
<li>Improved error checking for interleaved records. This also makes use of <code class="code docutils literal notranslate"><span class="pre">*2starch</span></code> conversion scripts with the <code class="code docutils literal notranslate"><span class="pre">--do-not-sort</span></code> option safer.</li>
</ul>
</li>
<li>Improved Mac OS X support<ul>
<li>New Mac OS X package installer makes installation of BEDOPS binaries and scripts very easy for OS X 10.6 - 10.8 hosts.</li>
<li>Installer resolves fatal library errors seen by some end users of older OS X BEDOPS releases.</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="v2-0-0b">
<h3>3.2.42. v2.0.0b<a class="headerlink" href="#v2-0-0b" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 19, 2013</strong></p>
<ul class="simple">
<li>Added <code class="code docutils literal notranslate"><span class="pre">starchcluster</span></code> script variant which supports task distribution with <a class="reference external" href="http://www.gnu.org/software/parallel/">GNU Parallel</a>.</li>
<li>Fixed minor problem with <code class="code docutils literal notranslate"><span class="pre">bam2bed</span></code> and <code class="code docutils literal notranslate"><span class="pre">sam2bed</span></code> conversion scripts.</li>
</ul>
</div>
<div class="section" id="v2-0-0a">
<h3>3.2.43. v2.0.0a<a class="headerlink" href="#v2-0-0a" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>February 7, 2013</strong></p>
<ul class="simple">
<li><a class="reference internal" href="reference/statistics/bedmap.html#bedmap"><span class="std std-ref">bedmap</span></a><ul>
<li>Takes in Starch-formatted archives as input, as well as raw BED (i.e., it is no longer required to extract a Starch archive to an intermediate, temporary file or named pipe before applying operations).</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">--chrom</span></code> operator jumps to and operates on information for specified chromosome only.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">--echo-map-id-uniq</span></code> operator lists unique IDs from overlapping mapping elements.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">--max-element</span></code> and <code class="code docutils literal notranslate"><span class="pre">--min-element</span></code> operators return the highest or lowest scoring overlapping map element.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/bedops.html#bedops"><span class="std std-ref">bedops</span></a><ul>
<li>Takes in Starch-formatted archives as input, as well as raw BED.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">--chrom</span></code> operator jumps to and operates on information for specified chromosome only.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/set-operations/closest-features.html#closest-features"><span class="std std-ref">closest-features</span></a><ul>
<li>Takes in Starch-formatted archives as input, as well as raw BED.</li>
<li>New <code class="code docutils literal notranslate"><span class="pre">--chrom</span></code> operator jumps to and operates on information for specified chromosome only.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/sorting/sort-bed.html#sort-bed"><span class="std std-ref">sort-bed</span></a> and <code class="docutils literal notranslate"><span class="pre">bbms</span></code><ul>
<li>New <code class="code docutils literal notranslate"><span class="pre">--max-mem</span></code> option to limit system memory on large BED inputs.</li>
<li>Incorporated <code class="code docutils literal notranslate"><span class="pre">bbms</span></code> functionality into <code class="code docutils literal notranslate"><span class="pre">sort-bed</span></code> with use of <code class="code docutils literal notranslate"><span class="pre">--max-mem</span></code> operator.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/compression/starch.html#starch"><span class="std std-ref">starch</span></a>, <a class="reference internal" href="reference/file-management/compression/starchcat.html#starchcat"><span class="std std-ref">starchcat</span></a> and <a class="reference internal" href="reference/file-management/compression/unstarch.html#unstarch"><span class="std std-ref">unstarch</span></a><ul>
<li>New metadata enhancements to Starch-format archival and extraction, including: <code class="code docutils literal notranslate"><span class="pre">--note</span></code>, <code class="code docutils literal notranslate"><span class="pre">--elements</span></code>, <code class="code docutils literal notranslate"><span class="pre">--bases</span></code>, <code class="code docutils literal notranslate"><span class="pre">--bases-uniq</span></code>, <code class="code docutils literal notranslate"><span class="pre">--list-chromosomes</span></code>, <code class="code docutils literal notranslate"><span class="pre">--archive-timestamp</span></code>, <code class="code docutils literal notranslate"><span class="pre">--archive-type</span></code> and <code class="code docutils literal notranslate"><span class="pre">--archive-version</span></code> (see <code class="code docutils literal notranslate"><span class="pre">--help</span></code> to <code class="code docutils literal notranslate"><span class="pre">starch</span></code>, <code class="code docutils literal notranslate"><span class="pre">starchcat</span></code> and <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code> binaries, or view the documentation for these applications for more detail).</li>
<li>Adds 20-35% performance boost to creating Starch archives with <code class="code docutils literal notranslate"><span class="pre">starch</span></code> utility.</li>
<li>New documentation with technical overview of the Starch format specification.</li>
</ul>
</li>
<li><a class="reference internal" href="reference/file-management/conversion.html#conversion-scripts"><span class="std std-ref">conversion scripts</span></a><ul>
<li>New <code class="code docutils literal notranslate"><span class="pre">gtf2bed</span></code> conversion script, converting GTF (v2.2) to BED.</li>
</ul>
</li>
<li>Scripts are now part of main download; it is no longer necessary to download the BEDOPS companion separately.</li>
</ul>
</div>
<div class="section" id="v1-2-5b">
<h3>3.2.44. v1.2.5b<a class="headerlink" href="#v1-2-5b" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>January 14, 2013</strong></p>
<ul class="simple">
<li>Adds support for Apple 32- and 64-bit Intel hardware running OS X 10.5 through 10.8.</li>
<li>Adds <code class="code docutils literal notranslate"><span class="pre">README</span></code> for companion download.</li>
<li>Removes some obsolete code.</li>
</ul>
</div>
<div class="section" id="v1-2-5">
<h3>3.2.45. v1.2.5<a class="headerlink" href="#v1-2-5" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>October 13, 2012</strong></p>
<ul class="simple">
<li>Fixed unusual bug with <code class="code docutils literal notranslate"><span class="pre">unstarch</span></code>, where an extra (and incorrect) line of BED data can potentially be extracted from an archive.</li>
<li>Updated companion download with updated <code class="code docutils literal notranslate"><span class="pre">bam2bed</span></code> and <code class="code docutils literal notranslate"><span class="pre">sam2bed</span></code> conversion scripts to address 0-indexing error with previous revisions.</li>
</ul>
</div>
<div class="section" id="v1-2-3">
<h3>3.2.46. v1.2.3<a class="headerlink" href="#v1-2-3" title="Permalink to this headline">¶</a></h3>
<p>Released: <strong>August 17, 2012</strong></p>
<ul class="simple">
<li>Added <code class="code docutils literal notranslate"><span class="pre">--indicator</span></code> option to <code class="code docutils literal notranslate"><span class="pre">bedmap</span></code>.</li>
<li>Assorted changes to conversion scripts and associated companion download.</li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/logo_with_label_v3.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">3. Revision history</a><ul>
<li><a class="reference internal" href="#current-version">3.1. Current version</a><ul>
<li><a class="reference internal" href="#v2-4-36">3.1.1. v2.4.36</a></li>
</ul>
</li>
<li><a class="reference internal" href="#previous-versions">3.2. Previous versions</a><ul>
<li><a class="reference internal" href="#v2-4-35">3.2.1. v2.4.35</a></li>
<li><a class="reference internal" href="#v2-4-34">3.2.2. v2.4.34</a></li>
<li><a class="reference internal" href="#v2-4-33">3.2.3. v2.4.33</a></li>
<li><a class="reference internal" href="#v2-4-32">3.2.4. v2.4.32</a></li>
<li><a class="reference internal" href="#v2-4-31">3.2.5. v2.4.31</a></li>
<li><a class="reference internal" href="#v2-4-30">3.2.6. v2.4.30</a></li>
<li><a class="reference internal" href="#v2-4-29">3.2.7. v2.4.29</a></li>
<li><a class="reference internal" href="#v2-4-28">3.2.8. v2.4.28</a></li>
<li><a class="reference internal" href="#v2-4-27">3.2.9. v2.4.27</a></li>
<li><a class="reference internal" href="#v2-4-26">3.2.10. v2.4.26</a></li>
<li><a class="reference internal" href="#v2-4-25">3.2.11. v2.4.25</a></li>
<li><a class="reference internal" href="#v2-4-24">3.2.12. v2.4.24</a></li>
<li><a class="reference internal" href="#v2-4-23">3.2.13. v2.4.23</a></li>
<li><a class="reference internal" href="#v2-4-22">3.2.14. v2.4.22</a></li>
<li><a class="reference internal" href="#v2-4-21">3.2.15. v2.4.21</a></li>
<li><a class="reference internal" href="#v2-4-20">3.2.16. v2.4.20</a></li>
<li><a class="reference internal" href="#v2-4-19">3.2.17. v2.4.19</a></li>
<li><a class="reference internal" href="#v2-4-18">3.2.18. v2.4.18</a></li>
<li><a class="reference internal" href="#v2-4-17">3.2.19. v2.4.17</a></li>
<li><a class="reference internal" href="#v2-4-16">3.2.20. v2.4.16</a></li>
<li><a class="reference internal" href="#v2-4-15">3.2.21. v2.4.15</a></li>
<li><a class="reference internal" href="#v2-4-14">3.2.22. v2.4.14</a></li>
<li><a class="reference internal" href="#v2-4-13">3.2.23. v2.4.13</a></li>
<li><a class="reference internal" href="#v2-4-12">3.2.24. v2.4.12</a></li>
<li><a class="reference internal" href="#v2-4-11">3.2.25. v2.4.11</a></li>
<li><a class="reference internal" href="#v2-4-10">3.2.26. v2.4.10</a></li>
<li><a class="reference internal" href="#v2-4-9">3.2.27. v2.4.9</a></li>
<li><a class="reference internal" href="#v2-4-8">3.2.28. v2.4.8</a></li>
<li><a class="reference internal" href="#v2-4-7">3.2.29. v2.4.7</a></li>
<li><a class="reference internal" href="#v2-4-6">3.2.30. v2.4.6</a></li>
<li><a class="reference internal" href="#v2-4-5">3.2.31. v2.4.5</a></li>
<li><a class="reference internal" href="#v2-4-4">3.2.32. v2.4.4</a></li>
<li><a class="reference internal" href="#v2-4-3">3.2.33. v2.4.3</a></li>
<li><a class="reference internal" href="#v2-4-2">3.2.34. v2.4.2</a></li>
<li><a class="reference internal" href="#v2-4-1">3.2.35. v2.4.1</a></li>
<li><a class="reference internal" href="#v2-4-0">3.2.36. v2.4.0</a></li>
<li><a class="reference internal" href="#v2-3-0">3.2.37. v2.3.0</a></li>
<li><a class="reference internal" href="#v2-2-0b">3.2.38. v2.2.0b</a></li>
<li><a class="reference internal" href="#v2-2-0">3.2.39. v2.2.0</a></li>
<li><a class="reference internal" href="#v2-1-1">3.2.40. v2.1.1</a></li>
<li><a class="reference internal" href="#v2-1-0">3.2.41. v2.1.0</a></li>
<li><a class="reference internal" href="#v2-0-0b">3.2.42. v2.0.0b</a></li>
<li><a class="reference internal" href="#v2-0-0a">3.2.43. v2.0.0a</a></li>
<li><a class="reference internal" href="#v1-2-5b">3.2.44. v1.2.5b</a></li>
<li><a class="reference internal" href="#v1-2-5">3.2.45. v1.2.5</a></li>
<li><a class="reference internal" href="#v1-2-3">3.2.46. v1.2.3</a></li>
</ul>
</li>
</ul>
</li>
</ul>

<form class="search" action="../search.html" method="get">
  <input type="text" name="q"
   placeholder="type to search" />
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
  <div class="related bottom">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="installation.html" title="Previous document">2. Installation</a>
        </li>
        <li>
          <a href="usage-examples.html" title="Next document">4. Usage examples</a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../index.html">Home</a></li> 
    </ul>
  </nav>
  </div>
  <footer id="pagefooter">&copy; 2011-2019, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a>
      1.8.5
        with the <a href="http://github.com/irskep/sphinx-better-theme">
          better</a> theme.

  </footer>

  
  </body>
</html>