<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>6.1.2. bedextract &mdash; BEDOPS 2.3 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '2.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../_static/sidebar.js"></script>
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
    <link rel="top" title="BEDOPS 2.3 documentation" href="../../../index.html" />
    <link rel="up" title="6.1. Set operations" href="../set-operations.html" />
    <link rel="next" title="6.1.3. closest-features" href="closest-features.html" />
    <link rel="prev" title="6.1.1. bedops" href="bedops.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="closest-features.html" title="6.1.3. closest-features"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="bedops.html" title="6.1.1. bedops"
             accesskey="P">previous</a> |</li>
        <li><a href="../../../index.html">BEDOPS 2.3 documentation</a> &raquo;</li>
          <li><a href="../../reference.html" >6. Reference</a> &raquo;</li>
          <li><a href="../set-operations.html" accesskey="U">6.1. Set operations</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="bedextract">
<span id="id1"></span><h1>6.1.2. <cite>bedextract</cite><a class="headerlink" href="#bedextract" title="Permalink to this headline">¶</a></h1>
<p>The <tt class="docutils literal"><span class="pre">bedextract</span></tt> utility performs three primary tasks, with the goal of doing them very quickly:</p>
<ol class="arabic simple">
<li>Lists all the chromosomes in a sorted input BED file.</li>
<li>Extracts all the elements in a sorted input BED file, for a given chromosome.</li>
<li>Finds elements of one BED file, which overlap elements in a second, reference BED file (when specific element criteria are satisfied).</li>
</ol>
<p>One might ask why use this utility, when the first two tasks can already be performed with common UNIX text processing tools, such as <tt class="docutils literal"><span class="pre">cut</span></tt>, <tt class="docutils literal"><span class="pre">sort</span></tt>, <tt class="docutils literal"><span class="pre">uniq</span></tt>, and <tt class="docutils literal"><span class="pre">awk</span></tt>, and the third task can be performed with <a class="reference internal" href="bedops.html#bedops"><em>bedops</em></a> with the <tt class="docutils literal"><span class="pre">--element-of</span> <span class="pre">-1</span></tt> options?</p>
<p>The <tt class="docutils literal"><span class="pre">bedextract</span></tt> utility does the work of all those tools without streaming through an entire BED file, resulting in massive performance improvements. By using the hints provided by sorted BED input, the <a class="reference internal" href="#bedextract"><em>bedextract</em></a> tool can jump around, seeking very quick answers to these questions about your data.</p>
<div class="section" id="how-it-works">
<h2>6.1.2.1. How it works<a class="headerlink" href="#how-it-works" title="Permalink to this headline">¶</a></h2>
<p>Specifically, sorting with <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a> allows us to perform a <a class="reference external" href="http://en.wikipedia.org/wiki/Binary_search_algorithm">binary search</a>:</p>
<ol class="arabic simple">
<li>We jump to the middle byte of the BED file, stream to the nearest element, then parse and test the chromosome name.</li>
<li>Either we have a match, or we jump to the middle of the remaining left or right half (decided by dictionary order), parse and test again.</li>
<li>We repeat steps 1 and 2 until we have matches that define the bounds of the target chromosome.</li>
</ol>
<a class="reference internal image-reference" href="../../../_images/reference_bedextract_mechanism.png"><img alt="../../../_images/reference_bedextract_mechanism.png" src="../../../_images/reference_bedextract_mechanism.png" style="width: 99%;" /></a>
<p>To indicate the kind of speed gain that the <a class="reference internal" href="#bedextract"><em>bedextract</em></a> tool provides, in local testing, a naïve listing of chromosomes from a 36 GB BED input using UNIX <tt class="docutils literal"><span class="pre">cut</span></tt> and <tt class="docutils literal"><span class="pre">uniq</span></tt> utilities took approximately 20 minutes to complete on a typical Core 2 Duo-based Linux workstation. Retrieval of the same chromosome listing with <tt class="docutils literal"><span class="pre">bedextract</span> <span class="pre">--list-chr</span></tt> took only 2 seconds (cache flushed—no cheating!).</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">While listing chromosomes is perhaps a trivial task, 1200 seconds to 2 seconds is a 600-fold speedup. Similar improvements are gained from using <tt class="docutils literal"><span class="pre">--chrom</span></tt> and <tt class="docutils literal"><span class="pre">--faster</span></tt> options with other core BEDOPS tools like <a class="reference internal" href="bedops.html#bedops"><em>bedops</em></a> and <a class="reference internal" href="../statistics/bedmap.html#bedmap"><em>bedmap</em></a>. If your data meet the criteria for using this approach—and a lot of genomic datasets do—we strongly encourage adding this to your toolkit.</p>
</div>
</div>
<div class="section" id="inputs-and-outputs">
<h2>6.1.2.2. Inputs and outputs<a class="headerlink" href="#inputs-and-outputs" title="Permalink to this headline">¶</a></h2>
<div class="section" id="input">
<h3>6.1.2.2.1. Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>Depending on specified options, <a class="reference internal" href="#bedextract"><em>bedextract</em></a> requires one or two <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sorted</em></a> BED files.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">It is critical that inputs are <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sorted</em></a> as the information in a sorted file allows <a class="reference internal" href="#bedextract"><em>bedextract</em></a> to do its work correctly. If your datasets are output from other BEDOPS tools, then they are already sorted!</p>
</div>
</div>
<div class="section" id="output">
<h3>6.1.2.2.2. Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<p>Depending on specified options, the <a class="reference internal" href="#bedextract"><em>bedextract</em></a> program will send a list of chromosomes or BED elements to standard output.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">The use of UNIX-like standard streams allows easy downstream analysis or post-processing with other tools and scripts, including other BEDOPS utilities.</p>
</div>
</div>
</div>
<div class="section" id="usage">
<h2>6.1.2.3. Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">--help</span></tt> option describes the functionality available to the end user:</p>
<div class="highlight-python"><pre>bedextract
  citation: http://bioinformatics.oxfordjournals.org/content/28/14/1919.abstract
  version:  2.4.0
  authors:  Shane Neph &amp; Alex Reynolds

    Every input file must be sorted per sort-bed.

 USAGE:
   0) --help or --version           Print requested info and exit successfully.
   1) --list-chr &lt;input.bed&gt;        Print all unique chromosome names found in &lt;input.bed&gt;.
   2) &lt;chromosome&gt; &lt;input.bed&gt;      Retrieve all rows for chr8 with:  bedextract chr8 &lt;input.bed&gt;.
   3) &lt;query.bed&gt; &lt;target&gt;          Grab elements from the &lt;query.bed&gt; that overlap elements in &lt;target&gt;. Same as
                                     `bedops -e -1 &lt;query.bed&gt; &lt;target&gt;`, except that this option fails silently
                                      if &lt;query.bed&gt; contains fully-nested BED elements.  If no fully-nested
                                      element exists, bedextract can vastly improve upon the performance of bedops.
                                      &lt;target&gt; may be a BED or Starch file (with or without fully-nested elements).
                                      Using '-' for &lt;target&gt; indicates input (in BED format) comes from stdin.</pre>
</div>
<div class="section" id="listing-chromosomes">
<h3>6.1.2.3.1. Listing chromosomes<a class="headerlink" href="#listing-chromosomes" title="Permalink to this headline">¶</a></h3>
<p>Use the <tt class="docutils literal"><span class="pre">--list-chr</span></tt> option to quickly retrieve a listing of chromosomes from a given sorted BED input.</p>
<p>For example, the following lists the chromosomes in an example BED file of FIMO motif hits (see the <a class="reference internal" href="#bedextract-downloads"><em>Downloads</em></a> section):</p>
<div class="highlight-python"><pre>$ bedextract --list-chr motifs.bed
chr1
chr10
chr11
chr12
...
chr9
chrX</pre>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">bedextract</span> <span class="pre">--list-chr</span></tt> operation only works on BED files. If you have a Starch file, use <tt class="docutils literal"><span class="pre">unstarch</span> <span class="pre">--list-chr</span></tt> to list its chromosomes.</p>
</div>
</div>
<div class="section" id="retrieving-elements-from-a-specific-chromosome">
<h3>6.1.2.3.2. Retrieving elements from a specific chromosome<a class="headerlink" href="#retrieving-elements-from-a-specific-chromosome" title="Permalink to this headline">¶</a></h3>
<p>To quickly retrieve the subset of elements from a sorted BED file associated with a given chromosome, apply the second usage case and specify the chromosome as the argument.</p>
<p>For example, to retrieve <tt class="docutils literal"><span class="pre">chrX</span></tt> from the same motif sample:</p>
<div class="highlight-python"><pre>$ bedextract chrX motifs.bed
chrX    6775077 6775092 +V_SPZ1_01      4.92705e-06     +       GTTGGAGGGAAGGGC
chrX    6775168 6775179 +V_ELF5_01      8.57585e-06     +       TCAAGGAAGTA
chrX    6777790 6777799 +V_CKROX_Q2     8.90515e-06     +       TCCCTCCCC
...</pre>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">bedextract</span> <span class="pre">&lt;chromosome&gt;</span></tt> operation only works on BED files. If you have a Starch file, use <tt class="docutils literal"><span class="pre">unstarch</span> <span class="pre">&lt;chromosome&gt;</span></tt> to list the elements associated with that chromosome.</p>
</div>
</div>
<div class="section" id="retrieving-elements-which-overlap-target-elements">
<h3>6.1.2.3.3. Retrieving elements which overlap target elements<a class="headerlink" href="#retrieving-elements-which-overlap-target-elements" title="Permalink to this headline">¶</a></h3>
<p>A common <a class="reference internal" href="bedops.html#bedops"><em>bedops</em></a> query involves asking which elements overlap one or more bases between two BED datasets, which we will call here <tt class="docutils literal"><span class="pre">Query</span></tt> and <tt class="docutils literal"><span class="pre">Target</span></tt>.</p>
<p>One can already use <tt class="docutils literal"><span class="pre">bedops</span> <span class="pre">--element-of</span> <span class="pre">-1</span></tt> to accomplish this task, but if certain specific criteria are met (which we will describe shortly) then a much faster result can often be obtained by instead using <a class="reference internal" href="#bedextract"><em>bedextract</em></a>.</p>
<p>Three criteria make the use of <a class="reference internal" href="#bedextract"><em>bedextract</em></a> in this mode very successful in practice, with potentially massive speed improvements:</p>
<ol class="arabic simple">
<li><tt class="docutils literal"><span class="pre">Query</span></tt> is a huge file.</li>
<li>There are relatively few regions of interest in <tt class="docutils literal"><span class="pre">Target</span></tt> (say, roughly 30,000 or fewer).</li>
<li>There are <strong>no fully-nested elements</strong> in <tt class="docutils literal"><span class="pre">Query</span></tt> (but duplicate coordinates are fine).</li>
</ol>
<div class="section" id="what-are-nested-elements">
<span id="bedextract-nested-elements"></span><h4>6.1.2.3.3.1. What are nested elements?<a class="headerlink" href="#what-are-nested-elements" title="Permalink to this headline">¶</a></h4>
<p>An example of a sorted BED file which contains a nested element follows:</p>
<div class="highlight-python"><pre>chr1    1      100
chr1    100    200
chr1    125    150
chr1    150    1000</pre>
</div>
<p>While this dataset is sorted, the element <tt class="docutils literal"><span class="pre">chr1:125-150</span></tt> is entirely nested within <tt class="docutils literal"><span class="pre">chr1:100-200</span></tt>:</p>
<a class="reference internal image-reference" href="../../../_images/reference_bedextract_nested_elements.png"><img alt="../../../_images/reference_bedextract_nested_elements.png" src="../../../_images/reference_bedextract_nested_elements.png" style="width: 99%;" /></a>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Fully-nested elements are not a problem for the other two <a class="reference internal" href="#bedextract"><em>bedextract</em></a> features: 1) Listing all chromosomes, and 2) Retrieving all information for a single chromosome.</p>
</div>
<p>Fully-nested elements are only an issue if they exist in the <tt class="docutils literal"><span class="pre">Query</span></tt> dataset. Results are not affected if the <tt class="docutils literal"><span class="pre">Target</span></tt> dataset contains nested elements. Overlapping (but not fully-nested) elements in the <tt class="docutils literal"><span class="pre">Query</span></tt> input file are fine, as are duplicated genomic positions.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Our lab works with BED data of various types: cut-counts, hotspots, peaks, footprints, etc. These data generally do not contain nested elements and so are amenable to use with <a class="reference internal" href="#bedextract"><em>bedextract</em></a> for extracting overlapping elements.</p>
<p class="last">However, other types of <tt class="docutils literal"><span class="pre">Query</span></tt> datasets can be problematic. FIMO search results, for example, might cause trouble, where the boundaries of one motif hit can be contained within another larger hit. Or paired-end sequence data, where tags are not of a fixed length. Be sure to consider the makeup of your BED data before using <a class="reference internal" href="#bedextract"><em>bedextract</em></a>.</p>
</div>
</div>
<div class="section" id="demonstration">
<h4>6.1.2.3.3.2. Demonstration<a class="headerlink" href="#demonstration" title="Permalink to this headline">¶</a></h4>
<p>To demonstrate this use of <a class="reference internal" href="#bedextract"><em>bedextract</em></a>, for our <tt class="docutils literal"><span class="pre">Query</span></tt> dataset we will use the <tt class="docutils literal"><span class="pre">Map</span></tt> example from our <a class="reference internal" href="../statistics/bedmap.html#bedmap"><em>bedmap</em></a> documentation, which contains raw DNaseI hypersensitivity signal from a human K562 cell line (see the <a class="reference internal" href="#bedextract-downloads"><em>Downloads</em></a> section for sample data):</p>
<div class="highlight-python"><pre>$ cat query.bed
chr21   33031165        33031185        map-1   1.000000
chr21   33031185        33031205        map-2   3.000000
chr21   33031205        33031225        map-3   3.000000
chr21   33031225        33031245        map-4   3.000000
...
chr21   33032445        33032465        map-65  5.000000
chr21   33032465        33032485        map-66  6.000000</pre>
</div>
<p>Our <tt class="docutils literal"><span class="pre">Target</span></tt> data is simply an <em>ad-hoc</em> BED region which overlaps part of the <tt class="docutils literal"><span class="pre">Query</span></tt> dataset, stored in a <a class="reference internal" href="../file-management/compression/starch.html#starch"><em>Starch-formatted</em></a> archive:</p>
<div class="highlight-python"><pre>$ unstarch target.starch
chr21   33031600        33031700</pre>
</div>
<p>We can now ask which elements of <tt class="docutils literal"><span class="pre">Query</span></tt> overlap the element in <tt class="docutils literal"><span class="pre">Target</span></tt>:</p>
<div class="highlight-python"><pre>$ bedextract query.bed target.starch
chr21   33031585        33031605        map-22  26.000000
chr21   33031605        33031625        map-23  27.000000
chr21   33031625        33031645        map-24  29.000000
chr21   33031645        33031665        map-25  31.000000
chr21   33031665        33031685        map-26  31.000000
chr21   33031685        33031705        map-27  37.000000</pre>
</div>
<p>Our <tt class="docutils literal"><span class="pre">Target</span></tt> dataset is a Starch-formatted file. Note that we can also use &#8220;-&#8221; to denote standard input for the <tt class="docutils literal"><span class="pre">Target</span></tt> dataset, as well as a regular BED- or Starch-formatted file. In other words, we can pipe target elements from another process to <a class="reference internal" href="#bedextract"><em>bedextract</em></a>, <em>e.g.</em> we can query for an ad-hoc element as follows:</p>
<div class="highlight-python"><pre>$ echo -e "chr21\t33031590\t33031600" | bedextract query.bed -
chr21   33031585        33031605        map-22  26.000000</pre>
</div>
<p>Instead of an <em>ad-hoc</em> element as in this example, however, target elements could just as easily be piped in from upstream <a class="reference internal" href="../statistics/bedmap.html#bedmap"><em>bedmap</em></a> or <a class="reference internal" href="bedops.html#bedops"><em>bedops</em></a> operations, or extracted elements from a Starch archive, etc.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">The output of this particular use of <a class="reference internal" href="#bedextract"><em>bedextract</em></a> is made up of elements from the <tt class="docutils literal"><span class="pre">Query</span></tt> dataset and is therefore <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sorted</em></a> BED data, which can be piped to <a class="reference internal" href="bedops.html#bedops"><em>bedops</em></a>, <a class="reference internal" href="../statistics/bedmap.html#bedmap"><em>bedmap</em></a> and other BEDOPS utilities for further downstream processing.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Though <a class="reference internal" href="#bedextract"><em>bedextract</em></a> only supports the overlap equivalent of <tt class="docutils literal"><span class="pre">bedops</span> <span class="pre">--element-of</span> <span class="pre">-1</span></tt>, other overlap criteria are efficiently supported by combining <a class="reference internal" href="#bedextract"><em>bedextract</em></a> with <a class="reference internal" href="bedops.html#bedops"><em>bedops</em></a>.</p>
<p>Specifically, we can quickly filter through just the results given by <a class="reference internal" href="#bedextract"><em>bedextract</em></a> and implement other overlap criteria with <a class="reference internal" href="bedops.html#bedops"><em>bedops</em></a>, <em>e.g.</em>:</p>
<p class="last"><tt class="docutils literal"><span class="pre">$</span> <span class="pre">bedextract</span> <span class="pre">query.bed</span> <span class="pre">target.bed</span> <span class="pre">|</span> <span class="pre">bedops</span> <span class="pre">-e</span> <span class="pre">-50%</span> <span class="pre">-</span> <span class="pre">target.bed</span></tt></p>
</div>
</div>
</div>
</div>
<div class="section" id="downloads">
<span id="bedextract-downloads"></span><h2>6.1.2.4. Downloads<a class="headerlink" href="#downloads" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Sample <a class="reference download internal" href="../../../_downloads/reference_bedextract_motifs.bed"><tt class="xref download docutils literal"><span class="pre">FIMO</span> <span class="pre">motifs</span></tt></a></li>
<li>Sample <tt class="docutils literal"><span class="pre">Query</span></tt> dataset: <a class="reference download internal" href="../../../_downloads/reference_bedmap_map.bed"><tt class="xref download docutils literal"><span class="pre">DHS</span> <span class="pre">signal</span></tt></a></li>
<li>Sample <tt class="docutils literal"><span class="pre">Target</span></tt> dataset: <a class="reference download internal" href="../../../_downloads/reference_bedextract_target.starch"><tt class="xref download docutils literal"><span class="pre">ad-hoc</span> <span class="pre">coordinates</span></tt></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../../index.html">
              <img class="logo" src="../../../_static/logo_with_label_v2.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">6.1.2. <cite>bedextract</cite></a><ul>
<li><a class="reference internal" href="#how-it-works">6.1.2.1. How it works</a></li>
<li><a class="reference internal" href="#inputs-and-outputs">6.1.2.2. Inputs and outputs</a><ul>
<li><a class="reference internal" href="#input">6.1.2.2.1. Input</a></li>
<li><a class="reference internal" href="#output">6.1.2.2.2. Output</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">6.1.2.3. Usage</a><ul>
<li><a class="reference internal" href="#listing-chromosomes">6.1.2.3.1. Listing chromosomes</a></li>
<li><a class="reference internal" href="#retrieving-elements-from-a-specific-chromosome">6.1.2.3.2. Retrieving elements from a specific chromosome</a></li>
<li><a class="reference internal" href="#retrieving-elements-which-overlap-target-elements">6.1.2.3.3. Retrieving elements which overlap target elements</a><ul>
<li><a class="reference internal" href="#what-are-nested-elements">6.1.2.3.3.1. What are nested elements?</a></li>
<li><a class="reference internal" href="#demonstration">6.1.2.3.3.2. Demonstration</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#downloads">6.1.2.4. Downloads</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="bedops.html"
                        title="previous chapter">6.1.1. <cite>bedops</cite></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="closest-features.html"
                        title="next chapter">6.1.3. <cite>closest-features</cite></a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../../../_sources/content/reference/set-operations/bedextract.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="closest-features.html" title="6.1.3. closest-features"
             >next</a> |</li>
        <li class="right" >
          <a href="bedops.html" title="6.1.1. bedops"
             >previous</a> |</li>
        <li><a href="../../../index.html">BEDOPS 2.3 documentation</a> &raquo;</li>
          <li><a href="../../reference.html" >6. Reference</a> &raquo;</li>
          <li><a href="../set-operations.html" >6.1. Set operations</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.
    </div>
  </body>
</html>