<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>6.1.1. bedops &mdash; BEDOPS 2.4.1 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '2.4.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../_static/sidebar.js"></script>
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
    <link rel="top" title="BEDOPS 2.4.1 documentation" href="../../../index.html" />
    <link rel="up" title="6.1. Set operations" href="../set-operations.html" />
    <link rel="next" title="6.1.2. bedextract" href="bedextract.html" />
    <link rel="prev" title="6.1. Set operations" href="../set-operations.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="bedextract.html" title="6.1.2. bedextract"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../set-operations.html" title="6.1. Set operations"
             accesskey="P">previous</a> |</li>
        <li><a href="../../../index.html">BEDOPS 2.4.1 documentation</a> &raquo;</li>
          <li><a href="../../reference.html" >6. Reference</a> &raquo;</li>
          <li><a href="../set-operations.html" accesskey="U">6.1. Set operations</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="bedops">
<span id="id1"></span><h1>6.1.1. <cite>bedops</cite><a class="headerlink" href="#bedops" title="Permalink to this headline">¶</a></h1>
<p><tt class="docutils literal"><span class="pre">bedops</span></tt> is a core tool for finding relationships between two or more genomic datasets.</p>
<p>This is an important category of problems to solve. As examples, one might want to:</p>
<ul class="simple">
<li>Know how much overlap exists between the elements of two datasets, to quantitatively establish the degree to which they are similar.</li>
<li>Merge or filter elements. For example, retrieving non-overlapping, &#8220;unique&#8221; elements from multiple BED files.</li>
<li>Split elements from multiple BED files into disjoint subsets.</li>
</ul>
<p>The <a class="reference internal" href="#bedops"><em>bedops</em></a> program offers several Boolean set and multiset operations, including union, subset, and difference, to assist investigators with answering these types of questions.</p>
<p>Importantly, <a class="reference internal" href="#bedops"><em>bedops</em></a> handles any number of any-size inputs at once when computing results in order to maximize efficiency. This use case has <a class="reference internal" href="../../usage-examples/multiple-inputs.html#multiple-inputs"><em>serious practical consequences</em></a> for many genomic studies.</p>
<p>One can also use <a class="reference internal" href="#bedops"><em>bedops</em></a> to symmetrically or asymmetrically pad coordinates.</p>
<div class="section" id="inputs-and-outputs">
<h2>6.1.1.1. Inputs and outputs<a class="headerlink" href="#inputs-and-outputs" title="Permalink to this headline">¶</a></h2>
<div class="section" id="input">
<h3>6.1.1.1.1. Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="#bedops"><em>bedops</em></a> program reads <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sorted</em></a> BED data and BEDOPS <a class="reference internal" href="../file-management/compression/starch.html#starch"><em>Starch-formatted</em></a> archives as input.</p>
<p>Finally, <a class="reference internal" href="#bedops"><em>bedops</em></a> requires specification of a set operation (and, optionally, may include modifier options).</p>
<p>Support for common headers (including UCSC track headers) is offered through the <tt class="docutils literal"><span class="pre">--header</span></tt> option. Headers are stripped from output.</p>
</div>
<div class="section" id="output">
<h3>6.1.1.1.2. Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="#bedops"><em>bedops</em></a> program returns <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sorted</em></a> BED results to standard output. This output can be redirected to a file or piped to other utilities.</p>
</div>
</div>
<div class="section" id="usage">
<h2>6.1.1.2. Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="#bedops"><em>bedops</em></a> program takes sorted BED-formatted data as input, either from a file or streamed from standard input. It will process any number of input files in parallel.</p>
<p>If your data are unsorted, use BEDOPS <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a> to prepare data for <a class="reference internal" href="#bedops"><em>bedops</em></a>. You only need to sort once, as all BEDOPS tools read and write sorted BED data.</p>
<p>Because memory usage is very low, one can use sorted inputs of any size. Processing times generally follow a simple linear relationship with input sizes (<em>e.g.</em>, as the input size doubles, the processing time doubles accordingly).</p>
<p>The <tt class="docutils literal"><span class="pre">--help</span></tt> option describes the set operation and other options available to the end user:</p>
<div class="highlight-python"><pre>bedops
  citation: http://bioinformatics.oxfordjournals.org/content/28/14/1919.abstract
  version:  2.4.1
  authors:  Shane Neph &amp; Scott Kuehn

      USAGE: bedops [process-flags] &lt;operation&gt; &lt;File(s)&gt;*

          Every input file must be sorted per the sort-bed utility.
          Each operation requires a minimum number of files as shown below.
            There is no fixed maximum number of files that may be used.
          Input files must have at least the first 3 columns of the BED specification.
          The program accepts BED and starch file formats.
          May use '-' for a file to indicate reading from standard input (BED format only).

      Process Flags:
          --chrom &lt;chromosome&gt; Process data for given &lt;chromosome&gt; only.
          --ec                 Error check input files (slower).
          --header             Accept headers (VCF, GFF, SAM, BED, WIG) in any input file.
          --help               Print this message and exit successfully.
          --help-&lt;operation&gt;   Detailed help on &lt;operation&gt;.
                                 An example is --help-c or --help-complement
          --range L:R          Add 'L' bp to all start coordinates and 'R' bp to end
                                 coordinates. Either value may be + or - to grow or
                                 shrink regions.  With the -e/-n operations, the first
                                 (reference) file is not padded, unlike all other files.
          --range S            Pad input file(s) coordinates symmetrically by S.
                                 This is shorthand for: --range -S:S.
          --version            Print program information.

      Operations: (choose one of)
          -c, --complement [-L] File1 [File]*
          -d, --difference ReferenceFile File2 [File]*
          -e, --element-of [-number% | -number (in bp)] ReferenceFile File2 [File]*
                 by default, -e -100% is used.
          -i, --intersect File1 File2 [File]*
          -m, --merge File1 [File]*
          -n, --not-element-of [-number% | -number (in bp)] ReferenceFile File2 [File]*
                 by default, -n -100% is used.
          -p, --partition File1 [File]*
          -s, --symmdiff File1 File2 [File]*
          -u, --everything File1 [File]*

      Example: bedops --range 10 -u file1.bed
      NOTE: Only operations -e|n|u preserve all columns (no flattening)</pre>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Extended help is available for all operations in <a class="reference internal" href="#bedops"><em>bedops</em></a>. For example, the <tt class="docutils literal"><span class="pre">--help-symmdiff</span></tt> option in <a class="reference internal" href="#bedops"><em>bedops</em></a> gives detailed information on the <tt class="docutils literal"><span class="pre">--symmdiff</span></tt> operation.</p>
</div>
</div>
<div class="section" id="operations">
<h2>6.1.1.3. Operations<a class="headerlink" href="#operations" title="Permalink to this headline">¶</a></h2>
<p>To demonstrate the various operations in <a class="reference internal" href="#bedops"><em>bedops</em></a>, we start with two simple datasets <tt class="docutils literal"><span class="pre">A</span></tt> and <tt class="docutils literal"><span class="pre">B</span></tt>, containing genomic elements on generic chromsome <tt class="docutils literal"><span class="pre">chrN</span></tt>:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_inputs.png"><img alt="../../../_images/reference_setops_bedops_inputs.png" src="../../../_images/reference_setops_bedops_inputs.png" style="width: 99%;" /></a>
<p>These datasets can be <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sorted</em></a> BED or <a class="reference internal" href="../file-management/compression/starch.html#starch"><em>Starch-formatted</em></a> files or streams.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <a class="reference internal" href="#bedops"><em>bedops</em></a> tool can operate on multiple inputs, but we show the results of operations on just two sets to help demonstrate the basic principles.</p>
</div>
<div class="section" id="everything-u-everything">
<span id="bedops-everything"></span><h3>6.1.1.3.1. Everything (-u, &#8211;everything)<a class="headerlink" href="#everything-u-everything" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--everything</span></tt> option is equivalent to concatenating and sorting BED elements from multiple files, but works much faster:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_everything.png"><img alt="../../../_images/reference_setops_bedops_everything.png" src="../../../_images/reference_setops_bedops_everything.png" style="width: 99%;" /></a>
<p>As with all BEDOPS tools and operations, the output of this operation is <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sorted</em></a>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">--everything</span></tt> option preserves all columns from all inputs. This is useful for multiset unions of datasets with additional ID, score or other metadata.</p>
</div>
</div>
<div class="section" id="element-of-e-element-of">
<h3>6.1.1.3.2. Element-of (-e, &#8211;element-of)<a class="headerlink" href="#element-of-e-element-of" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--element-of</span></tt> operation shows the elements of the first (&#8220;<em>reference</em>&#8221;) file that overlap elements in the second and subsequent files by the specified length (in bases) or by percentage of length.</p>
<p>In the following example, we search for elements in the reference set <tt class="docutils literal"><span class="pre">A</span></tt> which overlap elements in <tt class="docutils literal"><span class="pre">B</span></tt> by at least one base:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_elementof_ab.png"><img alt="../../../_images/reference_setops_bedops_elementof_ab.png" src="../../../_images/reference_setops_bedops_elementof_ab.png" style="width: 99%;" /></a>
<p>Elements that are returned are always from the reference set (in this case, set <tt class="docutils literal"><span class="pre">A</span></tt>).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">--element-of</span></tt> option preserves all columns from the first (reference) input.</p>
</div>
<p>Note that <cite>&#8211;element-of</cite> is <em>not</em> a symmetric operation, as demonstrated by making <tt class="docutils literal"><span class="pre">B</span></tt> the reference set:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_elementof_ba.png"><img alt="../../../_images/reference_setops_bedops_elementof_ba.png" src="../../../_images/reference_setops_bedops_elementof_ba.png" style="width: 99%;" /></a>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The argument to <tt class="docutils literal"><span class="pre">--element-of</span></tt> is a value that species to degree of overlap for elements. The value is either integral for per-base overlap,  or fractional for overlap measured by length.</p>
</div>
<p>In sum, <tt class="docutils literal"><span class="pre">--element-of</span></tt> (<tt class="docutils literal"><span class="pre">-e</span></tt>) produces exactly everything that <tt class="docutils literal"><span class="pre">--not-element-of</span></tt> (<tt class="docutils literal"><span class="pre">-n</span></tt>) does not, given the same overlap criterion (which is 100% by default).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For a more in-depth discussion of <tt class="docutils literal"><span class="pre">--element-of</span></tt> and how overlaps are determined with three or more input files, please review the <a class="reference external" href="http://bedops.uwencode.org/forum/index.php?topic=20.0">BEDOPS forum discussion</a> on this subject.</p>
</div>
</div>
<div class="section" id="not-element-of-n-not-element-of">
<h3>6.1.1.3.3. Not-element-of (-n, &#8211;not-element-of)<a class="headerlink" href="#not-element-of-n-not-element-of" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--not-element-of</span></tt> operation shows elements in the reference file which do not overlap elements in all other sets. For example:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_notelementof_ab.png"><img alt="../../../_images/reference_setops_bedops_notelementof_ab.png" src="../../../_images/reference_setops_bedops_notelementof_ab.png" style="width: 99%;" /></a>
<p>As with the <tt class="docutils literal"><span class="pre">--element-of</span></tt> (<tt class="docutils literal"><span class="pre">-e</span></tt>) operator, the overlap criterion for <tt class="docutils literal"><span class="pre">--not-element-of</span></tt> (<tt class="docutils literal"><span class="pre">-n</span></tt>) can be specified by length (in bases) or by percentage of length. Similarly, this operation is not symmetric—the order of inputs will specify the reference set.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">--not-element-of</span></tt> operatior preserves columns from the first (reference) dataset.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The same caveat applies to use of <tt class="docutils literal"><span class="pre">--not-element-of</span></tt> (<tt class="docutils literal"><span class="pre">-n</span></tt>) as with <tt class="docutils literal"><span class="pre">--element-of</span></tt> (<tt class="docutils literal"><span class="pre">-e</span></tt>), namely that the second and all subsequent input files are merged before the set operation is applied. Please review the BEDOPS <a class="reference external" href="http://bedops.uwencode.org/forum/index.php?topic=20.0">forum discussion thread</a> on this topic for more details.</p>
</div>
</div>
<div class="section" id="complement-c-complement">
<h3>6.1.1.3.4. Complement (-c, &#8211;complement)<a class="headerlink" href="#complement-c-complement" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--complement</span></tt> operation calculates the genomic regions in the gaps between the contiguous per-chromosome ranges defined by one or more inputs. The following example shows the use of two inputs:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_complement_ab.png"><img alt="../../../_images/reference_setops_bedops_complement_ab.png" src="../../../_images/reference_setops_bedops_complement_ab.png" style="width: 99%;" /></a>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Note this <strong>computed result</strong> will lack ID, score and other columnar data other than the first three columns that contain positional data. That is, computed elements will not come from any of the input sets, but are new elements created from the input set space.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For a more in-depth discussion on using <tt class="docutils literal"><span class="pre">--complement</span></tt> with left and right bounds of input chromosomes, please review the BEDOPS <a class="reference external" href="http://bedops.uwencode.org/forum/index.php?topic=19.0">forum discussion</a> on this subject.</p>
</div>
</div>
<div class="section" id="difference-d-difference">
<h3>6.1.1.3.5. Difference (-d, &#8211;difference)<a class="headerlink" href="#difference-d-difference" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--difference</span></tt> operation calculates the genomic regions found within the first (reference) input file, excluding regions in all other input files:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_difference_ab.png"><img alt="../../../_images/reference_setops_bedops_difference_ab.png" src="../../../_images/reference_setops_bedops_difference_ab.png" style="width: 99%;" /></a>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">As with <tt class="docutils literal"><span class="pre">--element-of</span></tt> and <tt class="docutils literal"><span class="pre">--not-element-of</span></tt>, this operation is not symmetric. While <tt class="docutils literal"><span class="pre">--not-element-of</span></tt> preserves all columns of elements found in the reference input and allows one to define overlaps, the <tt class="docutils literal"><span class="pre">--difference</span></tt> operator simply reports every genomic range as three-column BED, which does not overlap elements found in the second and subsequent input files by any amount.</p>
</div>
</div>
<div class="section" id="symmetric-difference-s-symmdiff">
<h3>6.1.1.3.6. Symmetric difference (-s, &#8211;symmdiff)<a class="headerlink" href="#symmetric-difference-s-symmdiff" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--symmdiff</span></tt> operation calculates the genomic range that is exclusive to each input, excluding any ranges shared across inputs:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_symmetricdifference_ab.png"><img alt="../../../_images/reference_setops_bedops_symmetricdifference_ab.png" src="../../../_images/reference_setops_bedops_symmetricdifference_ab.png" style="width: 99%;" /></a>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">It has been observed that <tt class="docutils literal"><span class="pre">--symmdiff</span></tt> (<tt class="docutils literal"><span class="pre">-s</span></tt>) is the same as <tt class="docutils literal"><span class="pre">--difference</span> <span class="pre">A</span> <span class="pre">B</span></tt> unionized with <tt class="docutils literal"><span class="pre">--difference</span> <span class="pre">B</span> <span class="pre">A</span></tt>, but <tt class="docutils literal"><span class="pre">--symmdiff</span></tt> runs faster in practice.</p>
</div>
</div>
<div class="section" id="intersect-i-intersect">
<h3>6.1.1.3.7. Intersect (-i, &#8211;intersect)<a class="headerlink" href="#intersect-i-intersect" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--intersect</span></tt> operation determines genomic regions common to all input sets:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_intersect_ab.png"><img alt="../../../_images/reference_setops_bedops_intersect_ab.png" src="../../../_images/reference_setops_bedops_intersect_ab.png" style="width: 99%;" /></a>
</div>
<div class="section" id="merge-m-merge">
<h3>6.1.1.3.8. Merge (-m, &#8211;merge)<a class="headerlink" href="#merge-m-merge" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--merge</span></tt> operation flattens all overlapping and adjoining elements into contiguous regions:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_merge_ab.png"><img alt="../../../_images/reference_setops_bedops_merge_ab.png" src="../../../_images/reference_setops_bedops_merge_ab.png" style="width: 99%;" /></a>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">The preceding example shows use of <tt class="docutils literal"><span class="pre">--merge</span></tt> (<tt class="docutils literal"><span class="pre">-m</span></tt>) with two inputs, but the merge operation works just as well with one input, collapsing regions which overlap or which are directly adjoining.</p>
</div>
</div>
<div class="section" id="partition-p-partition">
<h3>6.1.1.3.9. Partition (-p, &#8211;partition)<a class="headerlink" href="#partition-p-partition" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--partition</span></tt> operator splits all overlapping input regions into a set of disjoint segments. One or more input files may be provided; this option will segment regions from all inputs:</p>
<a class="reference internal image-reference" href="../../../_images/reference_setops_bedops_partition_ab.png"><img alt="../../../_images/reference_setops_bedops_partition_ab.png" src="../../../_images/reference_setops_bedops_partition_ab.png" style="width: 99%;" /></a>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">As with <tt class="docutils literal"><span class="pre">--merge</span></tt>, <tt class="docutils literal"><span class="pre">--complement</span></tt> and other &#8220;computing&#8221; operations, note the lack of ID, score and other columnar data in this computed result.</p>
</div>
</div>
<div class="section" id="per-chromosome-operations-chrom">
<h3>6.1.1.3.10. Per-chromosome operations (&#8211;chrom)<a class="headerlink" href="#per-chromosome-operations-chrom" title="Permalink to this headline">¶</a></h3>
<p>All operations on inputs can be restricted to one chromosome, by adding the <tt class="docutils literal"><span class="pre">--chrom</span> <span class="pre">&lt;val&gt;</span></tt> operator.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This operator is highly useful for cluster-based work, where operations on large BED inputs can be split up by chromosome and pushed to separate cluster nodes. (See the <a class="reference internal" href="../../usage-examples/starchcluster.html#starchcluster"><em>Efficiently creating Starch-formatted archives with a cluster</em></a> documentation for a demonstration of this technique in action.)</p>
</div>
</div>
<div class="section" id="range-range">
<span id="bedops-range"></span><h3>6.1.1.3.11. Range (&#8211;range)<a class="headerlink" href="#range-range" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">--range</span></tt> operation works in conjunction with other operations.</p>
<p>When used with one value (<tt class="docutils literal"><span class="pre">--range</span> <span class="pre">S</span></tt>), this operation <strong>symmetrically</strong> pads all elements of input sets by the specified integral value <tt class="docutils literal"><span class="pre">S</span></tt>. When the specified value is positive, every genomic segment grows in size. An element will grow asymmetrically to prevent growth beyond base position 0, if needed. Otherwise, when negative, elements shrink, and any element with zero (or less) length is discarded.</p>
<p>Alternatively, when used with two values (<tt class="docutils literal"><span class="pre">--range</span> <span class="pre">L:R</span></tt>), this operation <strong>asymmetrically</strong> pads elements, adding <tt class="docutils literal"><span class="pre">L</span></tt> to each start coordinate, and adding <tt class="docutils literal"><span class="pre">R</span></tt> to each stop coordinate. Negative values may be specified to grow or shrink the region, accordingly.</p>
<p id="bed-index-shifting">This option is immediately useful for adjusting the coordinate index of BED files. For example, to shift from 1-based to 0-based coordinate indexing:</p>
<div class="highlight-python"><pre>$ bedops --range -1:-1 --everything my1BasedCoordinates.bed &gt; my0BasedCoordinates.bed</pre>
</div>
<p>And, likewise, for 0-based to 1-based indexing:</p>
<div class="highlight-python"><pre>$ bedops --range 1:1 --everything my0BasedCoordinates.bed &gt; my1BasedCoordinates.bed</pre>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">--range</span></tt> value is applied to inputs prior to the application of other operations (such as <tt class="docutils literal"><span class="pre">--intersect</span></tt> or <tt class="docutils literal"><span class="pre">--merge</span></tt>, etc.).</p>
</div>
<p>Padding elements with <a class="reference internal" href="#bedops"><em>bedops</em></a> is much more efficient that doing so with <tt class="docutils literal"><span class="pre">awk</span></tt> or some other script, <em>and you do not need to go back and resort your data</em>. Even symmetric padding can cause data to become unsorted in non-obvious ways. Using <tt class="docutils literal"><span class="pre">--range</span></tt> ensures that your data remain sorted and it works efficiently with any set operation.</p>
<p>Also, note that the <tt class="docutils literal"><span class="pre">--element-of</span></tt> and <tt class="docutils literal"><span class="pre">--not-element-of</span></tt> operations behave differently with <tt class="docutils literal"><span class="pre">--range</span></tt>, in that only the second and subsequent input files are padded.</p>
</div>
</div>
<div class="section" id="starch-support">
<h2>6.1.1.4. Starch support<a class="headerlink" href="#starch-support" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="#bedops"><em>bedops</em></a> application supports use of <a class="reference internal" href="../file-management/compression/starch.html#starch"><em>Starch</em></a>-formatted archives as inputs, as well as text-based BED data. One or multiple inputs may be Starch archives.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">By combining the <tt class="docutils literal"><span class="pre">--chrom</span></tt> operator with operations on <a class="reference internal" href="../file-management/compression/starch.html#starch"><em>Starch</em></a> archives, the end user can achieve improved computing performance and disk space savings, particularly where <a class="reference internal" href="#bedops"><em>bedops</em></a>, <a class="reference internal" href="../statistics/bedmap.html#bedmap"><em>bedmap</em></a> and <a class="reference internal" href="closest-features.html#closest-features"><em>closest-features</em></a> operations are applied with a computational cluster on separate chromosomes.</p>
</div>
</div>
<div class="section" id="error-checking-ec">
<h2>6.1.1.5. Error checking (&#8211;ec)<a class="headerlink" href="#error-checking-ec" title="Permalink to this headline">¶</a></h2>
<p>Use the <tt class="docutils literal"><span class="pre">--ec</span></tt> option in conjunction with any aforementioned operation to do more stringent checking of the inputs&#8217; compliance to <a class="reference internal" href="#bedops"><em>bedops</em></a> requirements, including sorting checks, delimiter checks, among others.</p>
<p>To demonstrate, we can deliberately introduce a typo in dataset <cite>A</cite>, using the <tt class="docutils literal"><span class="pre">--ec</span></tt> option to try to catch it:</p>
<div class="highlight-python"><pre>$ bedops --ec --everything BEDFileA
May use bedops --help for more help.

Error: in BEDFileA
First column should not have spaces.  Consider 'chr1' vs. 'chr1 '.  These are different names.
See row: 3</pre>
</div>
<p>The typo introduced was the addition of a space within the third line of dataset <tt class="docutils literal"><span class="pre">A</span></tt>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Use of the <tt class="docutils literal"><span class="pre">--ec</span></tt> option will roughly <em>double</em> the running times of set operations, but it provides stringent error checking to ensure inputs and outputs are valid. <tt class="docutils literal"><span class="pre">--ec</span></tt> can help check problematic input and offers helpful hints for any needed corrections, when problems are detected.</p>
</div>
</div>
<div class="section" id="tips">
<h2>6.1.1.6. Tips<a class="headerlink" href="#tips" title="Permalink to this headline">¶</a></h2>
<div class="section" id="chaining-operations">
<h3>6.1.1.6.1. Chaining operations<a class="headerlink" href="#chaining-operations" title="Permalink to this headline">¶</a></h3>
<p>You can efficiently chain operations together, <em>e.g.</em>:</p>
<div class="highlight-python"><pre>$ bedops --range 50 --merge A | bedops --intersect - B &gt; answer.bed</pre>
</div>
<p>In this example, elements from <tt class="docutils literal"><span class="pre">A</span></tt> are padded 50 bases up- and downstream and merged, before intersecting with coordinates in <tt class="docutils literal"><span class="pre">B</span></tt>.</p>
</div>
<div class="section" id="sorting-inputs">
<h3>6.1.1.6.2. Sorting inputs<a class="headerlink" href="#sorting-inputs" title="Permalink to this headline">¶</a></h3>
<p>For unsorted input, be sure to first use <a class="reference internal" href="../file-management/sorting/sort-bed.html#sort-bed"><em>sort-bed</em></a> to presort the data stream before using with <a class="reference internal" href="#bedops"><em>bedops</em></a>. Unsorted input will not work properly with BEDOPS tools.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">If you will use an initially-unsorted file more than once, save the results of sorting. You only need to sort once! BEDOPS tools take in and export sorted data.</p>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../../index.html">
              <img class="logo" src="../../../_static/logo_with_label_v2.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">6.1.1. <cite>bedops</cite></a><ul>
<li><a class="reference internal" href="#inputs-and-outputs">6.1.1.1. Inputs and outputs</a><ul>
<li><a class="reference internal" href="#input">6.1.1.1.1. Input</a></li>
<li><a class="reference internal" href="#output">6.1.1.1.2. Output</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">6.1.1.2. Usage</a></li>
<li><a class="reference internal" href="#operations">6.1.1.3. Operations</a><ul>
<li><a class="reference internal" href="#everything-u-everything">6.1.1.3.1. Everything (-u, &#8211;everything)</a></li>
<li><a class="reference internal" href="#element-of-e-element-of">6.1.1.3.2. Element-of (-e, &#8211;element-of)</a></li>
<li><a class="reference internal" href="#not-element-of-n-not-element-of">6.1.1.3.3. Not-element-of (-n, &#8211;not-element-of)</a></li>
<li><a class="reference internal" href="#complement-c-complement">6.1.1.3.4. Complement (-c, &#8211;complement)</a></li>
<li><a class="reference internal" href="#difference-d-difference">6.1.1.3.5. Difference (-d, &#8211;difference)</a></li>
<li><a class="reference internal" href="#symmetric-difference-s-symmdiff">6.1.1.3.6. Symmetric difference (-s, &#8211;symmdiff)</a></li>
<li><a class="reference internal" href="#intersect-i-intersect">6.1.1.3.7. Intersect (-i, &#8211;intersect)</a></li>
<li><a class="reference internal" href="#merge-m-merge">6.1.1.3.8. Merge (-m, &#8211;merge)</a></li>
<li><a class="reference internal" href="#partition-p-partition">6.1.1.3.9. Partition (-p, &#8211;partition)</a></li>
<li><a class="reference internal" href="#per-chromosome-operations-chrom">6.1.1.3.10. Per-chromosome operations (&#8211;chrom)</a></li>
<li><a class="reference internal" href="#range-range">6.1.1.3.11. Range (&#8211;range)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#starch-support">6.1.1.4. Starch support</a></li>
<li><a class="reference internal" href="#error-checking-ec">6.1.1.5. Error checking (&#8211;ec)</a></li>
<li><a class="reference internal" href="#tips">6.1.1.6. Tips</a><ul>
<li><a class="reference internal" href="#chaining-operations">6.1.1.6.1. Chaining operations</a></li>
<li><a class="reference internal" href="#sorting-inputs">6.1.1.6.2. Sorting inputs</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../set-operations.html"
                        title="previous chapter">6.1. Set operations</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="bedextract.html"
                        title="next chapter">6.1.2. <cite>bedextract</cite></a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../../../_sources/content/reference/set-operations/bedops.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="bedextract.html" title="6.1.2. bedextract"
             >next</a> |</li>
        <li class="right" >
          <a href="../set-operations.html" title="6.1. Set operations"
             >previous</a> |</li>
        <li><a href="../../../index.html">BEDOPS 2.4.1 documentation</a> &raquo;</li>
          <li><a href="../../reference.html" >6. Reference</a> &raquo;</li>
          <li><a href="../set-operations.html" >6.1. Set operations</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.
    </div>
  </body>
</html>