<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>6.1.3. closest-features &mdash; BEDOPS v2.4.4</title>
    
    <link rel="stylesheet" href="../../../_static/better.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '2.4.4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://c328740.ssl.cf1.rackcdn.com/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
    <link rel="top" title="BEDOPS v2.4.4" href="../../../index.html" />
    <link rel="up" title="6.1. Set operations" href="../set-operations.html" />
    <link rel="next" title="6.1.4. Nested elements" href="nested-elements.html" />
    <link rel="prev" title="6.1.2. bedextract" href="bedextract.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  </head>
  <body role="document">
    <header id="pageheader"><h1><a href="../../../index.html ">
        BEDOPS v2.4.4
    </a></h1></header>
  <div class="related top">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="bedextract.html" title="Previous document">6.1.2. <cite>bedextract</cite></a>
        </li>
        <li>
          <a href="nested-elements.html" title="Next document">6.1.4. Nested elements</a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../../../index.html">BEDOPS v2.4.4</a></li>
      <li>
        <a href="../../reference.html">6. Reference</a>
      </li>
      <li>
        <a href="../set-operations.html">6.1. Set operations</a>
      </li> 
    </ul>
  </nav>
  </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="closest-features">
<span id="id1"></span><h1>6.1.3. <cite>closest-features</cite><a class="headerlink" href="#closest-features" title="Permalink to this headline">¶</a></h1>
<p>The <code class="docutils literal"><span class="pre">closest-features</span></code> program efficiently associates nearest features between two sorted inputs, based upon genomic distance measures.</p>
<p>An application of this tool in our own research is <a class="reference internal" href="../../usage-examples/distance-frequencies.html#distance-frequencies"><em>finding the nearest DNase hypersensitive sites</em></a> upstream and downstream from a given SNP, as well as signed distances. The <a class="reference internal" href="#closest-features"><em>closest-features</em></a> program can report both results.</p>
<p>As another example of what one can do with this utility, we can identify the closest transcriptional start site for a given putative replication origin. Suppose we have a sorted BED file named <code class="docutils literal"><span class="pre">TSS.bed</span></code> that contains all transcriptional start sites of all genes in some genome. Further, suppose that we have a set of measurements showing probable replication origins for the same species in a sorted BED file named <code class="docutils literal"><span class="pre">RepOrigins.bed</span></code>. The following command gives the closest TSS to each origin:</p>
<div class="highlight-python"><div class="highlight"><pre>$ closest-features --closest RepOrigins.bed TSS.bed
</pre></div>
</div>
<p>By default, the program will echo each entry from <code class="docutils literal"><span class="pre">RepOrigins.bed</span></code>, followed by the two closest elements in <code class="docutils literal"><span class="pre">TSS.bed</span></code> (the closest element to each side of the entry from <code class="docutils literal"><span class="pre">RepOrigins.bed</span></code>), with output columns separated by a pipe (<code class="docutils literal"><span class="pre">|</span></code>). With the <code class="docutils literal"><span class="pre">--shortest</span></code> option, the echoed entry from <code class="docutils literal"><span class="pre">RepOrigins.bed</span></code> and only the single nearest element in <code class="docutils literal"><span class="pre">TSS.bed</span></code> will be part of the output.</p>
<div class="section" id="inputs-and-outputs">
<h2>6.1.3.1. Inputs and outputs<a class="headerlink" href="#inputs-and-outputs" title="Permalink to this headline">¶</a></h2>
<div class="section" id="input">
<h3>6.1.3.1.1. Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="#closest-features"><em>closest-features</em></a> program takes two sorted BED files (a so-called <em>reference</em> file and a <em>map</em> file), as well as optional arguments for modifying behavior and outputs.</p>
<p>Alternatively, <a class="reference internal" href="#closest-features"><em>closest-features</em></a> can accept <a class="reference internal" href="../file-management/compression/starch.html#starch"><em>Starch-formatted archives</em></a> as inputs, with no need to extract archive data to intermediate BED files!</p>
<p>Support for common headers (such as UCSC track headers) is offered through the <code class="docutils literal"><span class="pre">--header</span></code> option. Headers are stripped from output.</p>
</div>
<div class="section" id="output">
<h3>6.1.3.1.2. Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="#closest-features"><em>closest-features</em></a> program returns summary data to standard output, which may include reference and nearest elements and distance values (depending on provided options).</p>
</div>
</div>
<div class="section" id="usage">
<h2>6.1.3.2. Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">--help</span></code> option describes the various operations and options available to the end user:</p>
<div class="highlight-python"><div class="highlight"><pre>closest-features
  citation: http://bioinformatics.oxfordjournals.org/content/28/14/1919.abstract
  version:  2.4.3
  authors:  Shane Neph &amp; Scott Kuehn

USAGE: closest-features [Process-Flags] &lt;input-file&gt; &lt;query-file&gt;
   All input files must be sorted per sort-bed.
   The program accepts BED and Starch file formats
   May use &#39;-&#39; for a file to indicate reading from standard input (BED format only).

   For every element in &lt;input-file&gt;, determine the two elements from &lt;query-file&gt; falling
     nearest to its left and right edges (See NOTES below).  By default, echo the &lt;input-file&gt;
     element, followed by those left and right elements found in &lt;query-file&gt;.

  Process Flags:
    --chrom &lt;chromosome&gt;   : Process data for given &lt;chromosome&gt; only.
    --closest              : Choose the closest element for output only.  Ties go the left element.
    --delim &lt;delim&gt;        : Change output delimiter from &#39;|&#39; to &lt;delim&gt; between columns (e.g. &#39;\t&#39;)
    --dist                 : Print the signed distances to the &lt;input-file&gt; element as additional
                               columns of output.  An overlapping element has a distance of 0.
    --ec                   : Error check all input files (slower).
    --header               : Accept headers (VCF, GFF, SAM, BED, WIG) in any input file.
    --help                 : Print this message and exit successfully.
    --no-overlaps          : Overlapping elements from &lt;query-file&gt; will not be reported.
    --no-ref               : Do not echo elements from &lt;input-file&gt;.
    --version              : Print program information.

  NOTES:
    If an element from &lt;query-file&gt; overlaps the &lt;input-file&gt; element, its distance is zero.
      An overlapping element take precedence over all non-overlapping elements.  This is true
      even when the overlapping element&#39;s edge-to-edge distance to the &lt;input-file&gt;&#39;s element
      is greater than the edge-to-edge distance from a non-overlapping element.
    Overlapping elements may be ignored completely (no precedence) with --no-overlaps.
    Elements reported as closest to the left and right edges are never the same.
    When no qualifying element from &lt;query-file&gt; exists as a closest feature, &#39;NA&#39; is reported.
</pre></div>
</div>
</div>
<div class="section" id="per-chromosome-operations-chrom">
<h2>6.1.3.3. Per-chromosome operations (&#8211;chrom)<a class="headerlink" href="#per-chromosome-operations-chrom" title="Permalink to this headline">¶</a></h2>
<p>All operations on inputs can be restricted to one chromosome, by adding the <code class="docutils literal"><span class="pre">--chrom</span> <span class="pre">&lt;val&gt;</span></code> operator.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">This option is highly useful for cluster-based work, where operations on large BED inputs can be split up by chromosome and pushed to separate cluster nodes.</p>
</div>
<p>To demonstrate the use of this option, we take two sample Starch-archived BED datasets <code class="docutils literal"><span class="pre">A</span></code> and <code class="docutils literal"><span class="pre">B</span></code> (refer to the <a class="reference internal" href="#closest-features-downloads"><em>Downloads</em></a> section for sample inputs) which contain regions from multiple chromosomes:</p>
<div class="highlight-python"><div class="highlight"><pre>$ unstarch A.starch
chr1    100     200     id-001A
chr1    400     500     id-002A
chr2    100     300     id-003A

$ unstarch B.starch
chr1    150     300     id-001B
chr1    500     600     id-002B
chr2    100     150     id-003B
chr2    180     500     id-004B
</pre></div>
</div>
<p>Now we want to ask, what is the closest element from <code class="docutils literal"><span class="pre">chr2</span></code> in <code class="docutils literal"><span class="pre">A</span></code>, to <code class="docutils literal"><span class="pre">chr2</span></code> elements in <code class="docutils literal"><span class="pre">B</span></code>:</p>
<div class="highlight-python"><div class="highlight"><pre>$ closest-features --chrom chr2 --closest A.starch B.starch
chr2    100     300     id-003A|chr2    100     150     id-003B
</pre></div>
</div>
<p>As we expect, element <code class="docutils literal"><span class="pre">id-003A</span></code> is closest to element <code class="docutils literal"><span class="pre">id-003B</span></code> between the two datasets.</p>
</div>
<div class="section" id="error-checking">
<h2>6.1.3.4. Error checking<a class="headerlink" href="#error-checking" title="Permalink to this headline">¶</a></h2>
<p>For performance reasons, no error checking of input is done, by default. Add <code class="docutils literal"><span class="pre">--ec</span></code> for stringent error checking and debugging purposes.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Using <code class="docutils literal"><span class="pre">--ec</span></code> will slow down analysis considerably. We recommend using this option to test and debug pipelines and then removing it for use in production.</p>
</div>
</div>
<div class="section" id="downloads">
<span id="closest-features-downloads"></span><h2>6.1.3.5. Downloads<a class="headerlink" href="#downloads" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Sample dataset <a class="reference download internal" href="../../../_downloads/reference_closestfeatures_a.starch"><code class="xref download docutils literal"><span class="pre">A</span></code></a></li>
<li>Sample dataset <a class="reference download internal" href="../../../_downloads/reference_closestfeatures_b.starch"><code class="xref download docutils literal"><span class="pre">B</span></code></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../../../index.html">
              <img class="logo" src="../../../_static/logo_with_label_v3.png" alt="Logo"/>
            </a></p>
  <h3><a href="../../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">6.1.3. <cite>closest-features</cite></a><ul>
<li><a class="reference internal" href="#inputs-and-outputs">6.1.3.1. Inputs and outputs</a><ul>
<li><a class="reference internal" href="#input">6.1.3.1.1. Input</a></li>
<li><a class="reference internal" href="#output">6.1.3.1.2. Output</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">6.1.3.2. Usage</a></li>
<li><a class="reference internal" href="#per-chromosome-operations-chrom">6.1.3.3. Per-chromosome operations (&#8211;chrom)</a></li>
<li><a class="reference internal" href="#error-checking">6.1.3.4. Error checking</a></li>
<li><a class="reference internal" href="#downloads">6.1.3.5. Downloads</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="bedextract.html"
                        title="previous chapter">6.1.2. <cite>bedextract</cite></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="nested-elements.html"
                        title="next chapter">6.1.4. Nested elements</a></p>
<div class="sourcelink">
  <a href="../../../_sources/content/reference/set-operations/closest-features.txt" rel="nofollow">
    Show page source
  </a>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>

        </div>
      </div>
      <div class="clearer"></div>
    </div>
  <div class="related bottom">
  <nav id="rellinks">
    <ul>
        <li>
          &larr;
          <a href="bedextract.html" title="Previous document">6.1.2. <cite>bedextract</cite></a>
        </li>
        <li>
          <a href="nested-elements.html" title="Next document">6.1.4. Nested elements</a>
          &rarr;
        </li>
    </ul>
  </nav>
  <nav id="breadcrumbs">
    <ul>
      <li><a href="../../../index.html">BEDOPS v2.4.4</a></li>
      <li>
        <a href="../../reference.html">6. Reference</a>
      </li>
      <li>
        <a href="../set-operations.html">6.1. Set operations</a>
      </li> 
    </ul>
  </nav>
  </div>
  <footer id="pagefooter">&copy; 2011-2015, Shane Neph, Alex Reynolds.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a>
      1.3b2
        with the <a href="http://github.com/irskep/sphinx-better-theme">
          better</a> theme.

  </footer>

  
  </body>
</html>